%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{times}
\expandafter\ifx\csname T@LGR\endcsname\relax
\else
% LGR was declared as font encoding
  \substitutefont{LGR}{\rmdefault}{cmr}
  \substitutefont{LGR}{\sfdefault}{cmss}
  \substitutefont{LGR}{\ttdefault}{cmtt}
\fi
\expandafter\ifx\csname T@X2\endcsname\relax
  \expandafter\ifx\csname T@T2A\endcsname\relax
  \else
  % T2A was declared as font encoding
    \substitutefont{T2A}{\rmdefault}{cmr}
    \substitutefont{T2A}{\sfdefault}{cmss}
    \substitutefont{T2A}{\ttdefault}{cmtt}
  \fi
\else
% X2 was declared as font encoding
  \substitutefont{X2}{\rmdefault}{cmr}
  \substitutefont{X2}{\sfdefault}{cmss}
  \substitutefont{X2}{\ttdefault}{cmtt}
\fi


\usepackage[Bjarne]{fncychap}
\usepackage{sphinx}

\fvset{fontsize=\small}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}


\usepackage{sphinxmessages}
\setcounter{tocdepth}{1}



\title{IPTK}
\date{Sep 24, 2020}
\release{0.3.1}
\author{Hesham ElAbd}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{index::doc}}


Analysis, visualize, compare and integrate Immunopeptidomics data !
\begin{quote}
\begin{description}
\item[{\textgreater{}\textgreater{}\textgreater{}print(‘Hello World!’)}] \leavevmode
Hello World

\end{description}
\end{quote}


\chapter{Introduction:}
\label{\detokenize{index:introduction}}
will write it later


\section{Guide}
\label{\detokenize{index:guide}}

\subsection{License}
\label{\detokenize{license:license}}\label{\detokenize{license::doc}}
Will be written here


\subsection{Contact}
\label{\detokenize{license:contact}}
for further question and communication please contact \sphinxhref{mailto:h.elabd@ikmb.uni-kiel.de}{h.elabd@ikmb.uni\sphinxhyphen{}kiel.de}


\subsection{Get Started!}
\label{\detokenize{get_started:get-started}}\label{\detokenize{get_started::doc}}
To get started with using the library check the Interactive Tutorials available at \sphinxurl{https://github.com/ikmb/iptoolkit/tree/master/Tutorials}


\subsection{IPTK}
\label{\detokenize{modules:iptk}}\label{\detokenize{modules::doc}}

\subsubsection{IPTK package}
\label{\detokenize{IPTK:iptk-package}}\label{\detokenize{IPTK::doc}}

\paragraph{Subpackages}
\label{\detokenize{IPTK:subpackages}}

\subparagraph{IPTK.Analysis package}
\label{\detokenize{IPTK.Analysis:iptk-analysis-package}}\label{\detokenize{IPTK.Analysis::doc}}

\subparagraph{Submodules}
\label{\detokenize{IPTK.Analysis:submodules}}

\subparagraph{IPTK.Analysis.AnalysisFunction module}
\label{\detokenize{IPTK.Analysis:module-IPTK.Analysis.AnalysisFunction}}\label{\detokenize{IPTK.Analysis:iptk-analysis-analysisfunction-module}}\index{module@\spxentry{module}!IPTK.Analysis.AnalysisFunction@\spxentry{IPTK.Analysis.AnalysisFunction}}\index{IPTK.Analysis.AnalysisFunction@\spxentry{IPTK.Analysis.AnalysisFunction}!module@\spxentry{module}}
The module contain a collection of analysis function that can be used by the methods of
the classes defined in the classes module.
\index{compute\_binary\_distance() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{compute\_binary\_distance()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.compute_binary_distance}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{compute\_binary\_distance}}}{\emph{\DUrole{n}{peptides}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}, \emph{\DUrole{n}{dist\_func}\DUrole{p}{:} \DUrole{n}{Callable}}}{{ $\rightarrow$ numpy.ndarray}}
compare the distance between every pair of peptides in a collection of peptides. 
@param: peptides: a collection of peptides sequences.
@param: dist\_func: function to compute the distance between each pair of peptides. 
@note:
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptides}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a collection of peptides sequences.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{dist\_func}} (\sphinxstyleliteralemphasis{\sphinxupquote{Callable}}) \textendash{} a function to compute the distance between each pair of peptides.

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeError}} \textendash{} make sure that the dist\_function is suitable with the peptides which might have different lengths.

\item[{Returns}] \leavevmode
the distance between each pair of peptides in the provided list of peptides

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_change\_in\_protein\_representation() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{compute\_change\_in\_protein\_representation()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.compute_change_in_protein_representation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{compute\_change\_in\_protein\_representation}}}{\emph{\DUrole{n}{mapped\_prot\_cond1}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}, \emph{\DUrole{n}{mapped\_prot\_cond2}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}}{{ $\rightarrow$ float}}
Compute the change in the protein representation between two conditions, by computing 
the difference in the area under the curve, AUC.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{mapped\_prot\_cond1}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} a mapped protein instance containing the protein coverage in the first condition

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{mapped\_prot\_cond2}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} a mapped protein instance containing the protein coverage in the second condition

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the provided pair of proteins is of different length

\item[{Returns}] \leavevmode
the difference in the area under the coverage curve between the two experiments.

\item[{Return type}] \leavevmode
float

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_difference\_in\_representation() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{compute\_difference\_in\_representation()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.compute_difference_in_representation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{compute\_difference\_in\_representation}}}{\emph{\DUrole{n}{mapped\_prot\_cond1}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}, \emph{\DUrole{n}{mapped\_prot\_cond2}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}}{{ $\rightarrow$ numpy.ndarray}}
return the difference in the representation of a protein between two conditions
by substracting the coverage of the first protein from the second proteins.

@param: mapped\_prot\_cond1: a mapped protein instance containing the protein coverage in the first condition
@param: mapped\_prot\_cond2: a mapped protein instance containing the protein coverage in the second condition
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{mapped\_prot\_cond1}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} a mapped protein instance containing the protein coverage in the first condition

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{mapped\_prot\_cond2}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} a mapped protein instance containing the protein coverage in the second condition

\end{itemize}

\item[{Returns}] \leavevmode
an array that shows the difference in coverage between the two proteins at each amino acid position.

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_expression\_correlation() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{compute\_expression\_correlation()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.compute_expression_correlation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{compute\_expression\_correlation}}}{\emph{\DUrole{n}{exp1}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}, \emph{\DUrole{n}{exp2}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}}{{ $\rightarrow$ float}}
compute the correlation in the gene expression between two experiments by constructing a union
of all the proteins expressed in the first and second experiments, extract the gene expression 
of these genes and then compute the correlation using SciPy stat module.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp1}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Experiment}}}}}) \textendash{} The first experimental object

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp2}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Experiment}}}}}) \textendash{} he second experimental object

\end{itemize}

\item[{Returns}] \leavevmode
the correlation in gene expression of the proteins inferred in the provided pair of experiment

\item[{Return type}] \leavevmode
float

\end{description}\end{quote}

\end{fulllineitems}

\index{download\_structure\_file() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{download\_structure\_file()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.download_structure_file}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{download\_structure\_file}}}{\emph{\DUrole{n}{pdb\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ None}}
Download PDB/mmCIF file containing the pbd\_id from PDB using BioPython library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pdb\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the protein id in protein databank

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_binnary\_peptide\_overlap() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{get\_binnary\_peptide\_overlap()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.get_binnary_peptide_overlap}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{get\_binnary\_peptide\_overlap}}}{\emph{\DUrole{n}{exp1}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}, \emph{\DUrole{n}{exp2}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
compare the peptide overlap between two experimental objects.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp1}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Experiment}}}}}) \textendash{} an instance of class Experiment.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp2}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Experiment}}}}}) \textendash{} an instance of class Experiment.

\end{itemize}

\item[{Returns}] \leavevmode
a list of peptides that have been identified in both experiments.

\item[{Return type}] \leavevmode
Peptides

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_binnary\_protein\_overlap() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{get\_binnary\_protein\_overlap()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.get_binnary_protein_overlap}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{get\_binnary\_protein\_overlap}}}{\emph{\DUrole{n}{exp1}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}, \emph{\DUrole{n}{exp2}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
compare the protein overlap between two experimental objects.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp1}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Experiment}}}}}) \textendash{} an instance of class Experiment.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp2}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{Experiment}}}}}) \textendash{} an instance of class Experiment.

\end{itemize}

\item[{Returns}] \leavevmode
a list of proteins that have been identified in both experiments.

\item[{Return type}] \leavevmode
Proteins

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_sequence\_motif() (in module IPTK.Analysis.AnalysisFunction)@\spxentry{get\_sequence\_motif()}\spxextra{in module IPTK.Analysis.AnalysisFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Analysis:IPTK.Analysis.AnalysisFunction.get_sequence_motif}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Analysis.AnalysisFunction.}}\sphinxbfcode{\sphinxupquote{get\_sequence\_motif}}}{\emph{\DUrole{n}{peptides}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}, \emph{\DUrole{n}{temp\_dir}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}./TEMP\_DIR\textquotesingle{}}}, \emph{\DUrole{n}{verbose}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{meme\_params}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}}{{ $\rightarrow$ None}}
compute the sequences motif from a collection of peptide sequences using meme software.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptides}} (\sphinxstyleliteralemphasis{\sphinxupquote{Peptides}}) \textendash{} a list of string containing the peptide sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{temp\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} he temp directory to write temp\sphinxhyphen{}files to it, defaults to “./TEMP\_DIR”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to print the output of the motif discovery tool to the stdout, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{meme\_params}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object that contain meme controlling parameters, defaults to \{\}

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{FileNotFoundError}} \textendash{} incase meme is not installed or could not be found in the path!

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} incase the peptides have different length!

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{Module contents}
\label{\detokenize{IPTK.Analysis:module-IPTK.Analysis}}\label{\detokenize{IPTK.Analysis:module-contents}}\index{module@\spxentry{module}!IPTK.Analysis@\spxentry{IPTK.Analysis}}\index{IPTK.Analysis@\spxentry{IPTK.Analysis}!module@\spxentry{module}}

\subparagraph{IPTK.Classes package}
\label{\detokenize{IPTK.Classes:iptk-classes-package}}\label{\detokenize{IPTK.Classes::doc}}

\subparagraph{Submodules}
\label{\detokenize{IPTK.Classes:submodules}}

\subparagraph{IPTK.Classes.Database module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.Database}}\label{\detokenize{IPTK.Classes:iptk-classes-database-module}}\index{module@\spxentry{module}!IPTK.Classes.Database@\spxentry{IPTK.Classes.Database}}\index{IPTK.Classes.Database@\spxentry{IPTK.Classes.Database}!module@\spxentry{module}}
This submodule define a collection of container classes that are used through the library
\index{CellularLocationDB (class in IPTK.Classes.Database)@\spxentry{CellularLocationDB}\spxextra{class in IPTK.Classes.Database}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Database.}}\sphinxbfcode{\sphinxupquote{CellularLocationDB}}}{\emph{\DUrole{n}{path2data}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{sep}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}\textbackslash{}t\textquotesingle{}}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

The class provides an API to access the cellular location information from a database   the follow the structure of the human Proteome Atlas sub\sphinxhyphen{}cellular location database. See \sphinxurl{https://www.proteinatlas.org/about/download}    for more details.
\index{add\_to\_database() (IPTK.Classes.Database.CellularLocationDB method)@\spxentry{add\_to\_database()}\spxextra{IPTK.Classes.Database.CellularLocationDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB.add_to_database}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_to\_database}}}{\emph{\DUrole{n}{genes\_to\_add}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB}]{\sphinxcrossref{IPTK.Classes.Database.CellularLocationDB}}}}}}{{ $\rightarrow$ None}}
add the the location of more proteins to the database.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{genes\_to\_add}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{CellularLocationDB}}}}}) \textendash{} a CellularLocationDB instance containing the genes that shall be added to the database.

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the genes to add to the database are already defined in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeError}} \textendash{} Incase any other error has been encountered while merging the tables.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_approved\_location() (IPTK.Classes.Database.CellularLocationDB method)@\spxentry{get\_approved\_location()}\spxextra{IPTK.Classes.Database.CellularLocationDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB.get_approved_location}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_approved\_location}}}{\emph{\DUrole{n}{gene\_id}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}, \emph{\DUrole{n}{gene\_name}\DUrole{o}{=}\DUrole{default_value}{None}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return the location of the provided gene id or gene name
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the id of the gene of interest , defaults to None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the name of the gene of interest , defaults to None

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if both gene\_id and gene\_name are None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_id is None and gene\_name is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_name is None and gene\_id is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeError}} \textendash{} incase some error was encountered while running retriving the elements from the database

\end{itemize}

\item[{Returns}] \leavevmode
the approved location where the protein the corresponds to the provided name or id is located.

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_gene\_names() (IPTK.Classes.Database.CellularLocationDB method)@\spxentry{get\_gene\_names()}\spxextra{IPTK.Classes.Database.CellularLocationDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB.get_gene_names}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_gene\_names}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return a list of all gene names in the dataset
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the names of all genes in the database

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_genes() (IPTK.Classes.Database.CellularLocationDB method)@\spxentry{get\_genes()}\spxextra{IPTK.Classes.Database.CellularLocationDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB.get_genes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_genes}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return a list of all gene ids in the dataset
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
all genes ids currently defined in the database

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_go\_names() (IPTK.Classes.Database.CellularLocationDB method)@\spxentry{get\_go\_names()}\spxextra{IPTK.Classes.Database.CellularLocationDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB.get_go_names}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_go\_names}}}{\emph{\DUrole{n}{gene\_id}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}, \emph{\DUrole{n}{gene\_name}\DUrole{o}{=}\DUrole{default_value}{None}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return the location of the provided gene id or gene name
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the id of the gene of interest , defaults to None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the name of the gene of interest , defaults to None

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if both gene\_id and gene\_name are None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_id is None and gene\_name is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_name is None and gene\_id is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeError}} \textendash{} incase some error was encountered while running retriving the elements from the database

\end{itemize}

\item[{Returns}] \leavevmode
the gene ontology, GO,  location where the protein the corresponds to the provided name or id is located.

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_main\_location() (IPTK.Classes.Database.CellularLocationDB method)@\spxentry{get\_main\_location()}\spxextra{IPTK.Classes.Database.CellularLocationDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB.get_main_location}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_main\_location}}}{\emph{\DUrole{n}{gene\_id}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}, \emph{\DUrole{n}{corresponds}\DUrole{o}{=}\DUrole{default_value}{None}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return the main location(s) of the provided gene id or gene name. 
If both gene Id and gene name are provided, both gene\_id has a higher precedence
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the id of the gene of interest , defaults to None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the name of the gene of interest , defaults to None

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if both gene\_id and gene\_name are None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_id is None and gene\_name is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_name is None and gene\_id is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeError}} \textendash{} incase some error was encountered while running retriving the elements from the database

\end{itemize}

\item[{Returns}] \leavevmode
the main location where the protein the corresponds to the provided name or id is located.

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_table() (IPTK.Classes.Database.CellularLocationDB method)@\spxentry{get\_table()}\spxextra{IPTK.Classes.Database.CellularLocationDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB.get_table}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_table}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return the instance table
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the location table of the instance.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{GeneExpressionDB (class in IPTK.Classes.Database)@\spxentry{GeneExpressionDB}\spxextra{class in IPTK.Classes.Database}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Database.}}\sphinxbfcode{\sphinxupquote{GeneExpressionDB}}}{\emph{\DUrole{n}{path2data}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}https://www.proteinatlas.org/download/rna\_tissue\_consensus.tsv.zip\textquotesingle{}}}, \emph{\DUrole{n}{sep}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}\textbackslash{}t\textquotesingle{}}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

provides an API to access gene expression data stored in table that follows the same structure as 
the Human proteome Atlas Normalized RNA Expression see  \sphinxurl{https://www.proteinatlas.org/about/download} for more details
\index{get\_expression() (IPTK.Classes.Database.GeneExpressionDB method)@\spxentry{get\_expression()}\spxextra{IPTK.Classes.Database.GeneExpressionDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB.get_expression}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_expression}}}{\emph{\DUrole{n}{gene\_name}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}, \emph{\DUrole{n}{gene\_id}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
Return a table summarizing the expression of the provided gene name or gene id accross different tissues.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the id of the gene of interest , defaults to None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the name of the gene of interest , defaults to None

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if both gene\_id and gene\_name are None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_id is None and gene\_name is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if gene\_name is None and gene\_id is not in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeError}} \textendash{} incase some error was encountered while running retriving the elements from the database

\end{itemize}

\item[{Returns}] \leavevmode
A table summarizing the expression of the provided gene accross all tissues in the database

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_expression\_in\_tissue() (IPTK.Classes.Database.GeneExpressionDB method)@\spxentry{get\_expression\_in\_tissue()}\spxextra{IPTK.Classes.Database.GeneExpressionDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB.get_expression_in_tissue}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_expression\_in\_tissue}}}{\emph{\DUrole{n}{tissue\_name}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return the expression profile of the provided tissue
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{tissue\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the tissue

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} incase the provided tissue is not provided in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeError}} \textendash{} in case any error was encountered while generating the expression profile.

\end{itemize}

\item[{Returns}] \leavevmode
a table summarizing the expression of all genes in the provided tissue.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_gene\_names() (IPTK.Classes.Database.GeneExpressionDB method)@\spxentry{get\_gene\_names()}\spxextra{IPTK.Classes.Database.GeneExpressionDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB.get_gene_names}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_gene\_names}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return a list of the UNIQUE gene names currently in the database
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list of the UNIQUE gene names currently in the database

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_genes() (IPTK.Classes.Database.GeneExpressionDB method)@\spxentry{get\_genes()}\spxextra{IPTK.Classes.Database.GeneExpressionDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB.get_genes}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_genes}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return a list of the UNIQUE gene ids currently in the database
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list of the UNIQUE gene ids currently in the database

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_table() (IPTK.Classes.Database.GeneExpressionDB method)@\spxentry{get\_table()}\spxextra{IPTK.Classes.Database.GeneExpressionDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB.get_table}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_table}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return a table containing the expression value of all the genes accross all tissues in the current instance
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The expression of all genes accross all tissues in the database.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_tissues() (IPTK.Classes.Database.GeneExpressionDB method)@\spxentry{get\_tissues()}\spxextra{IPTK.Classes.Database.GeneExpressionDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB.get_tissues}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_tissues}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return a list of the tissues in the current database
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list containing the names of the UNIQUE tissues in the database.

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{OrganismDB (class in IPTK.Classes.Database)@\spxentry{OrganismDB}\spxextra{class in IPTK.Classes.Database}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.OrganismDB}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Database.}}\sphinxbfcode{\sphinxupquote{OrganismDB}}}{\emph{\DUrole{n}{path2Fasta}\DUrole{p}{:} \DUrole{n}{str}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

Extract information about the source organsim of a collection of protein sequences 
from a fasta file and provides an API to query the results.     The function expect the input fasta file to have header written in the UNIPROT format.
\index{get\_number\_protein\_per\_organism() (IPTK.Classes.Database.OrganismDB method)@\spxentry{get\_number\_protein\_per\_organism()}\spxextra{IPTK.Classes.Database.OrganismDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.OrganismDB.get_number_protein_per_organism}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_number\_protein\_per\_organism}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
provides a table containing the number of proteins per organism.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a table containing the number of proteins per organism

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_org() (IPTK.Classes.Database.OrganismDB method)@\spxentry{get\_org()}\spxextra{IPTK.Classes.Database.OrganismDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.OrganismDB.get_org}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_org}}}{\emph{\DUrole{n}{prot\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ str}}
return the parent organism of the provided proteins
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{prot\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the id of the protein of interest

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} incase the provided identifier is not in the database

\item[{Returns}] \leavevmode
the name of the parent organism, i.e. the source organism.

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_unique\_orgs() (IPTK.Classes.Database.OrganismDB method)@\spxentry{get\_unique\_orgs()}\spxextra{IPTK.Classes.Database.OrganismDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.OrganismDB.get_unique_orgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_unique\_orgs}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
get the number of unique organisms in the database
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list of all unique organisms in the current instance

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{SeqDB (class in IPTK.Classes.Database)@\spxentry{SeqDB}\spxextra{class in IPTK.Classes.Database}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.SeqDB}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Database.}}\sphinxbfcode{\sphinxupquote{SeqDB}}}{\emph{\DUrole{n}{path2fasta}\DUrole{p}{:} \DUrole{n}{str}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

load a fasta file and constructs a lock up dictionary where sequence ids are  keys and sequences are values.
\index{get\_seq() (IPTK.Classes.Database.SeqDB method)@\spxentry{get\_seq()}\spxextra{IPTK.Classes.Database.SeqDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.SeqDB.get_seq}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_seq}}}{\emph{\DUrole{n}{protein\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ str}}
returns the corresponding sequence if the provided protein\sphinxhyphen{}id is defined in the database.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{protein\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The protein id to retrive its sequence.

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} If the provided protein does not exist in the database

\item[{Returns}] \leavevmode
the protein sequence

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_sequence() (IPTK.Classes.Database.SeqDB method)@\spxentry{has\_sequence()}\spxextra{IPTK.Classes.Database.SeqDB method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Database.SeqDB.has_sequence}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_sequence}}}{\emph{\DUrole{n}{sequence\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}
check if the provided sequence id is an element of the database or not
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{sequence\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The id of the sequence

\item[{Returns}] \leavevmode
True if the database has this id, False otherwise.

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.Experiment module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.Experiment}}\label{\detokenize{IPTK.Classes:iptk-classes-experiment-module}}\index{module@\spxentry{module}!IPTK.Classes.Experiment@\spxentry{IPTK.Classes.Experiment}}\index{IPTK.Classes.Experiment@\spxentry{IPTK.Classes.Experiment}!module@\spxentry{module}}
This module provides an abstraction for an IP experiment.
\index{Experiment (class in IPTK.Classes.Experiment)@\spxentry{Experiment}\spxextra{class in IPTK.Classes.Experiment}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Experiment.}}\sphinxbfcode{\sphinxupquote{Experiment}}}{\emph{\DUrole{n}{proband}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Proband.Proband}]{\sphinxcrossref{IPTK.Classes.Proband.Proband}}}}}, \emph{\DUrole{n}{hla\_set}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet}]{\sphinxcrossref{IPTK.Classes.HLASet.HLASet}}}}}, \emph{\DUrole{n}{tissue}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Tissue.Tissue}]{\sphinxcrossref{IPTK.Classes.Tissue.Tissue}}}}}, \emph{\DUrole{n}{database}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.SeqDB}]{\sphinxcrossref{IPTK.Classes.Database.SeqDB}}}}}, \emph{\DUrole{n}{ident\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

A representation of an immunopeptidomic experiment.
\index{add\_org\_info() (IPTK.Classes.Experiment.Experiment method)@\spxentry{add\_org\_info()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.add_org_info}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_org\_info}}}{\emph{\DUrole{n}{prot2org}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}}}}{{ $\rightarrow$ None}}
annotated the inferred proteins with their source organism
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{prot2org}} (\sphinxstyleliteralemphasis{\sphinxupquote{ProteinSource}}) \textendash{} a dict that contain the protein id as keys and its source organism as values           and add this info to each protein inferred in the current experiment.

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{RuntimeWarning}} \textendash{} If the provided dictionary does cover all proteins in the experimental object.

\end{description}\end{quote}

\end{fulllineitems}

\index{annotate\_proteins() (IPTK.Classes.Experiment.Experiment method)@\spxentry{annotate\_proteins()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.annotate_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{annotate\_proteins}}}{\emph{\DUrole{n}{organisms\_db}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.OrganismDB}]{\sphinxcrossref{IPTK.Classes.Database.OrganismDB}}}}}}{{ $\rightarrow$ None}}
Extract the parent organisms of each protein in the experiment from an organism database instance.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{organisms\_db}} ({\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.OrganismDB}]{\sphinxcrossref{\sphinxstyleliteralemphasis{\sphinxupquote{OrganismDB}}}}}) \textendash{} an OrgansimDB instance that will be used to annotate the proteins          identified in the experiment.

\end{description}\end{quote}

\end{fulllineitems}

\index{drop\_peptide\_belong\_to\_org() (IPTK.Classes.Experiment.Experiment method)@\spxentry{drop\_peptide\_belong\_to\_org()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.drop_peptide_belong_to_org}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{drop\_peptide\_belong\_to\_org}}}{\emph{\DUrole{n}{org}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ None}}
Drop the all the peptides that belong to a user provided organism.              Note that, this function will IRREVERSIBLY remove the peptide from the experimental object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{org}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the organims name

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_binarized\_results() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_binarized\_results()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_binarized_results}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_binarized\_results}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}numpy.ndarray\DUrole{p}{{]}}}}
Return a list of NumPy arrays where each array represents a child peptide, parent protein mapped pair.
Please note that, The function treat each peptide\sphinxhyphen{}protein pair individually, that is if two peptides originating from the same protein, 
it treat them independently and the same protein will be represented twice with the two different peptides. Incase an integrative mapping is needed,
the function @get\_integrated\_binarized\_results@ shall be used.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list of NumPy arrays containing the mapping between each peptide protein pair.

\item[{Return type}] \leavevmode
MappedProtein

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_c\_terminal\_flanked\_seqs() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_c\_terminal\_flanked\_seqs()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_c_terminal_flanked_seqs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_c\_terminal\_flanked\_seqs}}}{\emph{\DUrole{n}{flank\_length}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}
return the c\sphinxhyphen{}terminal flanking sequences
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{flank\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the length of the peptide downstream of the C\sphinxhyphen{}terminal of the peptide

\item[{Returns}] \leavevmode
a list sequences contain the N\sphinxhyphen{}terminal flanking sequence for each peptide in the instance.

\item[{Return type}] \leavevmode
Peptides

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_experiment\_reference\_tissue\_expression() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_experiment\_reference\_tissue\_expression()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_experiment_reference_tissue_expression}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_experiment\_reference\_tissue\_expression}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return the reference gene expression for the current tissue
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A table that contain the expression value for ALL the genes in the instance Tissue

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_expression\_of\_parent\_proteins() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_expression\_of\_parent\_proteins()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_expression_of_parent_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_expression\_of\_parent\_proteins}}}{\emph{\DUrole{n}{non\_mapped\_dval}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return a table containing the expression of the proteins inferred in the current experiment from the current tissue.
This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{non\_mapped\_dval}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} A default value to be added incase the parent protein is not define in the expression database, defaults to \sphinxhyphen{}1

\item[{Returns}] \leavevmode
a table that contain the expression of the protein inferred in the database

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_flanked\_peptides() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_flanked\_peptides()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_flanked_peptides}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_flanked\_peptides}}}{\emph{\DUrole{n}{flank\_length}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
returns a list of sequences containing the peptides identified in the experiment padded with
the flanking regions from all the parents of each peptide.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{flank\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the length of the flanking region

\item[{Returns}] \leavevmode
a list of the peptides + the flanking region.

\item[{Return type}] \leavevmode
Sequences

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_go\_location\_id\_parent\_proteins() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_go\_location\_id\_parent\_proteins()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_go_location_id_parent_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_go\_location\_id\_parent\_proteins}}}{\emph{\DUrole{n}{not\_mapped\_val}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}UNK\textquotesingle{}}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
retrun the gene ontology,GO, location terms for all the identified proteins.

@brief: 
@param: not\_mapped\_val: the default value to return incase the GO term of the protein can not be extracted. 
@note: This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{not\_mapped\_val}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default value to return incase the GO term of the protein can not be extracted, defaults to ‘UNK’

\item[{Returns}] \leavevmode
A table that contain the GO\sphinxhyphen{}location term for each protein in the current instance.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_hla\_allele() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_hla\_allele()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_hla_allele}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_hla\_allele}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the set of HLA alleles from which the instance peptides have been eluted

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_hla\_class() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_hla\_class()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_hla_class}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_hla\_class}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the HLA class

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_main\_sub\_cellular\_location\_of\_parent\_proteins() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_main\_sub\_cellular\_location\_of\_parent\_proteins()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_main_sub_cellular_location_of_parent_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_main\_sub\_cellular\_location\_of\_parent\_proteins}}}{\emph{\DUrole{n}{not\_mapped\_val}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}UNK\textquotesingle{}}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
retrun the main cellular location for the identified proteins.
This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{not\_mapped\_val}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default value to return incase the location of a protein can not be extracted, defaults to ‘UNK’

\item[{Returns}] \leavevmode
A table that contain the main cellular compartment for each protein in the current instance.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_mapped\_protein() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_mapped\_protein()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_mapped_protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_mapped\_protein}}}{\emph{\DUrole{n}{pro\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ numpy.ndarray}}
return an NumPy array of shape 1 x protein length where each number in the array represents 
the total number of peptides identified in the experiment that have originated from the said position 
in the protein.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pro\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the protein id

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if the provided protein id was inferred from the current experiment

\item[{Returns}] \leavevmode
a NumPy array that contain the coverage of the protein.

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_mapped\_proteins() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_mapped\_proteins()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_mapped_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_mapped\_proteins}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }List\DUrole{p}{{[}}numpy.ndarray\DUrole{p}{{]}}\DUrole{p}{{]}}}}
return a dictionary of all the proteins identified in the current experiment with all inferred
peptides mapped to them.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a dictionary that contain the mapped proteins for all the proteins in the current instance.

\item[{Return type}] \leavevmode
MappedProteins

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_mono\_parent\_peptides() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_mono\_parent\_peptides()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_mono_parent_peptides}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_mono\_parent\_peptides}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}
return a list of peptides that have only one parent protein
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
list of peptide instance

\item[{Return type}] \leavevmode
Peptides

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_n\_terminal\_flanked\_seqs() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_n\_terminal\_flanked\_seqs()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_n_terminal_flanked_seqs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_n\_terminal\_flanked\_seqs}}}{\emph{\DUrole{n}{flank\_length}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}
return the n\sphinxhyphen{}terminal flanking sequences
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{flank\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the length of the flanking region upstream of the N\sphinxhyphen{}terminal of the peptide

\item[{Returns}] \leavevmode
a list sequences contain the N\sphinxhyphen{}terminal flanking sequence for each peptide in the instance.

\item[{Return type}] \leavevmode
Peptides

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_negative\_example() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_negative\_example()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_negative_example}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_negative\_example}}}{\emph{\DUrole{n}{fold}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{2}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
generate negative examples, i.e., non\sphinxhyphen{}bounding peptides from the proteins identified in the current experiment.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{fold}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the number of negative example to generate relative to the number of unique identified peptides, defaults to 2

\item[{Returns}] \leavevmode
list of non\sphinxhyphen{}presented peptides from all inferred proteins.

\item[{Return type}] \leavevmode
Sequences

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_num\_peptide\_expression\_table() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_num\_peptide\_expression\_table()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_num_peptide_expression_table}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_num\_peptide\_expression\_table}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}~\begin{description}
\item[{Get a table that contain the id of all parent proteins, number of peptide per\sphinxhyphen{}proteins and the expression value }] \leavevmode
of these parent transcripts. Please note, this method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.

\end{description}
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the number of peptides per protein table

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_num\_peptide\_per\_go\_term() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_num\_peptide\_per\_go\_term()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_num_peptide_per_go_term}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_num\_peptide\_per\_go\_term}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
retrun the number of peptides per each GO\sphinxhyphen{}Term 
:return: A table that has two columns, namely, GO\sphinxhyphen{}Terms and Counts. 
:rtype: pd.DataFrame

\end{fulllineitems}

\index{get\_num\_peptide\_per\_location() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_num\_peptide\_per\_location()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_num_peptide_per_location}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_num\_peptide\_per\_location}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
retrun the number of peptides obtained from proteins localized to different sub\sphinxhyphen{}cellular compartments
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A table that has two columns, namely, Compartment and Counts.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_number\_of\_children() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_number\_of\_children()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_number_of_children}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_number\_of\_children}}}{\emph{\DUrole{n}{pro\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ int}}
return the number of children, i.e. number of peptides belonging to a parent protein
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pro\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the id of the parent protein

\item[{Returns}] \leavevmode
the number of peptides

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_number\_of\_proteins\_per\_compartment() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_number\_of\_proteins\_per\_compartment()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_compartment}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_number\_of\_proteins\_per\_compartment}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
get the number of proteins from each compartment
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A table that has two columns, namely, Compartment and Counts.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_number\_of\_proteins\_per\_go\_term() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_number\_of\_proteins\_per\_go\_term()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_go_term}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_number\_of\_proteins\_per\_go\_term}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
get the number of proteins from each GO\sphinxhyphen{}Term
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A table that has two columns, namely, GO\sphinxhyphen{}Terms and Counts.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_orgs() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_orgs()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_orgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_orgs}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
return a list containing the UNIQUE organisms identified in the current experiment
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
list of all UNIQUE organisms inferred from the inferred proteins.

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptide() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_peptide()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_peptide}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptide}}}{\emph{\DUrole{n}{pep\_seq}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ {\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}}}
return a peptide instance corresponding to the user provided peptide sequence.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pep\_seq}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the peptide sequence

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if the peptide sequence has not been inferred from the current database.

\item[{Returns}] \leavevmode
the peptide instance with the coresponding sequence

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{Peptide}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptide\_number\_parent() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_peptide\_number\_parent()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_peptide_number_parent}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptide\_number\_parent}}}{\emph{\DUrole{n}{ascending}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return a pandas dataframe with the peptide sequence in the first columns and the 
number of parent proteins in the second column.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ascending}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} ascending sort the peptide by their number of parent proteins, defaults to False

\item[{Returns}] \leavevmode
the number of parents for each peptide

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptides() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_peptides()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_peptides}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptides}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a set of all the peptide stored in the experimental object

\item[{Return type}] \leavevmode
Peptides

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptides\_length() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_peptides\_length()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_peptides_length}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptides\_length}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}int\DUrole{p}{{]}}}}
return a list containing the length of each unique peptide in the database.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
peptides’ lengths

\item[{Return type}] \leavevmode
List{[}int{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptides\_per\_organism() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_peptides\_per\_organism()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_peptides_per_organism}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptides\_per\_organism}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return a pandas dataframe that contain the count of peptides belonging to each organism in
the database
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a table with two columns, namely, Organisms and Counts

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptides\_per\_protein() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_peptides\_per\_protein()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_peptides_per_protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptides\_per\_protein}}}{\emph{\DUrole{n}{ascending}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
return a pandas dataframe that contain the number of peptides belonging to each protein 
inferred in the experiment
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ascending}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} ascending sort the proteins by their number of parent number of child peptides, defaults to False

\item[{Returns}] \leavevmode
a table with the following columns, Proteins and Number\_of\_Peptides

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_poly\_parental\_peptides() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_poly\_parental\_peptides()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_poly_parental_peptides}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_poly\_parental\_peptides}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}
return a list of peptides that have more than one parent proteins
:return: {[}list of peptide instance 
:rtype: Peptides

\end{fulllineitems}

\index{get\_proband\_name() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_proband\_name()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_proband_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_proband\_name}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the proband name

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_proteins() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_proteins()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_proteins}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein}]{\sphinxcrossref{IPTK.Classes.Protein.Protein}}}\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a set of all the proteins in the experimental object

\item[{Return type}] \leavevmode
Proteins

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_tissue() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_tissue()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_tissue}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_tissue}}}{}{{ $\rightarrow$ {\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Tissue.Tissue}]{\sphinxcrossref{IPTK.Classes.Tissue.Tissue}}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the tissue of the current experiment.

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Tissue.Tissue}]{\sphinxcrossref{Tissue}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_tissue\_name() (IPTK.Classes.Experiment.Experiment method)@\spxentry{get\_tissue\_name()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.get_tissue_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_tissue\_name}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the tissue name

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_allele\_group() (IPTK.Classes.Experiment.Experiment method)@\spxentry{has\_allele\_group()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.has_allele_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_allele\_group}}}{\emph{\DUrole{n}{gene\_group}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}
return whether or not the experiment contain peptides eluted from an HLA\sphinxhyphen{}alleles belonging to the provided allele group or not
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{gene\_group}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the gene group to query the hla\_set against

\item[{Returns}] \leavevmode
True if the gene group has a member that is a member of the instance HLASet and False otherwise

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_gene() (IPTK.Classes.Experiment.Experiment method)@\spxentry{has\_gene()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.has_gene}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_gene}}}{\emph{\DUrole{n}{locus}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}
return whether or not the experiment contain peptides eluted from an HLA\sphinxhyphen{}alleles belonging to the provided locus or not
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{locus}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the locus of the allele to query the hla\_set against

\item[{Returns}] \leavevmode
True if the locus has a member that is a member of the instance HLASet and False otherwise

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_hla\_allele() (IPTK.Classes.Experiment.Experiment method)@\spxentry{has\_hla\_allele()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.has_hla_allele}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_hla\_allele}}}{\emph{\DUrole{n}{individual}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}
return whether or not the experiment contain an eluted peptides from the provided alleles
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{individual}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} is the name of the allele as a string

\item[{Returns}] \leavevmode
True if the allele is a member of the instance HLASet and False otherwise.

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_protein\_group() (IPTK.Classes.Experiment.Experiment method)@\spxentry{has\_protein\_group()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.has_protein_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_protein\_group}}}{\emph{\DUrole{n}{protein\_group}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}
return whether or not the experiment contain peptides eluted from an HLA\sphinxhyphen{}alleles belonging to the provided protein group or not
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{protein\_group}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The protein group to query the hla\_set against

\item[{Returns}] \leavevmode
True if the locus has a member that is a member of the instance HLASet and False otherwise

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{is\_a\_parent\_protein() (IPTK.Classes.Experiment.Experiment method)@\spxentry{is\_a\_parent\_protein()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.is_a_parent_protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{is\_a\_parent\_protein}}}{\emph{\DUrole{n}{protein}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{protein}} \textendash{} check if the protein is a member of the instance proteins or not.

\item[{Returns}] \leavevmode
True if the protein has been identified in the current instance, False otherwise.

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{is\_member() (IPTK.Classes.Experiment.Experiment method)@\spxentry{is\_member()}\spxextra{IPTK.Classes.Experiment.Experiment method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment.is_member}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{is\_member}}}{\emph{\DUrole{n}{peptide}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{peptide}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} check if the peptide is a member of the instance peptides or not.

\item[{Returns}] \leavevmode
True if the peptide has been identified in the current instance, False otherwise.

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.ExperimentalSet module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.ExperimentalSet}}\label{\detokenize{IPTK.Classes:iptk-classes-experimentalset-module}}\index{module@\spxentry{module}!IPTK.Classes.ExperimentalSet@\spxentry{IPTK.Classes.ExperimentalSet}}\index{IPTK.Classes.ExperimentalSet@\spxentry{IPTK.Classes.ExperimentalSet}!module@\spxentry{module}}
An Experimental set which is a collection of experiments.
The class provides an API for integrating and comparing different experimental instances.
\index{ExperimentSet (class in IPTK.Classes.ExperimentalSet)@\spxentry{ExperimentSet}\spxextra{class in IPTK.Classes.ExperimentalSet}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.ExperimentalSet.}}\sphinxbfcode{\sphinxupquote{ExperimentSet}}}{\emph{\DUrole{o}{**}\DUrole{n}{exp\_id\_pair}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

an API for integrating and comparing different experimental instances
\index{add\_experiment() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{add\_experiment()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.add_experiment}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_experiment}}}{\emph{\DUrole{o}{**}\DUrole{n}{exp\_id\_pair}}}{{ $\rightarrow$ None}}
add an arbitrary number of experiments to the set

\end{fulllineitems}

\index{compare\_org\_count\_among\_exps() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compare\_org\_count\_among\_exps()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compare_org_count_among_exps}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compare\_org\_count\_among\_exps}}}{\emph{\DUrole{n}{org}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{abs\_count}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{org}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the organism to query the database for it.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{abs\_count}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The absolute count, defaults to False

\end{itemize}

\item[{Returns}] \leavevmode
the count of the peptides that belong to a specific organism in the database.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{compare\_peptide\_counts() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compare\_peptide\_counts()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compare_peptide_counts}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compare\_peptide\_counts}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A table that contain the total number of peptides and per\sphinxhyphen{}organism peptide counts         among all experiments in the set

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_average\_distance\_between\_exps() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_average\_distance\_between\_exps()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_average_distance_between_exps}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_average\_distance\_between\_exps}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
compute the average distance between experiments by taking the average over the z\sphinxhyphen{}axis
of the 3D tensor computed by the function compute\_change\_in\_protein\_representation.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A 2D tensor with shape of (num\sphinxhyphen{}experiments, num\sphinxhyphen{}experiments)

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_change\_in\_protein\_representation() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_change\_in\_protein\_representation()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_change_in_protein_representation}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_change\_in\_protein\_representation}}}{}{{ $\rightarrow$ numpy.ndarray}}
Compute the change in protein representation among the proteins that are presented/ detect in all of the         instance’s experiments.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a 3D tensor, T, with shape of (num\sphinxhyphen{}experiments, num\sphinxhyphen{}experiments, num\sphinxhyphen{}proteins),         where T{[}i,j,k{]} is a the difference between experiment i \& j with respect to the k th protein         :rtype: np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_correlation\_in\_experssion() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_correlation\_in\_experssion()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_correlation_in_experssion}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_correlation\_in\_experssion}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
compute the correlation in parent protein gene\sphinxhyphen{}expression across all the experiments
in the set. See the function \sphinxstylestrong{compute\_binary\_correlation} in the analysis module 
for information about the computational logic.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
return a 2D matrix containing the coorelation in gene expression between each pair of experiments inside the current instance collection of experiments.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_peptide\_length\_table() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_peptide\_length\_table()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_length_table}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_peptide\_length\_table}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A table that contain the length of each peptide in the experiment

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_peptide\_overlap\_matrix() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_peptide\_overlap\_matrix()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_overlap_matrix}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_peptide\_overlap\_matrix}}}{}{{ $\rightarrow$ numpy.ndarray}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a 2D matrix containing the number of peptide overlapping between each pair of experiments inside the current instance collection of experiment.

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_peptide\_representation\_count() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_peptide\_representation\_count()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_representation_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_peptide\_representation\_count}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }int\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The number of times a peptide was observed accross  experiments in the set

\item[{Return type}] \leavevmode
Counts

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_protein\_coverage\_over\_the\_set() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_protein\_coverage\_over\_the\_set()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_coverage_over_the_set}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_protein\_coverage\_over\_the\_set}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the mapped representation for each protein in accross the entire  set

\item[{Return type}] \leavevmode
Dict{[}str, np.ndarray{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_protein\_overlap\_matrix() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_protein\_overlap\_matrix()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_overlap_matrix}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_protein\_overlap\_matrix}}}{}{{ $\rightarrow$ numpy.ndarray}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
return a 2D matrix containing the number of proteins overlapping between each pair of experiments inside the current instance collection of experiment.

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{compute\_protein\_representation\_count() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{compute\_protein\_representation\_count()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_representation_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{compute\_protein\_representation\_count}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }int\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The number of times a protein was observed accross the experiment in the set

\item[{Return type}] \leavevmode
Counts

\end{description}\end{quote}

\end{fulllineitems}

\index{drop\_peptides\_belong\_to\_org() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{drop\_peptides\_belong\_to\_org()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.drop_peptides_belong_to_org}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{drop\_peptides\_belong\_to\_org}}}{\emph{\DUrole{n}{org\_name}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ None}}
drop all the peptides that belong to the provided organisms from all experiments in the set.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{org\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the organism to drop

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_allele\_count() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_allele\_count()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_allele_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_allele\_count}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }int\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the number of experiments obtained from each allele in the instance.

\item[{Return type}] \leavevmode
Counts

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_experiment() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_experiment()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiment}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_experiment}}}{\emph{\DUrole{n}{exp\_name}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ {\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}
return the experiment pointed to by the provided experimental name
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{exp\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the experiment

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if the provided experimental name is not in the dataset.

\item[{Returns}] \leavevmode
the experiment corresponds to the provided name

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{Experiment}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_experimental\_names() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_experimental\_names()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_experimental_names}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_experimental\_names}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list with all the identifiers of the experiments in the set

\item[{Return type}] \leavevmode
Names

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_experiments() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_experiments()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiments}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_experiments}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}Experiment\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
return a dict with all the experiments stored in the instance as value of ids as keys.

\item[{Return type}] \leavevmode
Dict{[}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{Experiment}}}{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_num\_experiments\_in\_the\_set() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_num\_experiments\_in\_the\_set()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_num_experiments_in_the_set}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_num\_experiments\_in\_the\_set}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The number of experiments currently in the set

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptides\_present\_in\_all() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_peptides\_present\_in\_all()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_peptides_present_in_all}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptides\_present\_in\_all}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the peptides that are observed in every experiments in the set.

\item[{Return type}] \leavevmode
Peptides

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_proband\_count() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_proband\_count()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_proband_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_proband\_count}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }int\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The number of experiments obtained from each proband in the ExperimentalSet.

\item[{Return type}] \leavevmode
Counts

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_proteins\_present\_in\_all() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_proteins\_present\_in\_all()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_proteins_present_in_all}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_proteins\_present\_in\_all}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the proteins that are inferred in all experiments of the set

\item[{Return type}] \leavevmode
Proteins

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_tissue\_counts() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_tissue\_counts()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_tissue_counts}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_tissue\_counts}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }int\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The number of experiments obtained from each tissue in the current instance

\item[{Return type}] \leavevmode
Counts

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_total\_peptide\_per\_org\_count() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_total\_peptide\_per\_org\_count()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_total_peptide_per_org_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_total\_peptide\_per\_org\_count}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The total count of peptides per organism accross the all experiments in the set.

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_unique\_orgs() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_unique\_orgs()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_orgs}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_unique\_orgs}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list of the unique organisms in the set

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_unique\_peptides() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_unique\_peptides()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_peptides}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_unique\_peptides}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The set of unique peptides in the experimentalSet

\item[{Return type}] \leavevmode
Peptides

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_unique\_proteins() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{get\_unique\_proteins()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_unique\_proteins}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the set of unique proteins in the experimentalset

\item[{Return type}] \leavevmode
Proteins

\end{description}\end{quote}

\end{fulllineitems}

\index{group\_by\_proband() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{group\_by\_proband()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_proband}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{group\_by\_proband}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet}]{\sphinxcrossref{IPTK.Classes.ExperimentalSet.ExperimentSet}}}\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a map between each proband and an Experimentalset object represent all the experiments objects         belonging to this proband.

\item[{Return type}] \leavevmode
Dict{[}str,{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet}]{\sphinxcrossref{ExperimentSet}}}{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{group\_by\_tissue() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{group\_by\_tissue()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_tissue}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{group\_by\_tissue}}}{}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet}]{\sphinxcrossref{IPTK.Classes.ExperimentalSet.ExperimentSet}}}\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A map between each tissue and an ExperimentalSet object representing all experiments belonging to that tissue.

\item[{Return type}] \leavevmode
Dict{[}str,{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet}]{\sphinxcrossref{ExperimentSet}}}{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{is\_peptide\_present\_in\_all() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{is\_peptide\_present\_in\_all()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.is_peptide_present_in_all}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{is\_peptide\_present\_in\_all}}}{\emph{\DUrole{n}{peptide}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{peptide}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The peptide sequence to search its occurrences in every experiment contained in the set

\item[{Returns}] \leavevmode
True if peptide is present in all experiments inside the instance, False otherwise

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{is\_protein\_present\_in\_all() (IPTK.Classes.ExperimentalSet.ExperimentSet method)@\spxentry{is\_protein\_present\_in\_all()}\spxextra{IPTK.Classes.ExperimentalSet.ExperimentSet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.ExperimentalSet.ExperimentSet.is_protein_present_in_all}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{is\_protein\_present\_in\_all}}}{\emph{\DUrole{n}{protein}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{protein}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the protein id to search its occurrences in every experimental in the set

\item[{Returns}] \leavevmode
True if peptide is present in all experiments inside the instance, False otherwise

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.HLAChain module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.HLAChain}}\label{\detokenize{IPTK.Classes:iptk-classes-hlachain-module}}\index{module@\spxentry{module}!IPTK.Classes.HLAChain@\spxentry{IPTK.Classes.HLAChain}}\index{IPTK.Classes.HLAChain@\spxentry{IPTK.Classes.HLAChain}!module@\spxentry{module}}
The implementation of an HLA molecules
\index{HLAChain (class in IPTK.Classes.HLAChain)@\spxentry{HLAChain}\spxextra{class in IPTK.Classes.HLAChain}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAChain.HLAChain}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.HLAChain.}}\sphinxbfcode{\sphinxupquote{HLAChain}}}{\emph{\DUrole{n}{name}\DUrole{p}{:} \DUrole{n}{str}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}
\index{get\_allele\_group() (IPTK.Classes.HLAChain.HLAChain method)@\spxentry{get\_allele\_group()}\spxextra{IPTK.Classes.HLAChain.HLAChain method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAChain.HLAChain.get_allele_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_allele\_group}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The allele group

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_chain\_class() (IPTK.Classes.HLAChain.HLAChain method)@\spxentry{get\_chain\_class()}\spxextra{IPTK.Classes.HLAChain.HLAChain method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAChain.HLAChain.get_chain_class}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_chain\_class}}}{\emph{\DUrole{n}{gene\_name}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the gene

\item[{Returns}] \leavevmode
1 if the gene belongs to class one and 2 if it belong to class two

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_class() (IPTK.Classes.HLAChain.HLAChain method)@\spxentry{get\_class()}\spxextra{IPTK.Classes.HLAChain.HLAChain method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAChain.HLAChain.get_class}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_class}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The HLA class

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_gene() (IPTK.Classes.HLAChain.HLAChain method)@\spxentry{get\_gene()}\spxextra{IPTK.Classes.HLAChain.HLAChain method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAChain.HLAChain.get_gene}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_gene}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the gene name

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_name() (IPTK.Classes.HLAChain.HLAChain method)@\spxentry{get\_name()}\spxextra{IPTK.Classes.HLAChain.HLAChain method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAChain.HLAChain.get_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_name}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The chain name

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_protein\_group() (IPTK.Classes.HLAChain.HLAChain method)@\spxentry{get\_protein\_group()}\spxextra{IPTK.Classes.HLAChain.HLAChain method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAChain.HLAChain.get_protein_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_protein\_group}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the protein name

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.HLAMolecules module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.HLAMolecules}}\label{\detokenize{IPTK.Classes:iptk-classes-hlamolecules-module}}\index{module@\spxentry{module}!IPTK.Classes.HLAMolecules@\spxentry{IPTK.Classes.HLAMolecules}}\index{IPTK.Classes.HLAMolecules@\spxentry{IPTK.Classes.HLAMolecules}!module@\spxentry{module}}
a representation of an HLA molecules
\index{HLAMolecule (class in IPTK.Classes.HLAMolecules)@\spxentry{HLAMolecule}\spxextra{class in IPTK.Classes.HLAMolecules}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAMolecules.HLAMolecule}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.HLAMolecules.}}\sphinxbfcode{\sphinxupquote{HLAMolecule}}}{\emph{\DUrole{o}{**}\DUrole{n}{hla\_chains}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}
\index{get\_allele\_group() (IPTK.Classes.HLAMolecules.HLAMolecule method)@\spxentry{get\_allele\_group()}\spxextra{IPTK.Classes.HLAMolecules.HLAMolecule method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAMolecules.HLAMolecule.get_allele_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_allele\_group}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the allele group for the instance chain/pair of chains

\item[{Return type}] \leavevmode
AlleleGroup

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_class() (IPTK.Classes.HLAMolecules.HLAMolecule method)@\spxentry{get\_class()}\spxextra{IPTK.Classes.HLAMolecules.HLAMolecule method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAMolecules.HLAMolecule.get_class}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_class}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The class of the HLA molecules

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_gene() (IPTK.Classes.HLAMolecules.HLAMolecule method)@\spxentry{get\_gene()}\spxextra{IPTK.Classes.HLAMolecules.HLAMolecule method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAMolecules.HLAMolecule.get_gene}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_gene}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
gene/pair of genes coding for the current HLA molecules

\item[{Return type}] \leavevmode
Genes

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_name() (IPTK.Classes.HLAMolecules.HLAMolecule method)@\spxentry{get\_name()}\spxextra{IPTK.Classes.HLAMolecules.HLAMolecule method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAMolecules.HLAMolecule.get_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_name}}}{\emph{\DUrole{n}{sep}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}:\textquotesingle{}}}}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{sep}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the name of the allele by concatenating the names of the individual chains using         a separator, defaults to ‘:’

\item[{Returns}] \leavevmode
{[}description{]}

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_protein\_group() (IPTK.Classes.HLAMolecules.HLAMolecule method)@\spxentry{get\_protein\_group()}\spxextra{IPTK.Classes.HLAMolecules.HLAMolecule method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLAMolecules.HLAMolecule.get_protein_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_protein\_group}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The protein group for the instance chain/pair of chains

\item[{Return type}] \leavevmode
ProteinGroup

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.HLASet module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.HLASet}}\label{\detokenize{IPTK.Classes:iptk-classes-hlaset-module}}\index{module@\spxentry{module}!IPTK.Classes.HLASet@\spxentry{IPTK.Classes.HLASet}}\index{IPTK.Classes.HLASet@\spxentry{IPTK.Classes.HLASet}!module@\spxentry{module}}
@author: Hesham ElAbd
@contact: \sphinxhref{mailto:h.elabd@ikmb.uni-kiel.de}{h.elabd@ikmb.uni\sphinxhyphen{}kiel.de}
\index{HLASet (class in IPTK.Classes.HLASet)@\spxentry{HLASet}\spxextra{class in IPTK.Classes.HLASet}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.HLASet.}}\sphinxbfcode{\sphinxupquote{HLASet}}}{\emph{\DUrole{n}{hlas}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}, \emph{\DUrole{n}{gene\_sep}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}:\textquotesingle{}}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}
\index{get\_alleles() (IPTK.Classes.HLASet.HLASet method)@\spxentry{get\_alleles()}\spxextra{IPTK.Classes.HLASet.HLASet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet.get_alleles}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_alleles}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The class of the HLA\sphinxhyphen{}alleles in the current instance

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_class() (IPTK.Classes.HLASet.HLASet method)@\spxentry{get\_class()}\spxextra{IPTK.Classes.HLASet.HLASet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet.get_class}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_class}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
The class of the HLA\sphinxhyphen{}alleles in the current instance

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_hla\_count() (IPTK.Classes.HLASet.HLASet method)@\spxentry{get\_hla\_count()}\spxextra{IPTK.Classes.HLASet.HLASet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet.get_hla_count}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_hla\_count}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the count of HLA molecules in the set

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_allele() (IPTK.Classes.HLASet.HLASet method)@\spxentry{has\_allele()}\spxextra{IPTK.Classes.HLASet.HLASet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet.has_allele}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_allele}}}{\emph{\DUrole{n}{allele}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{allele}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the allele to check the instance for

\item[{Returns}] \leavevmode
True, if the provided allele is in the current instance, False otherwise.

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_allele\_group() (IPTK.Classes.HLASet.HLASet method)@\spxentry{has\_allele\_group()}\spxextra{IPTK.Classes.HLASet.HLASet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet.has_allele_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_allele\_group}}}{\emph{\DUrole{n}{allele\_group}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{allele\_group}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the allele group to search the set for

\item[{Returns}] \leavevmode
True, if at least one allele in the set belongs to the provided allele group, False otherwise.

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_gene() (IPTK.Classes.HLASet.HLASet method)@\spxentry{has\_gene()}\spxextra{IPTK.Classes.HLASet.HLASet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet.has_gene}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_gene}}}{\emph{\DUrole{n}{gene\_name}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the gene name to search the set against.

\item[{Returns}] \leavevmode
True, if at least one of the alleles in the set belongs to the provided gene. False otherwise

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{has\_protein\_group() (IPTK.Classes.HLASet.HLASet method)@\spxentry{has\_protein\_group()}\spxextra{IPTK.Classes.HLASet.HLASet method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.HLASet.HLASet.has_protein_group}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{has\_protein\_group}}}{\emph{\DUrole{n}{protein\_group}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{protein\_group}} \textendash{} The protein group to search the set for

\item[{Returns}] \leavevmode
True, if at least one allele in the set belongs to the provided protein group

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.Peptide module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.Peptide}}\label{\detokenize{IPTK.Classes:iptk-classes-peptide-module}}\index{module@\spxentry{module}!IPTK.Classes.Peptide@\spxentry{IPTK.Classes.Peptide}}\index{IPTK.Classes.Peptide@\spxentry{IPTK.Classes.Peptide}!module@\spxentry{module}}
A representation of the eluted peptides and its identified proteins.
\index{Peptide (class in IPTK.Classes.Peptide)@\spxentry{Peptide}\spxextra{class in IPTK.Classes.Peptide}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Peptide.}}\sphinxbfcode{\sphinxupquote{Peptide}}}{\emph{\DUrole{n}{pep\_seq}\DUrole{p}{:} \DUrole{n}{str}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

An representation of an eluted peptide.
\index{add\_org\_2\_parent() (IPTK.Classes.Peptide.Peptide method)@\spxentry{add\_org\_2\_parent()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.add_org_2_parent}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_org\_2\_parent}}}{\emph{\DUrole{n}{prot\_name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{org}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ None}}
add the source organism of one of the instance parent protein
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{prot\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the protein, i.e. the identifier of the protein

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{org}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the organism

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} incase the provided protein is not a parent of the provided peptide

\end{description}\end{quote}

\end{fulllineitems}

\index{add\_parent\_protein() (IPTK.Classes.Peptide.Peptide method)@\spxentry{add\_parent\_protein()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.add_parent_protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{add\_parent\_protein}}}{\emph{\DUrole{n}{parent\_protein}}, \emph{\DUrole{n}{start\_index}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{end\_index}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ None}}
add a protein instance as a parent to the current peptide. 
The library use Python\sphinxhyphen{}based indexing where its 0\sphinxhyphen{}indexed and ranges are treated as {[}start, end). 
:param parent\_protein: a Protein instance that act as a parent to the peptide.
:type parent\_protein: Protein 
:param start\_index: the position in the parent protein where the peptide starts
:type start\_index: int 
:param end\_index: the index of the amino acid that occurs after the last amino acid in the peptide, 
:type start\_index: int

\end{fulllineitems}

\index{get\_c\_terminal\_flank\_seq() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_c\_terminal\_flank\_seq()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_c_terminal_flank_seq}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_c\_terminal\_flank\_seq}}}{\emph{\DUrole{n}{flank\_len}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
:param flank\_len:the length of the flanking regions 
:type flank\_len: int
:return: a list of string containing the sequences located downstream of the peptide in the parent protein. 
:rtype: {[}type{]}

\end{fulllineitems}

\index{get\_flanked\_peptide() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_flanked\_peptide()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_flanked_peptide}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_flanked\_peptide}}}{\emph{\DUrole{n}{flank\_len}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{flank\_len}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the length of the flanking regions

\item[{Returns}] \leavevmode
A list of string containing the length of the peptide + the flanking region from               both the N and C terminal of the instance peptide, from all proteins.

\item[{Return type}] \leavevmode
Sequences

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_length() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_length()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_length}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_length}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the length of the peptides

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_n\_terminal\_flank\_seq() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_n\_terminal\_flank\_seq()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_n_terminal_flank_seq}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_n\_terminal\_flank\_seq}}}{\emph{\DUrole{n}{flank\_len}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{flank\_len}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the length of the flanking regions

\item[{Returns}] \leavevmode
a list of string containing the sequences located upstream of the peptide in the parent protein.

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_non\_presented\_peptides() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_non\_presented\_peptides()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_non_presented_peptides}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_non\_presented\_peptides}}}{\emph{\DUrole{n}{length}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The length, i.e. number of amino acids, for the non\sphinxhyphen{}presented peptide

\item[{Returns}] \leavevmode
non\sphinxhyphen{}presented peptide from all the parent protein of the current peptide instance.

\item[{Return type}] \leavevmode
Sequences

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_number\_of\_parents() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_number\_of\_parents()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_number_of_parents}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_number\_of\_parents}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the number of instance parent proteins

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_number\_parent\_protein() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_number\_parent\_protein()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_number_parent_protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_number\_parent\_protein}}}{}{{ $\rightarrow$ int}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the number of parent proteins this instance has

\item[{Return type}] \leavevmode
int

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_parent() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_parent()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_parent}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_parent}}}{\emph{\DUrole{n}{pro\_id}\DUrole{p}{:} \DUrole{n}{str}}}{}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pro\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The protein identifer

\item[{Returns}] \leavevmode
the parent protein that has an id matching the user defined pro\_id

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein}]{\sphinxcrossref{Protein}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_parent\_proteins() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_parent\_proteins()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_parent_proteins}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_parent\_proteins}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
\end{fulllineitems}

\index{get\_parents\_org() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_parents\_org()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_parents_org}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_parents\_org}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list containing the name of each parent protein source organisms

\item[{Return type}] \leavevmode
Organisms

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptide\_seq() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_peptide\_seq()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_peptide_seq}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptide\_seq}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the sequence of the peptide.

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_pos\_in\_parent() (IPTK.Classes.Peptide.Peptide method)@\spxentry{get\_pos\_in\_parent()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.get_pos_in_parent}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_pos\_in\_parent}}}{\emph{\DUrole{n}{pro\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ Tuple\DUrole{p}{{[}}int\DUrole{p}{, }int\DUrole{p}{{]}}}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pro\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the id of the parent protein

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} If the identifier is not a parent of the instance

\item[{Returns}] \leavevmode
the start and end position of the instance peptide in the parent pointed out by the provided identifier

\item[{Return type}] \leavevmode
Range

\end{description}\end{quote}

\end{fulllineitems}

\index{is\_child\_of() (IPTK.Classes.Peptide.Peptide method)@\spxentry{is\_child\_of()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.is_child_of}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{is\_child\_of}}}{\emph{\DUrole{n}{pro\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{pro\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} is the protein id

\item[{Returns}] \leavevmode
True if a user provided protein\sphinxhyphen{}id is a parent for the instance peptide, False otherwise

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}

\index{map\_to\_parent\_protein() (IPTK.Classes.Peptide.Peptide method)@\spxentry{map\_to\_parent\_protein()}\spxextra{IPTK.Classes.Peptide.Peptide method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide.map_to_parent_protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{map\_to\_parent\_protein}}}{}{{ $\rightarrow$ List\DUrole{p}{{[}}numpy.ndarray\DUrole{p}{{]}}}}
Mapped the instance peptide to the parent protein and returned a 
list of numpy arrays where each array has a size of 1 by protein length. 
within the protein the range representing the peptide is encoded as one while
the rest is zero.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
A list of binary encoded arrays represent this mapping.

\item[{Return type}] \leavevmode
MappedProtein

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.Proband module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.Proband}}\label{\detokenize{IPTK.Classes:iptk-classes-proband-module}}\index{module@\spxentry{module}!IPTK.Classes.Proband@\spxentry{IPTK.Classes.Proband}}\index{IPTK.Classes.Proband@\spxentry{IPTK.Classes.Proband}!module@\spxentry{module}}
A description for an IP proband
\index{Proband (class in IPTK.Classes.Proband)@\spxentry{Proband}\spxextra{class in IPTK.Classes.Proband}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Proband.Proband}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Proband.}}\sphinxbfcode{\sphinxupquote{Proband}}}{\emph{\DUrole{o}{**}\DUrole{n}{info}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}
\index{get\_meta\_data() (IPTK.Classes.Proband.Proband method)@\spxentry{get\_meta\_data()}\spxextra{IPTK.Classes.Proband.Proband method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Proband.Proband.get_meta_data}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_meta\_data}}}{}{{ $\rightarrow$ dict}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a dict that contain all the meta\sphinxhyphen{}data about the patient

\item[{Return type}] \leavevmode
dict

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_name() (IPTK.Classes.Proband.Proband method)@\spxentry{get\_name()}\spxextra{IPTK.Classes.Proband.Proband method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Proband.Proband.get_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_name}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the name of the patient

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{update\_info() (IPTK.Classes.Proband.Proband method)@\spxentry{update\_info()}\spxextra{IPTK.Classes.Proband.Proband method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Proband.Proband.update_info}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{update\_info}}}{\emph{\DUrole{o}{**}\DUrole{n}{info}}}{{ $\rightarrow$ None}}
add new or update existing info about the patient  using an arbitrary number of key\sphinxhyphen{}value pair to be added to added to the instance meta\sphinxhyphen{}info dict

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.Protein module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.Protein}}\label{\detokenize{IPTK.Classes:iptk-classes-protein-module}}\index{module@\spxentry{module}!IPTK.Classes.Protein@\spxentry{IPTK.Classes.Protein}}\index{IPTK.Classes.Protein@\spxentry{IPTK.Classes.Protein}!module@\spxentry{module}}
A representation of a protein that has been inferred from an IP experiment.
\index{Protein (class in IPTK.Classes.Protein)@\spxentry{Protein}\spxextra{class in IPTK.Classes.Protein}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Protein.}}\sphinxbfcode{\sphinxupquote{Protein}}}{\emph{\DUrole{n}{prot\_id}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{seq}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{org}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

representation of a protein that has been infered from an IP experiment.
\index{get\_id() (IPTK.Classes.Protein.Protein method)@\spxentry{get\_id()}\spxextra{IPTK.Classes.Protein.Protein method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein.get_id}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_id}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
return the protein identifier.

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_non\_presented\_peptide() (IPTK.Classes.Protein.Protein method)@\spxentry{get\_non\_presented\_peptide()}\spxextra{IPTK.Classes.Protein.Protein method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein.get_non_presented_peptide}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_non\_presented\_peptide}}}{\emph{\DUrole{n}{exc\_reg\_s\_idx}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{exc\_reg\_e\_idx}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{length}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ str}}
sample a peptide from the protein sequences where the sampled peptides is not part of the               experimentally identified regions.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exc\_reg\_s\_idx}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the start point in the reference protein sequence of the experimentally identified peptide.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exc\_reg\_e\_idx}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the end point in the reference protein sequence of the experimentally identified peptide.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} length the non\sphinxhyphen{}presented peptides.

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the length of the peptide is bigger than the protein length

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the length of the peptide is smaller than or equal to zero

\end{itemize}

\item[{Returns}] \leavevmode
a substring of the instance sequence

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_org() (IPTK.Classes.Protein.Protein method)@\spxentry{get\_org()}\spxextra{IPTK.Classes.Protein.Protein method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein.get_org}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_org}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the organism in which this instance protein belong.

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_peptides\_map() (IPTK.Classes.Protein.Protein method)@\spxentry{get\_peptides\_map()}\spxextra{IPTK.Classes.Protein.Protein method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein.get_peptides_map}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_peptides\_map}}}{\emph{\DUrole{n}{start\_idxs}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}int\DUrole{p}{{]}}}}, \emph{\DUrole{n}{end\_idxs}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}int\DUrole{p}{{]}}}}}{{ $\rightarrow$ numpy.ndarray}}
compute a coverage over the protein sequence
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{start\_idxs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Index}}) \textendash{} a list of integers representing the start positions

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{end\_idxs}} \textendash{} a list of integers representing the end positions

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if start\_indxs and end\_idxs MUST be of equal length are not of equal length

\item[{Returns}] \leavevmode
A numpy array with shape of 1 by the length of the protein where every element in the array donates the number of times, It has been observed in the experiment.

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_seq() (IPTK.Classes.Protein.Protein method)@\spxentry{get\_seq()}\spxextra{IPTK.Classes.Protein.Protein method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein.get_seq}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_seq}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the protein sequence.

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{set\_org() (IPTK.Classes.Protein.Protein method)@\spxentry{set\_org()}\spxextra{IPTK.Classes.Protein.Protein method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Protein.Protein.set_org}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{set\_org}}}{\emph{\DUrole{n}{org}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ None}}
a post\sphinxhyphen{}instantitation mechanism to set the organism for which the protein belong.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{org}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the organism

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{IPTK.Classes.Tissue module}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes.Tissue}}\label{\detokenize{IPTK.Classes:iptk-classes-tissue-module}}\index{module@\spxentry{module}!IPTK.Classes.Tissue@\spxentry{IPTK.Classes.Tissue}}\index{IPTK.Classes.Tissue@\spxentry{IPTK.Classes.Tissue}!module@\spxentry{module}}
A representation of the Tissue used in an IP Experiment.
\index{ExpressionProfile (class in IPTK.Classes.Tissue)@\spxentry{ExpressionProfile}\spxextra{class in IPTK.Classes.Tissue}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.ExpressionProfile}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Tissue.}}\sphinxbfcode{\sphinxupquote{ExpressionProfile}}}{\emph{\DUrole{n}{name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{expression\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{aux\_proteins}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}pandas.core.frame.DataFrame\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}

a representation of tissue reference expression value.
\index{get\_gene\_id\_expression() (IPTK.Classes.Tissue.ExpressionProfile method)@\spxentry{get\_gene\_id\_expression()}\spxextra{IPTK.Classes.Tissue.ExpressionProfile method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.ExpressionProfile.get_gene_id_expression}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_gene\_id\_expression}}}{\emph{\DUrole{n}{gene\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ float}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{gene\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the gene id to retrive its expression value from the database

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if the provided id is not defined in the instance table

\item[{Returns}] \leavevmode
the expression value of the provided gene id.

\item[{Return type}] \leavevmode
float

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_gene\_name\_expression() (IPTK.Classes.Tissue.ExpressionProfile method)@\spxentry{get\_gene\_name\_expression()}\spxextra{IPTK.Classes.Tissue.ExpressionProfile method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.ExpressionProfile.get_gene_name_expression}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_gene\_name\_expression}}}{\emph{\DUrole{n}{gene\_name}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ float}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{gene\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the gene name to retrive its expression value from the database

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if the provided id is not defined in the instance table

\item[{Returns}] \leavevmode
the expression value of the provided gene name.

\item[{Return type}] \leavevmode
float

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_name() (IPTK.Classes.Tissue.ExpressionProfile method)@\spxentry{get\_name()}\spxextra{IPTK.Classes.Tissue.ExpressionProfile method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.ExpressionProfile.get_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_name}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the name of the tissue which the instance profile its gene expression

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_table() (IPTK.Classes.Tissue.ExpressionProfile method)@\spxentry{get\_table()}\spxextra{IPTK.Classes.Tissue.ExpressionProfile method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.ExpressionProfile.get_table}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_table}}}{}{{ $\rightarrow$ pandas.core.frame.DataFrame}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
return a table that contain the expression of all the transcript in the current profile                including core and auxiliary proteins

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}

\index{Tissue (class in IPTK.Classes.Tissue)@\spxentry{Tissue}\spxextra{class in IPTK.Classes.Tissue}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.Tissue}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{class }}\sphinxcode{\sphinxupquote{IPTK.Classes.Tissue.}}\sphinxbfcode{\sphinxupquote{Tissue}}}{\emph{\DUrole{n}{name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{main\_exp\_value}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB}]{\sphinxcrossref{IPTK.Classes.Database.GeneExpressionDB}}}}}, \emph{\DUrole{n}{main\_location}\DUrole{p}{:} \DUrole{n}{{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB}]{\sphinxcrossref{IPTK.Classes.Database.CellularLocationDB}}}}}, \emph{\DUrole{n}{aux\_exp\_value}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.GeneExpressionDB}]{\sphinxcrossref{IPTK.Classes.Database.GeneExpressionDB}}}\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}, \emph{\DUrole{n}{aux\_location}\DUrole{p}{:} \DUrole{n}{Optional\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB}]{\sphinxcrossref{IPTK.Classes.Database.CellularLocationDB}}}\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{None}}}{}
Bases: \sphinxcode{\sphinxupquote{object}}
\index{get\_expression\_profile() (IPTK.Classes.Tissue.Tissue method)@\spxentry{get\_expression\_profile()}\spxextra{IPTK.Classes.Tissue.Tissue method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.Tissue.get_expression_profile}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_expression\_profile}}}{}{{ $\rightarrow$ {\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Tissue.ExpressionProfile}]{\sphinxcrossref{IPTK.Classes.Tissue.ExpressionProfile}}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the expresion profile of the current tissue

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Tissue.ExpressionProfile}]{\sphinxcrossref{ExpressionProfile}}}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_name() (IPTK.Classes.Tissue.Tissue method)@\spxentry{get\_name()}\spxextra{IPTK.Classes.Tissue.Tissue method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.Tissue.get_name}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_name}}}{}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the name of the tissue

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_subCellular\_locations() (IPTK.Classes.Tissue.Tissue method)@\spxentry{get\_subCellular\_locations()}\spxextra{IPTK.Classes.Tissue.Tissue method}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Classes:IPTK.Classes.Tissue.Tissue.get_subCellular_locations}}\pysiglinewithargsret{\sphinxbfcode{\sphinxupquote{get\_subCellular\_locations}}}{}{{ $\rightarrow$ {\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB}]{\sphinxcrossref{IPTK.Classes.Database.CellularLocationDB}}}}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
the sub\sphinxhyphen{}cellular localization of all the proteins stored in current instance resources.

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Database.CellularLocationDB}]{\sphinxcrossref{CellularLocationDB}}}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{Module contents}
\label{\detokenize{IPTK.Classes:module-IPTK.Classes}}\label{\detokenize{IPTK.Classes:module-contents}}\index{module@\spxentry{module}!IPTK.Classes@\spxentry{IPTK.Classes}}\index{IPTK.Classes@\spxentry{IPTK.Classes}!module@\spxentry{module}}

\subparagraph{IPTK.IO package}
\label{\detokenize{IPTK.IO:iptk-io-package}}\label{\detokenize{IPTK.IO::doc}}

\subparagraph{Submodules}
\label{\detokenize{IPTK.IO:submodules}}

\subparagraph{IPTK.IO.InFunctions module}
\label{\detokenize{IPTK.IO:module-IPTK.IO.InFunctions}}\label{\detokenize{IPTK.IO:iptk-io-infunctions-module}}\index{module@\spxentry{module}!IPTK.IO.InFunctions@\spxentry{IPTK.IO.InFunctions}}\index{IPTK.IO.InFunctions@\spxentry{IPTK.IO.InFunctions}!module@\spxentry{module}}
Parse different user inputs into a standard format/tables used by the library.
\index{download\_pdb\_entry() (in module IPTK.IO.InFunctions)@\spxentry{download\_pdb\_entry()}\spxextra{in module IPTK.IO.InFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.InFunctions.download_pdb_entry}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.InFunctions.}}\sphinxbfcode{\sphinxupquote{download\_pdb\_entry}}}{\emph{\DUrole{n}{prot\_id}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ str}}
Download the structure of a protein from protein databank form as mmCIF file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{prot\_id}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the protein id

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} incase downloading and accessing the data failed

\item[{Returns}] \leavevmode
the path to the downloaded file

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{fasta2dict() (in module IPTK.IO.InFunctions)@\spxentry{fasta2dict()}\spxextra{in module IPTK.IO.InFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.InFunctions.fasta2dict}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.InFunctions.}}\sphinxbfcode{\sphinxupquote{fasta2dict}}}{\emph{\DUrole{n}{path2fasta}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{filter\_decoy}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{True}}, \emph{\DUrole{n}{decoy\_string}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}DECOY\textquotesingle{}}}}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}}}
loads a fasta file and construct a dict object where ids are keys and sequences are the value
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2fasta}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to load the fasta file

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{filter\_decoy}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} A boolean of whether or not to filter the decoy sequences from the database, defaults to True

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{decoy\_string}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The decoy database prefix, only valid incase filter\_decoy is set to true, defaults to ‘DECOY’

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} {[}description{]}

\item[{Returns}] \leavevmode
a dict where the protein ids are the keys and the protein sequences are the value

\item[{Return type}] \leavevmode
Dict{[}str,str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_identification\_table() (in module IPTK.IO.InFunctions)@\spxentry{load\_identification\_table()}\spxextra{in module IPTK.IO.InFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.InFunctions.load_identification_table}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.InFunctions.}}\sphinxbfcode{\sphinxupquote{load\_identification\_table}}}{\emph{\DUrole{n}{input\_path}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{sep}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
load \& process an identification table
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{input\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the path two the identification table. with the following columns: peptides which hold the identified
peptide sequence, protein which hold the identified protein sequence, start\_index, and end\_index where
the last two columns define the position of the peptide in the parent protein.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sep}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The separator to parse the provided table.

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} {[}description{]}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} {[}description{]}

\end{itemize}

\item[{Returns}] \leavevmode
{[}description{]}

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{parse\_mzTab\_to\_identification\_table() (in module IPTK.IO.InFunctions)@\spxentry{parse\_mzTab\_to\_identification\_table()}\spxextra{in module IPTK.IO.InFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.InFunctions.parse_mzTab_to_identification_table}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.InFunctions.}}\sphinxbfcode{\sphinxupquote{parse\_mzTab\_to\_identification\_table}}}{\emph{\DUrole{n}{path2mzTab}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{path2fastaDB}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{fasta\_reader\_param}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\textquotesingle{}decoy\_string\textquotesingle{}: \textquotesingle{}DECOY\textquotesingle{}, \textquotesingle{}filter\_decoy\textquotesingle{}: True\}}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
parse a user provided mzTab to an identification table
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2mzTab}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the path to the input mzTab file

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2fastaDB}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the path to a fasta sequence database to obtain the protein sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{fasta\_reader\_param}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict of parameters for controlling the behavior of the fasta reader , defaults to \{‘filter\_decoy’:True, ‘decoy\_string’:’DECOY’ \}

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} if the mztab file could not be open and loaded or if the fasta database could not be read

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if a protein id defined in the mzTab file could not be extracted from a matched sequence database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the peptide can not be mapped to the identified protein

\end{itemize}

\item[{Returns}] \leavevmode
the identification table

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{parse\_text\_table() (in module IPTK.IO.InFunctions)@\spxentry{parse\_text\_table()}\spxextra{in module IPTK.IO.InFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.InFunctions.parse_text_table}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.InFunctions.}}\sphinxbfcode{\sphinxupquote{parse\_text\_table}}}{\emph{\DUrole{n}{path2file}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{path2fastaDB}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{sep}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}, \emph{\DUrole{n}{fasta\_reader\_param}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\textquotesingle{}decoy\_string\textquotesingle{}: \textquotesingle{}DECOY\textquotesingle{}, \textquotesingle{}filter\_decoy\textquotesingle{}: True\}}}, \emph{\DUrole{n}{seq\_column}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Sequence\textquotesingle{}}}, \emph{\DUrole{n}{accession\_column}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Protein Accessions\textquotesingle{}}}, \emph{\DUrole{n}{protein\_group\_sep}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{};\textquotesingle{}}}, \emph{\DUrole{n}{remove\_protein\_version}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{True}}, \emph{\DUrole{n}{remove\_if\_not\_matched}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{True}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
Parse a user defined table to extract the columns containing the identification table
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2file}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to load the CSV file holding the results

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2fastaDB}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to a fasta sequence database to obtain the protein sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sep}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The table separators, defaults to ‘,’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{fasta\_reader\_param}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict of parameters for controlling the behavior of the fasta reader, defaults to \{‘filter\_decoy’:True, ‘decoy\_string’:’DECOY’ \}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{seq\_column}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The name of the columns containing the peptide sequence, defaults to ‘Sequence’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{accession\_column}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The name of the column containing the protein accession , defaults to ‘Protein Accessions’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_group\_sep}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The separator for the protein group,, defaults to ‘;’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{remove\_protein\_version}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} A bool if true strip the version number from the protein , defaults to True

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{remove\_if\_not\_matched}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} remove the peptide if it could not be matched to the parent protein, defaults to True

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} Incase either the sequences database or the identification table can not be open and loaded

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} In case the provided column names not in the provided identification table.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} Incase the protein sequence can not be extract from the sequence database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} incase the peptide could not be located in the protein sequence

\end{itemize}

\item[{Returns}] \leavevmode
an identification table

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{parse\_xml\_based\_format\_to\_identification\_table() (in module IPTK.IO.InFunctions)@\spxentry{parse\_xml\_based\_format\_to\_identification\_table()}\spxextra{in module IPTK.IO.InFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.InFunctions.parse_xml_based_format_to_identification_table}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.InFunctions.}}\sphinxbfcode{\sphinxupquote{parse\_xml\_based\_format\_to\_identification\_table}}}{\emph{\DUrole{n}{path2XML\_file}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{path2fastaDB}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{decoy\_prefix}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}DECOY\textquotesingle{}}}, \emph{\DUrole{n}{is\_idXML}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{fasta\_reader\_param}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\textquotesingle{}decoy\_string\textquotesingle{}: \textquotesingle{}DECOY\textquotesingle{}, \textquotesingle{}filter\_decoy\textquotesingle{}: True\}}}, \emph{\DUrole{n}{remove\_if\_not\_matched}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{True}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
parse either a pepXML or an idXML file to generate an identification table ,
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2XML\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to the input pepXML files

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2fastaDB}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to a fasta sequence database to obtain the protein sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{decoy\_prefix}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the prefix of the decoy sequences, default is DECOY, defaults to ‘DECOY’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{is\_idXML}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} Whether or not the provided file is an idXML, default is false which assume the provided file is a pepXML file, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{fasta\_reader\_param}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict of parameters for controlling the behavior of the fasta reader , defaults to \{‘filter\_decoy’:True, ‘decoy\_string’:’DECOY’ \}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{remove\_if\_not\_matched}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} remove the peptide if it could not be matched to the parent protein,, defaults to True

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} if the fasta database could not be open

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the XML file can not be open

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{KeyError}} \textendash{} if a protein id defined in the mzTab file could not be extracted from a matched sequence database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the peptide can not be mapped to the identified protein

\end{itemize}

\item[{Returns}] \leavevmode
the identification table

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{IPTK.IO.MEMEInterface module}
\label{\detokenize{IPTK.IO:module-IPTK.IO.MEMEInterface}}\label{\detokenize{IPTK.IO:iptk-io-memeinterface-module}}\index{module@\spxentry{module}!IPTK.IO.MEMEInterface@\spxentry{IPTK.IO.MEMEInterface}}\index{IPTK.IO.MEMEInterface@\spxentry{IPTK.IO.MEMEInterface}!module@\spxentry{module}}
The module contains functions to to call meme software via a system call.
\index{call\_meme() (in module IPTK.IO.MEMEInterface)@\spxentry{call\_meme()}\spxextra{in module IPTK.IO.MEMEInterface}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.MEMEInterface.call_meme}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.MEMEInterface.}}\sphinxbfcode{\sphinxupquote{call\_meme}}}{\emph{\DUrole{n}{input\_fasta\_file}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{output\_dir}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{verbose}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{True}}, \emph{\DUrole{n}{objfunc}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}classic\textquotesingle{}}}, \emph{\DUrole{n}{test}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}mhg\textquotesingle{}}}, \emph{\DUrole{n}{use\_llr}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{shuf}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{2}}, \emph{\DUrole{n}{hsfrac}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{0.5}}, \emph{\DUrole{n}{cefrac}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{0.25}}, \emph{\DUrole{n}{searchsize}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{maxsize}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{norand}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{csites}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{seed}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{mod}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}oops\textquotesingle{}}}, \emph{\DUrole{n}{nmotifs}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{evt}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1.0}}, \emph{\DUrole{n}{time}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{nsite}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{minsites}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{maxsite}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{nsites}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{w}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{minw}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{maxw}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{nomatrim}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{wg}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{ws}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{noendgaps}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{maxiter}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{prior}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}dirichlet\textquotesingle{}}}, \emph{\DUrole{n}{b}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{p}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}}{{ $\rightarrow$ None}}
warper for making a system call to meme software for sequence motif finding         for the reset of the function parameters use the function             \sphinxstylestrong{get\_meme\_help} defined in the module IO, submodule MEMEInterface. 
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{input\_fasta\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to input FASTA files.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{output\_dir}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the output dir to write the results, \sphinxstylestrong{IT WILL OVERWRITE EXISTING DIRECTORY}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{verbose}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}) \textendash{} whether or not to print the output of calling meme to the screen, default is True.

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_meme\_help() (in module IPTK.IO.MEMEInterface)@\spxentry{get\_meme\_help()}\spxextra{in module IPTK.IO.MEMEInterface}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.MEMEInterface.get_meme_help}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.MEMEInterface.}}\sphinxbfcode{\sphinxupquote{get\_meme\_help}}}{}{{ $\rightarrow$ None}}
print the command line help interface for the meme tool
\begin{quote}\begin{description}
\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{FileNotFoundError}} \textendash{} if meme is not callable

\end{description}\end{quote}

\end{fulllineitems}

\index{is\_meme\_callable() (in module IPTK.IO.MEMEInterface)@\spxentry{is\_meme\_callable()}\spxextra{in module IPTK.IO.MEMEInterface}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.MEMEInterface.is_meme_callable}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.MEMEInterface.}}\sphinxbfcode{\sphinxupquote{is\_meme\_callable}}}{}{{ $\rightarrow$ bool}}~\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
True if meme is callable, False otherwise.

\item[{Return type}] \leavevmode
bool

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{IPTK.IO.OutFunctions module}
\label{\detokenize{IPTK.IO:module-IPTK.IO.OutFunctions}}\label{\detokenize{IPTK.IO:iptk-io-outfunctions-module}}\index{module@\spxentry{module}!IPTK.IO.OutFunctions@\spxentry{IPTK.IO.OutFunctions}}\index{IPTK.IO.OutFunctions@\spxentry{IPTK.IO.OutFunctions}!module@\spxentry{module}}
Write the results generated by the library into a wide variety of formats.
\index{write\_annotated\_sequences() (in module IPTK.IO.OutFunctions)@\spxentry{write\_annotated\_sequences()}\spxextra{in module IPTK.IO.OutFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.OutFunctions.write_annotated_sequences}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.OutFunctions.}}\sphinxbfcode{\sphinxupquote{write\_annotated\_sequences}}}{\emph{\DUrole{n}{peptides}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}, \emph{\DUrole{n}{labels}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}int\DUrole{p}{{]}}}}, \emph{\DUrole{n}{path2write}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{sep}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{},\textquotesingle{}}}, \emph{\DUrole{n}{shuffle}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{True}}}{{ $\rightarrow$ None}}
take a list of peptides along with it sequences and write the results to a CSV file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptides}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a list of peptide sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{labels}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a list of numerical labels associated with the peptides

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2write}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the path to write the generated file

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sep}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The separator in the resulting table,  defaults to ‘,’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{shuffle}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} Whether or not to shuffle the table , defaults to True

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} incase the length of the tables and labels is not matching

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} In case writing the output table failed

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{write\_auto\_named\_peptide\_to\_fasta() (in module IPTK.IO.OutFunctions)@\spxentry{write\_auto\_named\_peptide\_to\_fasta()}\spxextra{in module IPTK.IO.OutFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.OutFunctions.write_auto_named_peptide_to_fasta}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.OutFunctions.}}\sphinxbfcode{\sphinxupquote{write\_auto\_named\_peptide\_to\_fasta}}}{\emph{\DUrole{n}{peptides}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}, \emph{\DUrole{n}{output\_file}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ None}}
Takes a list of peptides, generate automatic names for the peptides and write the results to the disk as FASTA files
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptides}} (\sphinxstyleliteralemphasis{\sphinxupquote{Peptides}}) \textendash{} a list of peptide sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{output\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the output file to write the results to, it will OVERWRITE existing files

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{write\_mapped\_tensor\_to\_h5py() (in module IPTK.IO.OutFunctions)@\spxentry{write\_mapped\_tensor\_to\_h5py()}\spxextra{in module IPTK.IO.OutFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.OutFunctions.write_mapped_tensor_to_h5py}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.OutFunctions.}}\sphinxbfcode{\sphinxupquote{write\_mapped\_tensor\_to\_h5py}}}{\emph{\DUrole{n}{tensor}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}, \emph{\DUrole{n}{path2write}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{dataSet\_name}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}MAPPED\_TENSOR\textquotesingle{}}}}{{ $\rightarrow$ None}}
Write a mapped tensor to an hdf5 file
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tensor}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} The provided tensor to  write it to the hdf5 file.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2write}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path of the output file

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{dataSet\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The name of the dataset inside the mapped tensor, defaults to ‘MAPPED\_TENSOR’

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} In case opening the file for writing failed

\end{description}\end{quote}

\end{fulllineitems}

\index{write\_named\_peptides\_to\_fasta() (in module IPTK.IO.OutFunctions)@\spxentry{write\_named\_peptides\_to\_fasta()}\spxextra{in module IPTK.IO.OutFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.OutFunctions.write_named_peptides_to_fasta}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.OutFunctions.}}\sphinxbfcode{\sphinxupquote{write\_named\_peptides\_to\_fasta}}}{\emph{\DUrole{n}{names}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}, \emph{\DUrole{n}{peptides}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}, \emph{\DUrole{n}{output\_file}\DUrole{p}{:} \DUrole{n}{str}}}{}
Takes a list of names and peptide sequences and writes them as an output file to the disk as fasta files
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{names}} (\sphinxstyleliteralemphasis{\sphinxupquote{Names}}) \textendash{} A list of sequences names

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptides}} (\sphinxstyleliteralemphasis{\sphinxupquote{Peptides}}) \textendash{} A list of peptide sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{output\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the output file to write the results to, it will OVERWRITE existing files

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} Incase the length of the tables and labels is not matching

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} In case writing the output file failed

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{write\_pep\_file() (in module IPTK.IO.OutFunctions)@\spxentry{write\_pep\_file()}\spxextra{in module IPTK.IO.OutFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.IO:IPTK.IO.OutFunctions.write_pep_file}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.IO.OutFunctions.}}\sphinxbfcode{\sphinxupquote{write\_pep\_file}}}{\emph{\DUrole{n}{peptides}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Peptide.Peptide}]{\sphinxcrossref{IPTK.Classes.Peptide.Peptide}}}\DUrole{p}{{]}}}}, \emph{\DUrole{n}{output\_file}\DUrole{p}{:} \DUrole{n}{str}}}{}
Takes a file and write the peptides to .pep file which is a text file that contain one peptide per line
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptides}} (\sphinxstyleliteralemphasis{\sphinxupquote{Peptides}}) \textendash{} a list of peptide sequences

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{output\_file}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the name of the output file to write the results to, it will OVERWRITE existing files

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} In case writing the output file failed

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{Module contents}
\label{\detokenize{IPTK.IO:module-IPTK.IO}}\label{\detokenize{IPTK.IO:module-contents}}\index{module@\spxentry{module}!IPTK.IO@\spxentry{IPTK.IO}}\index{IPTK.IO@\spxentry{IPTK.IO}!module@\spxentry{module}}

\subparagraph{IPTK.Utils package}
\label{\detokenize{IPTK.Utils:iptk-utils-package}}\label{\detokenize{IPTK.Utils::doc}}

\subparagraph{Submodules}
\label{\detokenize{IPTK.Utils:submodules}}

\subparagraph{IPTK.Utils.DevFunctions module}
\label{\detokenize{IPTK.Utils:module-IPTK.Utils.DevFunctions}}\label{\detokenize{IPTK.Utils:iptk-utils-devfunctions-module}}\index{module@\spxentry{module}!IPTK.Utils.DevFunctions@\spxentry{IPTK.Utils.DevFunctions}}\index{IPTK.Utils.DevFunctions@\spxentry{IPTK.Utils.DevFunctions}!module@\spxentry{module}}
The module contain functions that can be used for developing \& testing other functions of the library
\index{generate\_random\_peptide\_seq() (in module IPTK.Utils.DevFunctions)@\spxentry{generate\_random\_peptide\_seq()}\spxextra{in module IPTK.Utils.DevFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.DevFunctions.generate_random_peptide_seq}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.DevFunctions.}}\sphinxbfcode{\sphinxupquote{generate\_random\_peptide\_seq}}}{\emph{\DUrole{n}{peptide\_length}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{num\_peptides}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
generate a list of random peptides for testing and developing purposes.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptide\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The peptide length

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_peptides}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of peptides in the generate list

\end{itemize}

\item[{Returns}] \leavevmode
a list of random peptides

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_an\_experimental\_ident\_table\_from\_fasta() (in module IPTK.Utils.DevFunctions)@\spxentry{simulate\_an\_experimental\_ident\_table\_from\_fasta()}\spxextra{in module IPTK.Utils.DevFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.DevFunctions.simulate_an_experimental_ident_table_from_fasta}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.DevFunctions.}}\sphinxbfcode{\sphinxupquote{simulate\_an\_experimental\_ident\_table\_from\_fasta}}}{\emph{\DUrole{n}{path2load}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{num\_pep}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{num\_prot}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
simulate an IP identification table from a fasta file. Please Note,  if the reminder of num\_pep over num\_prot does not equal to zero,
the floor of this ratio will be used to sample peptides from each proteins
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2load}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to load the Fasta files

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_pep}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of peptides in the tables

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_prot}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of UNIQUE proteins in the table

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if number of proteins or number of peptide is zero

\item[{Returns}] \leavevmode
an identification table

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_an\_expression\_table() (in module IPTK.Utils.DevFunctions)@\spxentry{simulate\_an\_expression\_table()}\spxextra{in module IPTK.Utils.DevFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.DevFunctions.simulate_an_expression_table}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.DevFunctions.}}\sphinxbfcode{\sphinxupquote{simulate\_an\_expression\_table}}}{\emph{\DUrole{n}{num\_transcripts}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{100}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
create a dummy expression table to be used for testing and developing Tissue based classes
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{num\_transcripts}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The number of transcripts that shall be contained in the transcript , defaults to 100

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} incase number of transcripts is 0

\item[{Returns}] \leavevmode
{[}description{]}

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_mapped\_array\_list() (in module IPTK.Utils.DevFunctions)@\spxentry{simulate\_mapped\_array\_list()}\spxextra{in module IPTK.Utils.DevFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.DevFunctions.simulate_mapped_array_list}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.DevFunctions.}}\sphinxbfcode{\sphinxupquote{simulate\_mapped\_array\_list}}}{\emph{\DUrole{n}{min\_len}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{20}}, \emph{\DUrole{n}{max\_len}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{100}}, \emph{\DUrole{n}{num\_elem}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{100}}}{{ $\rightarrow$ List\DUrole{p}{{[}}numpy.ndarray\DUrole{p}{{]}}}}
Simulate a list of mapped arrays proteins to be used for developing purposes
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{min\_len}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the minmum length of the protein , defaults to 20

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{max\_len}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the maximum length for the protein , defaults to 100

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_elem}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the number of arrays in the protein , defaults to 100

\end{itemize}

\item[{Returns}] \leavevmode
a list of simulated NumPy array that represent protein peptide coverage

\item[{Return type}] \leavevmode
List{[}np.ndarray{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_random\_experiment() (in module IPTK.Utils.DevFunctions)@\spxentry{simulate\_random\_experiment()}\spxextra{in module IPTK.Utils.DevFunctions}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.DevFunctions.simulate_random_experiment}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.DevFunctions.}}\sphinxbfcode{\sphinxupquote{simulate\_random\_experiment}}}{\emph{\DUrole{n}{alleles}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}, \emph{\DUrole{n}{path2fasta}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{tissue\_name}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}TEST\_TISSUE\textquotesingle{}}}, \emph{\DUrole{n}{num\_pep}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{10}}, \emph{\DUrole{n}{num\_prot}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{5}}, \emph{\DUrole{n}{proband\_name}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{None}}}{{ $\rightarrow$ {\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{IPTK.Classes.Experiment.Experiment}}}}}
Simulate a random experiment objects
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alleles}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a list of alleles names.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2fasta}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to load the database objects

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tissue\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the name of the tissue, defaults to ‘TEST\_TISSUE’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_pep}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the number of peptides in the table, defaults to 10

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_prot}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} number of proteins, defaults to 5

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{proband\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The name of the Proband, defaults to None

\end{itemize}

\item[{Returns}] \leavevmode
a simulated experimental object

\item[{Return type}] \leavevmode
{\hyperref[\detokenize{IPTK.Classes:IPTK.Classes.Experiment.Experiment}]{\sphinxcrossref{Experiment}}}

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{IPTK.Utils.Mapping module}
\label{\detokenize{IPTK.Utils:module-IPTK.Utils.Mapping}}\label{\detokenize{IPTK.Utils:iptk-utils-mapping-module}}\index{module@\spxentry{module}!IPTK.Utils.Mapping@\spxentry{IPTK.Utils.Mapping}}\index{IPTK.Utils.Mapping@\spxentry{IPTK.Utils.Mapping}!module@\spxentry{module}}
A submodule that contain function to map different database keys
\index{map\_from\_uniprot\_gene() (in module IPTK.Utils.Mapping)@\spxentry{map\_from\_uniprot\_gene()}\spxextra{in module IPTK.Utils.Mapping}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.Mapping.map_from_uniprot_gene}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.Mapping.}}\sphinxbfcode{\sphinxupquote{map\_from\_uniprot\_gene}}}{\emph{\DUrole{n}{uniprots}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
map from uniprot id to ensemble gene ids
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{uniprots}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a list of uniprot IDs

\item[{Returns}] \leavevmode
A table that contain the mapping between each uniprot and its corresponding Gene ID/IDs

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{map\_from\_uniprot\_pdb() (in module IPTK.Utils.Mapping)@\spxentry{map\_from\_uniprot\_pdb()}\spxextra{in module IPTK.Utils.Mapping}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.Mapping.map_from_uniprot_pdb}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.Mapping.}}\sphinxbfcode{\sphinxupquote{map\_from\_uniprot\_pdb}}}{\emph{\DUrole{n}{uniprots}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
map from uniprot id to protein data bank identifiers
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{uniprots}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a list of uniprot IDs

\item[{Returns}] \leavevmode
A table that contain the mapping between each uniprot and its corresponding PDB ID/IDs

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{IPTK.Utils.Types module}
\label{\detokenize{IPTK.Utils:module-IPTK.Utils.Types}}\label{\detokenize{IPTK.Utils:iptk-utils-types-module}}\index{module@\spxentry{module}!IPTK.Utils.Types@\spxentry{IPTK.Utils.Types}}\index{IPTK.Utils.Types@\spxentry{IPTK.Utils.Types}!module@\spxentry{module}}
Contain a definition of commonly used types through the library


\subparagraph{IPTK.Utils.UtilityFunction module}
\label{\detokenize{IPTK.Utils:module-IPTK.Utils.UtilityFunction}}\label{\detokenize{IPTK.Utils:iptk-utils-utilityfunction-module}}\index{module@\spxentry{module}!IPTK.Utils.UtilityFunction@\spxentry{IPTK.Utils.UtilityFunction}}\index{IPTK.Utils.UtilityFunction@\spxentry{IPTK.Utils.UtilityFunction}!module@\spxentry{module}}
Utility functions that are used through the library
\index{append\_to\_calling\_string() (in module IPTK.Utils.UtilityFunction)@\spxentry{append\_to\_calling\_string()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.append_to_calling_string}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{append\_to\_calling\_string}}}{\emph{\DUrole{n}{param}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{def\_value}}, \emph{\DUrole{n}{cur\_val}}, \emph{\DUrole{n}{calling\_string}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{is\_flag}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}}{{ $\rightarrow$ str}}
help function that take a calling string, a parameter, a default value and current value     if the parameter does not equal its default value the function append the parameter with its current 
value to the calling string adding a space before the calling\_string.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{param}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the parameter that will be append to the calling string

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{def\_value}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} The default value for the parameter

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{cur\_val}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} The current value for the parameter

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{calling\_string}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The calling string in which the parameter and the current value might be appended to it

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{is\_flag}} \textendash{} If the parameter is a control flag, i.e. a boolean switch, it append the parameter to the calling 

\end{itemize}

\end{description}\end{quote}

string without associating a value to it , defaults to False
:type is\_flag: bool, optional
:return: the updated version of the calling string 
:rtype: str

\end{fulllineitems}

\index{build\_sequence\_table() (in module IPTK.Utils.UtilityFunction)@\spxentry{build\_sequence\_table()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.build_sequence_table}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{build\_sequence\_table}}}{\emph{\DUrole{n}{sequence\_dict}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}}}}{{ $\rightarrow$ pandas.core.frame.DataFrame}}
construct a sequences database from sequences dict object
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{sequence\_dict}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a dict that contain the protein ids as keys and sequences as values.

\item[{Returns}] \leavevmode
pandas dataframe that contain the protein ID and the associated protein sequence

\item[{Return type}] \leavevmode
pd.DataFrame

\end{description}\end{quote}

\end{fulllineitems}

\index{check\_peptide\_made\_of\_std\_20\_aa() (in module IPTK.Utils.UtilityFunction)@\spxentry{check\_peptide\_made\_of\_std\_20\_aa()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.check_peptide_made_of_std_20_aa}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{check\_peptide\_made\_of\_std\_20\_aa}}}{\emph{\DUrole{n}{peptide}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ str}}
Check if the peptide is made of the standard 20 amino acids, if this is the case, 
it return the peptide sequence, otherwise it return an empty string
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{peptide}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} a peptide sequence to check its composition

\item[{Returns}] \leavevmode
True, if the peptide is made of the standard 20 amino acids, False otherwise.

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{generate\_color\_scale() (in module IPTK.Utils.UtilityFunction)@\spxentry{generate\_color\_scale()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.generate_color_scale}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{generate\_color\_scale}}}{\emph{\DUrole{n}{color\_ranges}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ matplotlib.colors.LinearSegmentedColormap}}
generate a color gradient with number of steps equal to color\_ranges \sphinxhyphen{}1
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{color\_ranges}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the number of colors in the range

\item[{Returns}] \leavevmode
a color gradient palette

\item[{Return type}] \leavevmode
matplotlib.colors.LinearSegmentedColormap

\end{description}\end{quote}

\end{fulllineitems}

\index{generate\_random\_name() (in module IPTK.Utils.UtilityFunction)@\spxentry{generate\_random\_name()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.generate_random_name}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{generate\_random\_name}}}{\emph{\DUrole{n}{name\_length}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ str}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{name\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} Generate a random ASCII based string

\item[{Returns}] \leavevmode
{[}description{]}

\item[{Return type}] \leavevmode
str

\end{description}\end{quote}

\end{fulllineitems}

\index{generate\_random\_protein\_mapping() (in module IPTK.Utils.UtilityFunction)@\spxentry{generate\_random\_protein\_mapping()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.generate_random_protein_mapping}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{generate\_random\_protein\_mapping}}}{\emph{\DUrole{n}{protein\_len}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{max\_coverage}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ numpy.ndarray}}
Generate a NumPy array with shape of 1 by protein\_len where the elements in the array 
is a random integer between zero \&  max\_coverage.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_len}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The length of the protein

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{max\_coverage}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The maximum peptide coverage at each position

\end{itemize}

\item[{Returns}] \leavevmode
a NumPy array contain a simulate protein coverage

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{get\_idx\_peptide\_in\_sequence\_table() (in module IPTK.Utils.UtilityFunction)@\spxentry{get\_idx\_peptide\_in\_sequence\_table()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.get_idx_peptide_in_sequence_table}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{get\_idx\_peptide\_in\_sequence\_table}}}{\emph{\DUrole{n}{sequence\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{peptide}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ List\DUrole{p}{{[}}str\DUrole{p}{{]}}}}
check the sequences table if the provided peptide is locate in one of its sequences and returns 
a list of protein identifiers containing the identifier of the hit proteins.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sequence\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} pandas dataframe that contain the protein ID and the associated protein sequence

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{peptide}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} the peptide sequence to query the protein with

\end{itemize}

\item[{Returns}] \leavevmode
a list of protein identifiers containing the identifier of the hit proteins

\item[{Return type}] \leavevmode
List{[}str{]}

\end{description}\end{quote}

\end{fulllineitems}

\index{load\_3d\_figure() (in module IPTK.Utils.UtilityFunction)@\spxentry{load\_3d\_figure()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.load_3d_figure}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{load\_3d\_figure}}}{\emph{\DUrole{n}{file\_path}\DUrole{p}{:} \DUrole{n}{str}}}{{ $\rightarrow$ matplotlib.figure.Figure}}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{file\_path}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} Load a pickled 3D figure from thr provided path

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} The path of the pickled figure.

\item[{Returns}] \leavevmode
a matplotlib figure

\item[{Return type}] \leavevmode
plt.Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{pad\_mapped\_proteins() (in module IPTK.Utils.UtilityFunction)@\spxentry{pad\_mapped\_proteins()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.pad_mapped_proteins}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{pad\_mapped\_proteins}}}{\emph{\DUrole{n}{list\_array}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}numpy.ndarray\DUrole{p}{{]}}}}, \emph{\DUrole{n}{pre\_pad}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{True}}, \emph{\DUrole{n}{padding\_char}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}}{{ $\rightarrow$ numpy.ndarray}}
Pad the provided list of array into a 2D tensor of shape number of arrays by maxlength.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{list\_array}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} A list of NumPy arrays where each array is a mapped\_protein array,     the expected shape of these arrays is 1 by protein length.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pre\_pad}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} pre or post padding of shorter array in the library.Default is pre\sphinxhyphen{}padding, defaults to True

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{padding\_char}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The padding char, defaults to \sphinxhyphen{}1

\end{itemize}

\item[{Returns}] \leavevmode
A 2D tensor of shape number of arrays by maxlength.

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{save\_3d\_figure() (in module IPTK.Utils.UtilityFunction)@\spxentry{save\_3d\_figure()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.save_3d_figure}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{save\_3d\_figure}}}{\emph{\DUrole{n}{outpath}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{fig2save}\DUrole{p}{:} \DUrole{n}{matplotlib.figure.Figure}}}{{ $\rightarrow$ None}}
write a pickled version of the a 3D figure so it can be loaded later for more interactive analysis
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{outpath}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The output path of the writer function

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{fig2save}} (\sphinxstyleliteralemphasis{\sphinxupquote{plt.Figure}}) \textendash{} The figure to save to the output file

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} In case writing the file failed

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_protein\_binary\_represention() (in module IPTK.Utils.UtilityFunction)@\spxentry{simulate\_protein\_binary\_represention()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.simulate_protein_binary_represention}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{simulate\_protein\_binary\_represention}}}{\emph{\DUrole{n}{num\_conditions}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{protein\_length}\DUrole{p}{:} \DUrole{n}{int}}}{}~\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_conditions}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The number of conditions to simulate

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} The Length of the protein

\end{itemize}

\item[{Returns}] \leavevmode
A 2D matrix of shape protein\_length by number of conditions, where each element can be either zero.

\item[{Return type}] \leavevmode
np.ndarray

\end{description}\end{quote}

\end{fulllineitems}

\index{simulate\_protein\_representation() (in module IPTK.Utils.UtilityFunction)@\spxentry{simulate\_protein\_representation()}\spxextra{in module IPTK.Utils.UtilityFunction}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Utils:IPTK.Utils.UtilityFunction.simulate_protein_representation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Utils.UtilityFunction.}}\sphinxbfcode{\sphinxupquote{simulate\_protein\_representation}}}{\emph{\DUrole{n}{num\_conditions}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{protein\_len}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{protein\_coverage}\DUrole{p}{:} \DUrole{n}{int}}}{{ $\rightarrow$ Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}}}
Simulate protein peptide coverage under\sphinxhyphen{}different conditions
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_conditions}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} The number of condition to simulate

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_len}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} The length of the protein

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_coverage}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} The maximum protein coverage

\end{itemize}

\item[{Returns}] \leavevmode
a dict of length num\_conditions contains that the condition index and a simulated protein array

\item[{Return type}] \leavevmode
Dict{[}str, np.ndarray{]}

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{Module contents}
\label{\detokenize{IPTK.Utils:module-IPTK.Utils}}\label{\detokenize{IPTK.Utils:module-contents}}\index{module@\spxentry{module}!IPTK.Utils@\spxentry{IPTK.Utils}}\index{IPTK.Utils@\spxentry{IPTK.Utils}!module@\spxentry{module}}

\subparagraph{IPTK.Visualization package}
\label{\detokenize{IPTK.Visualization:iptk-visualization-package}}\label{\detokenize{IPTK.Visualization::doc}}

\subparagraph{Submodules}
\label{\detokenize{IPTK.Visualization:submodules}}

\subparagraph{IPTK.Visualization.vizTools module}
\label{\detokenize{IPTK.Visualization:module-IPTK.Visualization.vizTools}}\label{\detokenize{IPTK.Visualization:iptk-visualization-viztools-module}}\index{module@\spxentry{module}!IPTK.Visualization.vizTools@\spxentry{IPTK.Visualization.vizTools}}\index{IPTK.Visualization.vizTools@\spxentry{IPTK.Visualization.vizTools}!module@\spxentry{module}}
The module contain visualization functions the can be used to plot the results obtained from the 
methods of the classes defined in the Class module or from the analysis functions defined in the Analysis Module.
\index{imposed\_coverage\_on\_3D\_structure() (in module IPTK.Visualization.vizTools)@\spxentry{imposed\_coverage\_on\_3D\_structure()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.imposed_coverage_on_3D_structure}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{imposed\_coverage\_on\_3D\_structure}}}{\emph{\DUrole{n}{path2mmCIF}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{mapped\_protein}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}, \emph{\DUrole{n}{background\_color}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}black\textquotesingle{}}}, \emph{\DUrole{n}{low}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}red\textquotesingle{}}}, \emph{\DUrole{n}{high}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}blue\textquotesingle{}}}}{{ $\rightarrow$ None}}
A function to impose the peptide coverage on top of a protein 3D structure     where the color of each position is marked by a color gradient that reflect the number of peptides 
aligned to this position. Note: Use the function with Jupyter\sphinxhyphen{}note book as it return an NGLWidget object that your can explore     and navigate from the browser.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{path2mmCIF}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The path to load the mmCIF file containing the protein structure

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{mapped\_protein}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} a Numpy array of containg the number of peptides originated from each position in the array

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{background\_color}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the color of the background, default is black , defaults to ‘black’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{low}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the color of low covered position, default is red. , defaults to ‘red’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{high}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the color of high covered position, default is violet., defaults to ‘blue’

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} incase the provided path to the structure file does not exists

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IOError}} \textendash{} if the structure file can not be loaded or if more than one file are located in the provided path

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_change\_in\_presentation\_between\_experiment() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_change\_in\_presentation\_between\_experiment()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_change_in_presentation_between_experiment}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_change\_in\_presentation\_between\_experiment}}}{\emph{\DUrole{n}{change\_in\_presentation\_array}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}, \emph{\DUrole{n}{index\_first}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{index\_second}\DUrole{p}{:} \DUrole{n}{int}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{title}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}Change in protein presentation\textquotesingle{}}}, \emph{\DUrole{n}{xlabel}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}Proteins\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{}magnitude of change in protein count\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the change in protein presentation between two experiment
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{change\_in\_presentation\_array}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} a 3D tensor of shape number of experiments by number of experiment by number of identified proteins.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{index\_first}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} {[}description{]}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{index\_second}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}) \textendash{} the index of the first experiment in the tensor

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.scatterplot function, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Change in protein presentation’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The axis on the x\sphinxhyphen{}axis , defaults to “Proteins”

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The axis on the y\sphinxhyphen{}axis, defaults to “magnitude of change in protein count”

\end{itemize}

\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the provided tensor is not of rank 3

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{IndexError}} \textendash{} if the provided indices are out of bound

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_experiment\_set\_counts() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_experiment\_set\_counts()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_experiment_set_counts}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_experiment\_set\_counts}}}{\emph{\DUrole{n}{counts\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{log\_scale}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}}{{ $\rightarrow$ matplotlib.pyplot.figure}}
visualize the number of peptides and number of peptides\sphinxhyphen{}per\sphinxhyphen{}organism per experiment.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{counts\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} a pandas dataframe that contain the count, organism name and the count

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log\_scale}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} Normalize the peptide counts one log 10, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.catplot function, defaults to \{\}

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_gene\_expression\_vs\_num\_peptides() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_gene\_expression\_vs\_num\_peptides()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_gene_expression_vs_num_peptides}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_gene\_expression\_vs\_num\_peptides}}}{\emph{\DUrole{n}{exp\_count\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{tissue\_name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{def\_value}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Expression value\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Peptides per protein Vs. Expression Level\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
Plot the correlation between the gene expression and the num of peptids per protein using seaborn library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp\_count\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the number of peptides and the expresion value for each protein in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tissue\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the tissue

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{def\_value}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default value for proteins that could not be mapped to the expression database, defaults to \sphinxhyphen{}1

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.scatter function, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis, defaults to ‘Expression value’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Peptides per protein Vs. Expression Level’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_motif() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_motif()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_motif}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_motif}}}{\emph{\DUrole{n}{pwm\_df}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\textquotesingle{}fade\_probabilities\textquotesingle{}: True\}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
A generic motif plotter that forward its argument to logomaker for making plots
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pwm\_df}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A pandas dataframe containing the position weighted matrix

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{PlottingKeywards}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} A dictionary of parameter to control the behavior of the logo plotter, defaults to \{‘fade\_probabilities’:True\}

\end{itemize}

\item[{Returns}] \leavevmode
a matplotlib figure instance containing the ploted motif

\item[{Return type}] \leavevmode
plt.Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_num\_peptide\_per\_go\_term() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_num\_peptide\_per\_go\_term()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_num_peptide_per_go_term}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_num\_peptide\_per\_go\_term}}}{\emph{\DUrole{n}{pep2goTerm}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}GO\sphinxhyphen{}Term\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per GO Term\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the number of peptides obtained per Go\sphinxhyphen{}Term  using matplotlib library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep2goTerm}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of peptides from each GO\sphinxhyphen{}Term

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.barplot function, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop peptide with unknown GO\sphinxhyphen{}term, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis, defaults to ‘GO\sphinxhyphen{}Term’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Number of peptides per GO Term’

\end{itemize}

\item[{Returns}] \leavevmode
{[}description{]}

\item[{Return type}] \leavevmode
plt.Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_num\_peptides\_per\_location() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_num\_peptides\_per\_location()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_num_peptides_per_location}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_num\_peptides\_per\_location}}}{\emph{\DUrole{n}{pep2loc}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Compartment\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per sub\sphinxhyphen{}cellular compartment\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the number of peptides obtained from each compartment using seaborn library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep2loc}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of peptides from each  location

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.barplot function, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop protein with unknown location, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the y\sphinxhyphen{}axis, defaults to ‘Compartment’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Number of peptides per sub\sphinxhyphen{}cellular compartment’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_num\_peptides\_per\_organism() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_num\_peptides\_per\_organism()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_num_peptides_per_organism}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_num\_peptides\_per\_organism}}}{\emph{\DUrole{n}{pep\_per\_org}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{log\_scale}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Organism\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per organism\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the number of peptides per each organism inferred from the experiment using seaborn and matlotlib.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep\_per\_org}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the number of peptides belonging to each organism

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log\_scale}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} Whether or not to scale the number of peptides using a log scale, default is False, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.barplot function, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the y\sphinxhyphen{}axis, defaults to ‘Organism’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Number of peptides per organism’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_num\_peptides\_per\_parent() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_num\_peptides\_per\_parent()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_num_peptides_per_parent}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_num\_peptides\_per\_parent}}}{\emph{\DUrole{n}{nums\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{num\_prot}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{x\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{y\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Protein ID\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per protein\textquotesingle{}}}}{}
Visualize a histogram of the eluted peptide length.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{nums\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} a pandas dataframe containing number of peptides identified from each protein.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_prot}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the number of protein to show relative to the first element, for example, the first 10, 20 etc.     If the default value of \sphinxhyphen{}1 is used then all protein will be plotted, however, this might lead to a crowded figure,     defaults to \sphinxhyphen{}1.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} \textendash{} a dict object containing parameters for the function     seaborn::distplot, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{x\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{y\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the y\sphinxhyphen{}axis, defaults to ‘Protein ID’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Number of peptides per protein’

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the num\_prot is bigger than the number of elements in the provided table

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_num\_protein\_per\_go\_term() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_num\_protein\_per\_go\_term()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_num_protein_per_go_term}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_num\_protein\_per\_go\_term}}}{\emph{\DUrole{n}{protein2goTerm}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{tissue\_name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of Proteins\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Compartment\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of proteins per sub\sphinxhyphen{}cellular compartment\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the number of proteins per each GO Term
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein2goTerm}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of proteins from each GO\sphinxhyphen{}Term

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tissue\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} a dict object containing parameters for the sns.barplot function.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.barplot function, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop protein with unknown location, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of Proteins’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis, defaults to ‘Compartment’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure, defaults to ‘Number of proteins per sub\sphinxhyphen{}cellular compartment’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_num\_protein\_per\_location() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_num\_protein\_per\_location()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_num_protein_per_location}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_num\_protein\_per\_location}}}{\emph{\DUrole{n}{protein\_loc}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of Proteins\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Compartment\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of proteins per sub\sphinxhyphen{}cellular compartment\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the number of proteins per each sub\sphinxhyphen{}cellular compartment
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_loc}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of protein from each location.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.barplot function, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop protein with unknown location, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of Proteins’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis, defaults to ‘Compartment’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure, defaults to ‘Number of proteins per sub\sphinxhyphen{}cellular compartment’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_overlap\_heatmap() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_overlap\_heatmap()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_overlap_heatmap}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_overlap\_heatmap}}}{\emph{\DUrole{n}{results\_df}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}}{{ $\rightarrow$ seaborn.matrix.ClusterGrid}}
Plot a user provided dataframe as a cluster heatmap using seaborn library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{results\_df}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A pandas dataframe table that hold the overlapping number

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{PlottingKeywards}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} forward parameter to the clustermap function, defaults to \{\}

\end{itemize}

\item[{Returns}] \leavevmode
the generated clustermap

\item[{Return type}] \leavevmode
sns.matrix.ClusterGrid

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_overlay\_representation() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_overlay\_representation()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_overlay_representation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_overlay\_representation}}}{\emph{\DUrole{n}{proteins}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}\DUrole{p}{{]}}}}, \emph{\DUrole{n}{alpha}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{0.5}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{None}}, \emph{\DUrole{n}{legend\_pos}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{2}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot an overlayed representation for the SAME protein or proteins OF EQUAL LENGTH in different conditions        in which the mapped represention of each protein are stacked on top of each other to generate        a representation for the protein representability under different conditions.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{proteins}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a nested dict object containing for each protein a child dict that contain the mapping array and the color in the figure.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the transparency between proteins , defaults to 0.5

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to None

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{legend\_pos}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the position of the legend , defaults to 2

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the provided protein have different lengths

\item[{Returns}] \leavevmode
a matplotlib figure containing the mapped representation

\item[{Return type}] \leavevmode
plt.Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_paired\_represention() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_paired\_represention()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_paired_represention}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_paired\_represention}}}{\emph{\DUrole{n}{protein\_one\_repr}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}}}, \emph{\DUrole{n}{protein\_two\_repr}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}}}, \emph{\DUrole{n}{color\_first}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}red\textquotesingle{}}}, \emph{\DUrole{n}{color\_second}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}blue\textquotesingle{}}}, \emph{\DUrole{n}{alpha}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{0.9}}, \emph{\DUrole{n}{title}\DUrole{o}{=}\DUrole{default_value}{\textquotesingle{} Parired protein representation\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
visualize the difference in representation for the same protein between two experiments using matplotlib library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_one\_repr}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a dict object containing the legand of the first condition along with its mapped array

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_two\_repr}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a dict object containing the legand of the second condition along with its mapped array

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{alpha}} (\sphinxstyleliteralemphasis{\sphinxupquote{the transparency of the figure.}}) \textendash{} the transparency of the figure.

\end{itemize}

\item[{Param}] \leavevmode
color\_first: the color of representation for the first condition

\item[{Param}] \leavevmode
color\_second: the color of the second condition

\item[{Param}] \leavevmode
title: the title of the figure.

\item[{Returns}] \leavevmode
A matplotlib Figure containing the representation

\item[{Return type}] \leavevmode
plt.Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_parent\_protein\_expression\_in\_tissue() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_parent\_protein\_expression\_in\_tissue()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_parent_protein_expression_in_tissue}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_parent\_protein\_expression\_in\_tissue}}}{\emph{\DUrole{n}{expression\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{ref\_expression}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{tissue\_name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{sampling\_num}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{10}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\textquotesingle{}orient\textquotesingle{}: \textquotesingle{}v\textquotesingle{}\}}}, \emph{\DUrole{n}{def\_value}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Normalized Expression\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
Plot the parent protein expression in tissue relative a sampled collection of non\sphinxhyphen{}presented data using seaborn library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{expression\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} The protein expression table which contains the expresion value for each parent protein

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ref\_expression}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} The reference expression of the tissue under investigation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tissue\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the tissue .

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sampling\_num}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The number of times to sample from the non\sphinxhyphen{}prsenter. , defaults to 10

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.violinplot function., defaults to \{‘orient’:’v’\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{def\_value}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default value for proteins that could not be mapped to the expression database , defaults to \sphinxhyphen{}1

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis. , defaults to ‘Normalized Expression’

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the reference gene expression table is smaller than the number of parents

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_peptide\_length\_dist() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_peptide\_length\_dist()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_peptide_length_dist}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_peptide\_length\_dist}}}{\emph{\DUrole{n}{pep\_length}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}int\DUrole{p}{{]}}}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}, \emph{\DUrole{n}{x\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Peptide Length\textquotesingle{}}}, \emph{\DUrole{n}{y\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Frequency\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Peptide Length distribution\textquotesingle{}}}}{}
Visualize a histogram of the eluted peptide length using seaborn library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} {[}description{]}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the function seaborn::distplot, defaults to \{\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{x\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the x\sphinxhyphen{}axis , defaults to ‘Peptide Length’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{y\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the y\sphinxhyphen{}axis , defaults to ‘Frequency’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure, defaults to ‘Peptide Length distribution’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_peptide\_length\_per\_experiment() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_peptide\_length\_per\_experiment()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_peptide_length_per_experiment}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_peptide\_length\_per\_experiment}}}{\emph{\DUrole{n}{counts\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{plotting\_kwargs}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }str\DUrole{p}{{]}}} \DUrole{o}{=} \DUrole{default_value}{\{\}}}}{{ $\rightarrow$ matplotlib.pyplot.figure}}
visualize the peptide length distribution among the experiments defined in the set
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{counts\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} a pandas dataframe that contain the length of each peptide defined in the experiment along with     the experiment name

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_kwargs}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{,}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict object containing parameters for the sns.catplot function, defaults to \{\}

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_protein\_coverage() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_protein\_coverage()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_protein_coverage}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_protein\_coverage}}}{\emph{\DUrole{n}{mapped\_protein}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}, \emph{\DUrole{n}{col}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}r\textquotesingle{}}}, \emph{\DUrole{n}{prot\_name}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{None}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the peptide coverage for a given protein.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{mapped\_protein}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} a NumPy array with shape of 1 by protein length or shape protein\sphinxhyphen{}length

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{col}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the color of the coverage respresentation, defaults to ‘r’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{prot\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default protein name, defaults to None

\end{itemize}

\item[{Return type}] \leavevmode
plt.Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plot\_protein\_presentation\_3D() (in module IPTK.Visualization.vizTools)@\spxentry{plot\_protein\_presentation\_3D()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plot_protein_presentation_3D}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plot\_protein\_presentation\_3D}}}{\emph{\DUrole{n}{proteins}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}\DUrole{p}{{]}}}}, \emph{\DUrole{n}{plotting\_args}\DUrole{o}{=}\DUrole{default_value}{\{\textquotesingle{}color\textquotesingle{}: \textquotesingle{}red\textquotesingle{}\}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{None}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot a 3D surface representation for the SAME protein or proteins OF EQAUL LENGTH     in different conditions.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{proteins}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a nested dict object containing for each protein a child dict that contain     the mapping array and the color in the figure.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{plotting\_args}} (\sphinxstyleliteralemphasis{\sphinxupquote{dict}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} a dict that contain further parameter to the plot\_surface functions, defaults to \{‘color’:’red’\}

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure, defaults to None

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the provided proteins are of different length

\item[{Return type}] \leavevmode
plt.Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_gene\_expression\_vs\_num\_peptides() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_gene\_expression\_vs\_num\_peptides()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_gene_expression_vs_num_peptides}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_gene\_expression\_vs\_num\_peptides}}}{\emph{\DUrole{n}{exp\_count\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{tissue\_name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{def\_value}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Expression value\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Peptides per protein Vs. Protein Expression Level\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
Plot the correlation between the gene expression and the number of peptids per protein using plotly library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{exp\_count\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the number of peptides and the expresion value for each protein in the database

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tissue\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the tissue

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{def\_value}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default value for proteins that could not be mapped to the expression database , defaults to \sphinxhyphen{}1

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the y\sphinxhyphen{}axis., defaults to ‘Expression value’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Peptides per protein Vs. Protein Expression Level’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_multi\_traced\_coverage\_representation() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_multi\_traced\_coverage\_representation()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_multi_traced_coverage_representation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_multi\_traced\_coverage\_representation}}}{\emph{\DUrole{n}{proteins}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}\DUrole{p}{{]}}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Protein Coverage Across  \textquotesingle{}}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
plot a multi\sphinxhyphen{}traced representation for the same protein accross  using plotly library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{proteins}} (\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{type}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} A dict object containing for each protein the corresponding mapped array.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure, defaults to “Protein Coverage Across  “

\end{itemize}

\item[{Returns}] \leavevmode
a multitraced traced figure showing the coverage of proteins accross different conditions

\item[{Return type}] \leavevmode
Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_num\_peptide\_per\_go\_term() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_num\_peptide\_per\_go\_term()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_num_peptide_per_go_term}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_num\_peptide\_per\_go\_term}}}{\emph{\DUrole{n}{pep2goTerm}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}GO\sphinxhyphen{}Term\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per GO Term\textquotesingle{}}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
plot the number of peptides obtained per Go\sphinxhyphen{}Term  using plotly library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep2goTerm}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of peptides from each GO\sphinxhyphen{}Term

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop peptide with unknown GO\sphinxhyphen{}term, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis, defaults to ‘GO\sphinxhyphen{}Term’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure, defaults to ‘Number of peptides per GO Term’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_num\_peptides\_per\_location() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_num\_peptides\_per\_location()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_num_peptides_per_location}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_num\_peptides\_per\_location}}}{\emph{\DUrole{n}{pep2loc}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Compartment\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per sub\sphinxhyphen{}cellular compartment\textquotesingle{}}}}{{ $\rightarrow$ matplotlib.figure.Figure}}
plot the number of peptides obtained from each compartment using plotly library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep2loc}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of peptides from each  location

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop protein with unknown location, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the x\sphinxhyphen{}axis, defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the y\sphinxhyphen{}axis, defaults to ‘Compartment’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The title of the figure, defaults to ‘Number of peptides per sub\sphinxhyphen{}cellular compartment’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_num\_peptides\_per\_organism() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_num\_peptides\_per\_organism()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_num_peptides_per_organism}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_num\_peptides\_per\_organism}}}{\emph{\DUrole{n}{pep\_per\_org}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{log\_scale}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of Peptides\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Organism\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per organism\textquotesingle{}}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
plot the number of peptides per each organism inferred from the experiment using plotly library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep\_per\_org}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of peptides from each organism

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{log\_scale}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} Whether or not to scale the number of peptide using a log scale, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the x\sphinxhyphen{}axis, defaults to ‘Number of Peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The label on the y\sphinxhyphen{}axis, defaults to ‘Organism’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure , defaults to ‘Number of peptides per organism’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_num\_peptides\_per\_parent() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_num\_peptides\_per\_parent()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_num_peptides_per_parent}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_num\_peptides\_per\_parent}}}{\emph{\DUrole{n}{nums\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{num\_prot}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{x\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides\textquotesingle{}}}, \emph{\DUrole{n}{y\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Protein ID\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of peptides per protein\textquotesingle{}}}}{}
Visualize a histogram of the the number of peptides per each inferred protein.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{nums\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} a pandas dataframe containing number of peptides identified from each protein.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{num\_prot}} \textendash{} the number of protein to show relative to the first element, for example, the first 10, 20 etc.     If the default value of \sphinxhyphen{}1 is used then all protein will be plotted, however, this might lead to a crowded figure., defaults to \sphinxhyphen{}1    :type num\_prot: int, optional

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{x\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the x\sphinxhyphen{}axis , defaults to ‘Number of peptides’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{y\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the y\sphinxhyphen{}axis , defaults to ‘Protein ID’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} title, defaults to ‘Number of peptides per protein’

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} if the num\_prot is bigger than the number of elements in the provided table

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_num\_protein\_per\_go\_term() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_num\_protein\_per\_go\_term()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_num_protein_per_go_term}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_num\_protein\_per\_go\_term}}}{\emph{\DUrole{n}{protein2goTerm}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of Proteins\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}GO\sphinxhyphen{}Term\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of proteins per GO\sphinxhyphen{}Term\textquotesingle{}}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
plot the number of proteins per each GO Term using plotly library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein2goTerm}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of proteins from each GO\sphinxhyphen{}Term

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop protein with unknown location, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of Proteins’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis, defaults to ‘GO\sphinxhyphen{}Term’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the title of the figure, defaults to ‘Number of proteins per GO\sphinxhyphen{}Term’

\end{itemize}

\item[{Returns}] \leavevmode
{[}description{]}

\item[{Return type}] \leavevmode
Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_num\_protein\_per\_location() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_num\_protein\_per\_location()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_num_protein_per_location}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_num\_protein\_per\_location}}}{\emph{\DUrole{n}{protein\_loc}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{drop\_unknown}\DUrole{p}{:} \DUrole{n}{bool} \DUrole{o}{=} \DUrole{default_value}{False}}, \emph{\DUrole{n}{xlabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of Proteins\textquotesingle{}}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Compartment\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Number of proteins per sub\sphinxhyphen{}cellular compartment\textquotesingle{}}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
plot the number of proteins per each sub\sphinxhyphen{}cellular compartment
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_loc}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A table that contain the count of protein from each location

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{drop\_unknown}} (\sphinxstyleliteralemphasis{\sphinxupquote{bool}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} whether or not to drop protein with unknown location, defaults to False

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{xlabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the x\sphinxhyphen{}axis, defaults to ‘Number of Proteins’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label on the y\sphinxhyphen{}axis, defaults to ‘Compartment’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} {[}description{]}, defaults to ‘Number of proteins per sub\sphinxhyphen{}cellular compartment’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_overlap\_heatmap() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_overlap\_heatmap()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_overlap_heatmap}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_overlap\_heatmap}}}{\emph{\DUrole{n}{results\_df}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
Plot a user provided dataframe as a heatmap using plotly library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{results\_df}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} A pandas dataframe table that hold the overlapping number.

\item[{Returns}] \leavevmode
a plotly Figure containing the heatmap

\item[{Return type}] \leavevmode
Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_paired\_representation() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_paired\_representation()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_paired_representation}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_paired\_representation}}}{\emph{\DUrole{n}{protein\_one\_repr}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}}}, \emph{\DUrole{n}{protein\_two\_repr}\DUrole{p}{:} \DUrole{n}{Dict\DUrole{p}{{[}}str\DUrole{p}{, }numpy.ndarray\DUrole{p}{{]}}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{} Parired protein representation\textquotesingle{}}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
compare the peptide coverage for the same protein under different conditions using the same protein using plotly library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_one\_repr}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a dict object containing the legand of the first condition along with its mapped array

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{protein\_two\_repr}} (\sphinxstyleliteralemphasis{\sphinxupquote{Dict}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a dict object containing the legand of the second condition along with its mapped array

\end{itemize}

\item[{Returns}] \leavevmode
A plotly Figure containing the representation

\item[{Return type}] \leavevmode
Figure

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_parent\_protein\_expression\_in\_tissue() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_parent\_protein\_expression\_in\_tissue()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_parent_protein_expression_in_tissue}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_parent\_protein\_expression\_in\_tissue}}}{\emph{\DUrole{n}{expression\_table}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{ref\_expression}\DUrole{p}{:} \DUrole{n}{pandas.core.frame.DataFrame}}, \emph{\DUrole{n}{tissue\_name}\DUrole{p}{:} \DUrole{n}{str}}, \emph{\DUrole{n}{sampling\_num}\DUrole{p}{:} \DUrole{n}{int} \DUrole{o}{=} \DUrole{default_value}{10}}, \emph{\DUrole{n}{def\_value}\DUrole{p}{:} \DUrole{n}{float} \DUrole{o}{=} \DUrole{default_value}{\sphinxhyphen{} 1}}, \emph{\DUrole{n}{ylabel}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Normalized Expression\textquotesingle{}}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
plot the parent protein expression in tissue relative a sampled collection of non\sphinxhyphen{}presented genes using plotly library.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{expression\_table}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} The protein expression table which contains the expresion value for each parent proteins.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ref\_expression}} (\sphinxstyleliteralemphasis{\sphinxupquote{pd.DataFrame}}) \textendash{} The reference expression of the tissue under investigation.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{tissue\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}) \textendash{} The name of the tissue.

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{sampling\_num}} (\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the number of times to sample from the non\sphinxhyphen{}prsenter, defaults to 10

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{def\_value}} (\sphinxstyleliteralemphasis{\sphinxupquote{float}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default value for proteins that could not be mapped to the expression database, defaults to \sphinxhyphen{}1

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{ylabel}} (\sphinxstyleliteralemphasis{\sphinxupquote{The label on the y\sphinxhyphen{}axis}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} {[}description{]}, defaults to ‘Normalized Expression’

\end{itemize}

\item[{Raises}] \leavevmode
\sphinxstyleliteralstrong{\sphinxupquote{ValueError}} \textendash{} If the reference gene expression table is smaller than the number of parents

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_peptide\_length\_dist() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_peptide\_length\_dist()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_peptide_length_dist}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_peptide\_length\_dist}}}{\emph{\DUrole{n}{pep\_length}\DUrole{p}{:} \DUrole{n}{List\DUrole{p}{{[}}int\DUrole{p}{{]}}}}, \emph{\DUrole{n}{x\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Peptide Length\textquotesingle{}}}, \emph{\DUrole{n}{y\_label}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Counts\textquotesingle{}}}, \emph{\DUrole{n}{title}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{\textquotesingle{}Peptide Length distribution\textquotesingle{}}}}{}
visualize a histogram of the eluted peptide length using plotly library
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{pep\_length}} (\sphinxstyleliteralemphasis{\sphinxupquote{List}}\sphinxstyleliteralemphasis{\sphinxupquote{{[}}}\sphinxstyleliteralemphasis{\sphinxupquote{int}}\sphinxstyleliteralemphasis{\sphinxupquote{{]}}}) \textendash{} a list of integer containing the peptides’ lengths

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{x\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the x\sphinxhyphen{}axis , defaults to ‘Peptide Length’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{y\_label}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the label of the y\sphinxhyphen{}axis, defaults to ‘Counts’

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{title}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} the figure title, defaults to ‘Peptide Length distribution’

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{plotly\_protein\_coverage() (in module IPTK.Visualization.vizTools)@\spxentry{plotly\_protein\_coverage()}\spxextra{in module IPTK.Visualization.vizTools}}

\begin{fulllineitems}
\phantomsection\label{\detokenize{IPTK.Visualization:IPTK.Visualization.vizTools.plotly_protein_coverage}}\pysiglinewithargsret{\sphinxcode{\sphinxupquote{IPTK.Visualization.vizTools.}}\sphinxbfcode{\sphinxupquote{plotly\_protein\_coverage}}}{\emph{\DUrole{n}{mapped\_protein}\DUrole{p}{:} \DUrole{n}{numpy.ndarray}}, \emph{\DUrole{n}{prot\_name}\DUrole{p}{:} \DUrole{n}{str} \DUrole{o}{=} \DUrole{default_value}{None}}}{{ $\rightarrow$ plotly.graph\_objs.\_figure.Figure}}
Plot the peptide coverage for a given protein
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{mapped\_protein}} (\sphinxstyleliteralemphasis{\sphinxupquote{np.ndarray}}) \textendash{} A NumPy array with shape of 1 by protein length or shape protein\sphinxhyphen{}length

\item {} 
\sphinxstyleliteralstrong{\sphinxupquote{prot\_name}} (\sphinxstyleliteralemphasis{\sphinxupquote{str}}\sphinxstyleliteralemphasis{\sphinxupquote{, }}\sphinxstyleliteralemphasis{\sphinxupquote{optional}}) \textendash{} The default protein name, defaults to None

\end{itemize}

\item[{Return type}] \leavevmode
Figure

\end{description}\end{quote}

\end{fulllineitems}



\subparagraph{Module contents}
\label{\detokenize{IPTK.Visualization:module-IPTK.Visualization}}\label{\detokenize{IPTK.Visualization:module-contents}}\index{module@\spxentry{module}!IPTK.Visualization@\spxentry{IPTK.Visualization}}\index{IPTK.Visualization@\spxentry{IPTK.Visualization}!module@\spxentry{module}}

\paragraph{Module contents}
\label{\detokenize{IPTK:module-IPTK}}\label{\detokenize{IPTK:module-contents}}\index{module@\spxentry{module}!IPTK@\spxentry{IPTK}}\index{IPTK@\spxentry{IPTK}!module@\spxentry{module}}

\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{sphinxtheindex}
\let\bigletter\sphinxstyleindexlettergroup
\bigletter{i}
\item\relax\sphinxstyleindexentry{IPTK}\sphinxstyleindexpageref{IPTK:\detokenize{module-IPTK}}
\item\relax\sphinxstyleindexentry{IPTK.Analysis}\sphinxstyleindexpageref{IPTK.Analysis:\detokenize{module-IPTK.Analysis}}
\item\relax\sphinxstyleindexentry{IPTK.Analysis.AnalysisFunction}\sphinxstyleindexpageref{IPTK.Analysis:\detokenize{module-IPTK.Analysis.AnalysisFunction}}
\item\relax\sphinxstyleindexentry{IPTK.Classes}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.Database}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.Database}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.Experiment}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.Experiment}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.ExperimentalSet}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.ExperimentalSet}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.HLAChain}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.HLAChain}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.HLAMolecules}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.HLAMolecules}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.HLASet}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.HLASet}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.Peptide}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.Peptide}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.Proband}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.Proband}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.Protein}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.Protein}}
\item\relax\sphinxstyleindexentry{IPTK.Classes.Tissue}\sphinxstyleindexpageref{IPTK.Classes:\detokenize{module-IPTK.Classes.Tissue}}
\item\relax\sphinxstyleindexentry{IPTK.IO}\sphinxstyleindexpageref{IPTK.IO:\detokenize{module-IPTK.IO}}
\item\relax\sphinxstyleindexentry{IPTK.IO.InFunctions}\sphinxstyleindexpageref{IPTK.IO:\detokenize{module-IPTK.IO.InFunctions}}
\item\relax\sphinxstyleindexentry{IPTK.IO.MEMEInterface}\sphinxstyleindexpageref{IPTK.IO:\detokenize{module-IPTK.IO.MEMEInterface}}
\item\relax\sphinxstyleindexentry{IPTK.IO.OutFunctions}\sphinxstyleindexpageref{IPTK.IO:\detokenize{module-IPTK.IO.OutFunctions}}
\item\relax\sphinxstyleindexentry{IPTK.Utils}\sphinxstyleindexpageref{IPTK.Utils:\detokenize{module-IPTK.Utils}}
\item\relax\sphinxstyleindexentry{IPTK.Utils.DevFunctions}\sphinxstyleindexpageref{IPTK.Utils:\detokenize{module-IPTK.Utils.DevFunctions}}
\item\relax\sphinxstyleindexentry{IPTK.Utils.Mapping}\sphinxstyleindexpageref{IPTK.Utils:\detokenize{module-IPTK.Utils.Mapping}}
\item\relax\sphinxstyleindexentry{IPTK.Utils.Types}\sphinxstyleindexpageref{IPTK.Utils:\detokenize{module-IPTK.Utils.Types}}
\item\relax\sphinxstyleindexentry{IPTK.Utils.UtilityFunction}\sphinxstyleindexpageref{IPTK.Utils:\detokenize{module-IPTK.Utils.UtilityFunction}}
\item\relax\sphinxstyleindexentry{IPTK.Visualization}\sphinxstyleindexpageref{IPTK.Visualization:\detokenize{module-IPTK.Visualization}}
\item\relax\sphinxstyleindexentry{IPTK.Visualization.vizTools}\sphinxstyleindexpageref{IPTK.Visualization:\detokenize{module-IPTK.Visualization.vizTools}}
\end{sphinxtheindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}