
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>IPTK 0.3.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="welcome-to-iptk-s-documentation">
<h1>Welcome to IPTK’s documentation!<a class="headerlink" href="#welcome-to-iptk-s-documentation" title="Permalink to this headline">¶</a></h1>
<img alt="_images/IKMB_LOGO.png" src="_images/IKMB_LOGO.png" />
<p>Analysis, Visualize, Compare and Integrate experimentally generated or in-silico predicted  Immunopeptidomics data,  !</p>
</div>
<div class="section" id="introduction">
<h1>Introduction:<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>IPTK is a Pythonic library specialized in the analysis of HLA-peptidomes identified through an Immunopeptiomics pipeline.
The library provides a high level API for analyzing and visualizing the identified peptides,
Integrating transcritomics and protein structure information for a rich analysis and for comparing different experiments and different runs.</p>
</div>
<div class="section" id="installation">
<h1>Installation:<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>The library can installed using pip as follow</p>
</div>
<div class="section" id="funding">
<h1>Funding:<a class="headerlink" href="#funding" title="Permalink to this headline">¶</a></h1>
<p>The project was funded by the German Research Foundation (DFG) (Research Training Group 1743, ‘Genes, Environment and Inflammation’)</p>
<img alt="_images/RTG1743.png" src="_images/RTG1743.png" />
<div class="section" id="guide">
<h2>Guide<a class="headerlink" href="#guide" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<span id="document-license"></span><div class="section" id="license">
<h3>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h3>
<p>Will be written here</p>
</div>
<div class="section" id="contact">
<h3>Contact<a class="headerlink" href="#contact" title="Permalink to this headline">¶</a></h3>
<p>for further question and communication please contact <a class="reference external" href="mailto:h&#46;elabd&#37;&#52;&#48;ikmb&#46;uni-kiel&#46;de">h<span>&#46;</span>elabd<span>&#64;</span>ikmb<span>&#46;</span>uni-kiel<span>&#46;</span>de</a></p>
</div>
<span id="document-get_started"></span><div class="section" id="get-started">
<h3>Get Started!<a class="headerlink" href="#get-started" title="Permalink to this headline">¶</a></h3>
<p>To get started with using the library check the Interactive Tutorials available at <a class="reference external" href="https://github.com/ikmb/iptoolkit/tree/master/Tutorials">https://github.com/ikmb/iptoolkit/tree/master/Tutorials</a></p>
</div>
<span id="document-modules"></span><div class="section" id="iptk">
<h3>IPTK<a class="headerlink" href="#iptk" title="Permalink to this headline">¶</a></h3>
<div class="toctree-wrapper compound">
<span id="document-IPTK"></span><div class="section" id="iptk-package">
<h4>IPTK package<a class="headerlink" href="#iptk-package" title="Permalink to this headline">¶</a></h4>
<div class="section" id="subpackages">
<h5>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h5>
<div class="toctree-wrapper compound">
<span id="document-IPTK.Analysis"></span><div class="section" id="iptk-analysis-package">
<h6>IPTK.Analysis package<a class="headerlink" href="#iptk-analysis-package" title="Permalink to this headline">¶</a></h6>
<div class="section" id="submodules">
<h7>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h7>
</div>
<div class="section" id="module-IPTK.Analysis.AnalysisFunction">
<span id="iptk-analysis-analysisfunction-module"></span><h7>IPTK.Analysis.AnalysisFunction module<a class="headerlink" href="#module-IPTK.Analysis.AnalysisFunction" title="Permalink to this headline">¶</a></h7>
<p>The module contain a collection of analysis function that can be used by the methods of
the classes defined in the classes module.</p>
<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.compute_binary_distance">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">compute_binary_distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptides</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">dist_func</span><span class="p">:</span> <span class="n">Callable</span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.compute_binary_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>compare the distance between every pair of peptides in a collection of peptides. 
&#64;param: peptides: a collection of peptides sequences.
&#64;param: dist_func: function to compute the distance between each pair of peptides. 
&#64;note:</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – a collection of peptides sequences.</p></li>
<li><p><strong>dist_func</strong> (<em>Callable</em>) – a function to compute the distance between each pair of peptides.</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>RuntimeError</strong> – make sure that the dist_function is suitable with the peptides which might have different lengths.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the distance between each pair of peptides in the provided list of peptides</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.compute_change_in_protein_representation">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">compute_change_in_protein_representation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapped_prot_cond1</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">mapped_prot_cond2</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.compute_change_in_protein_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the change in the protein representation between two conditions, by computing 
the difference in the area under the curve, AUC.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mapped_prot_cond1</strong> (<em>np.ndarray</em>) – a mapped protein instance containing the protein coverage in the first condition</p></li>
<li><p><strong>mapped_prot_cond2</strong> (<em>np.ndarray</em>) – a mapped protein instance containing the protein coverage in the second condition</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if the provided pair of proteins is of different length</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the difference in the area under the coverage curve between the two experiments.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.compute_difference_in_representation">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">compute_difference_in_representation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapped_prot_cond1</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">mapped_prot_cond2</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.compute_difference_in_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>return the difference in the representation of a protein between two conditions
by substracting the coverage of the first protein from the second proteins.</p>
<p>&#64;param: mapped_prot_cond1: a mapped protein instance containing the protein coverage in the first condition
&#64;param: mapped_prot_cond2: a mapped protein instance containing the protein coverage in the second condition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mapped_prot_cond1</strong> (<em>np.ndarray</em>) – a mapped protein instance containing the protein coverage in the first condition</p></li>
<li><p><strong>mapped_prot_cond2</strong> (<em>np.ndarray</em>) – a mapped protein instance containing the protein coverage in the second condition</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>an array that shows the difference in coverage between the two proteins at each amino acid position.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.compute_expression_correlation">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">compute_expression_correlation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exp1</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a></span></em>, <em class="sig-param"><span class="n">exp2</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a></span></em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.compute_expression_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>compute the correlation in the gene expression between two experiments by constructing a union
of all the proteins expressed in the first and second experiments, extract the gene expression 
of these genes and then compute the correlation using SciPy stat module.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp1</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment"><em>Experiment</em></a>) – The first experimental object</p></li>
<li><p><strong>exp2</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment"><em>Experiment</em></a>) – he second experimental object</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the correlation in gene expression of the proteins inferred in the provided pair of experiment</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.download_structure_file">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">download_structure_file</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pdb_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.download_structure_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Download PDB/mmCIF file containing the pbd_id from PDB using BioPython library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pdb_id</strong> (<em>str</em>) – the protein id in protein databank</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.get_binnary_peptide_overlap">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">get_binnary_peptide_overlap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exp1</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a></span></em>, <em class="sig-param"><span class="n">exp2</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a></span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.get_binnary_peptide_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>compare the peptide overlap between two experimental objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp1</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment"><em>Experiment</em></a>) – an instance of class Experiment.</p></li>
<li><p><strong>exp2</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment"><em>Experiment</em></a>) – an instance of class Experiment.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of peptides that have been identified in both experiments.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.get_binnary_protein_overlap">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">get_binnary_protein_overlap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exp1</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a></span></em>, <em class="sig-param"><span class="n">exp2</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a></span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.get_binnary_protein_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>compare the protein overlap between two experimental objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp1</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment"><em>Experiment</em></a>) – an instance of class Experiment.</p></li>
<li><p><strong>exp2</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment"><em>Experiment</em></a>) – an instance of class Experiment.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of proteins that have been identified in both experiments.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Proteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Analysis.AnalysisFunction.get_sequence_motif">
<code class="sig-prename descclassname">IPTK.Analysis.AnalysisFunction.</code><code class="sig-name descname">get_sequence_motif</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptides</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">temp_dir</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'./TEMP_DIR'</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">meme_params</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Analysis.AnalysisFunction.get_sequence_motif" title="Permalink to this definition">¶</a></dt>
<dd><p>compute the sequences motif from a collection of peptide sequences using meme software.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>Peptides</em>) – a list of string containing the peptide sequences</p></li>
<li><p><strong>temp_dir</strong> (<em>str</em><em>, </em><em>optional</em>) – he temp directory to write temp-files to it, defaults to “./TEMP_DIR”</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to print the output of the motif discovery tool to the stdout, defaults to False</p></li>
<li><p><strong>meme_params</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object that contain meme controlling parameters, defaults to {}</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – incase meme is not installed or could not be found in the path!</p></li>
<li><p><strong>ValueError</strong> – incase the peptides have different length!</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.Analysis">
<span id="module-contents"></span><h7>Module contents<a class="headerlink" href="#module-IPTK.Analysis" title="Permalink to this headline">¶</a></h7>
</div>
</div>
<span id="document-IPTK.Classes"></span><div class="section" id="iptk-classes-package">
<h6>IPTK.Classes package<a class="headerlink" href="#iptk-classes-package" title="Permalink to this headline">¶</a></h6>
<div class="section" id="submodules">
<h7>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h7>
</div>
<div class="section" id="module-IPTK.Classes.Database">
<span id="iptk-classes-database-module"></span><h7>IPTK.Classes.Database module<a class="headerlink" href="#module-IPTK.Classes.Database" title="Permalink to this headline">¶</a></h7>
<p>This submodule define a collection of container classes that are used through the library</p>
<dl class="py class">
<dt id="IPTK.Classes.Database.CellularLocationDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">CellularLocationDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2data</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The class provides an API to access the cellular location information from a database   the follow the structure of the human Proteome Atlas sub-cellular location database. See <a class="reference external" href="https://www.proteinatlas.org/about/download">https://www.proteinatlas.org/about/download</a>    for more details.</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.add_to_database">
<code class="sig-name descname">add_to_database</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genes_to_add</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a></span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.add_to_database" title="Permalink to this definition">¶</a></dt>
<dd><p>add the the location of more proteins to the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>genes_to_add</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB"><em>CellularLocationDB</em></a>) – a CellularLocationDB instance containing the genes that shall be added to the database.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if the genes to add to the database are already defined in the database</p></li>
<li><p><strong>RuntimeError</strong> – Incase any other error has been encountered while merging the tables.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_approved_location">
<code class="sig-name descname">get_approved_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_approved_location" title="Permalink to this definition">¶</a></dt>
<dd><p>return the location of the provided gene id or gene name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the approved location where the protein the corresponds to the provided name or id is located.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_gene_names">
<code class="sig-name descname">get_gene_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_gene_names" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of all gene names in the dataset</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the names of all genes in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_genes">
<code class="sig-name descname">get_genes</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of all gene ids in the dataset</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>all genes ids currently defined in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_go_names">
<code class="sig-name descname">get_go_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_go_names" title="Permalink to this definition">¶</a></dt>
<dd><p>return the location of the provided gene id or gene name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the gene ontology, GO,  location where the protein the corresponds to the provided name or id is located.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_main_location">
<code class="sig-name descname">get_main_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">corresponds</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_main_location" title="Permalink to this definition">¶</a></dt>
<dd><p>return the main location(s) of the provided gene id or gene name. 
If both gene Id and gene name are provided, both gene_id has a higher precedence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the main location where the protein the corresponds to the provided name or id is located.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_table">
<code class="sig-name descname">get_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_table" title="Permalink to this definition">¶</a></dt>
<dd><p>return the instance table</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the location table of the instance.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Database.GeneExpressionDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">GeneExpressionDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2data</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'https://www.proteinatlas.org/download/rna_tissue_consensus.tsv.zip'</span></em>, <em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>provides an API to access gene expression data stored in table that follows the same structure as 
the Human proteome Atlas Normalized RNA Expression see  <a class="reference external" href="https://www.proteinatlas.org/about/download">https://www.proteinatlas.org/about/download</a> for more details</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_expression">
<code class="sig-name descname">get_expression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a table summarizing the expression of the provided gene name or gene id accross different tissues.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table summarizing the expression of the provided gene accross all tissues in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_expression_in_tissue">
<code class="sig-name descname">get_expression_in_tissue</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_expression_in_tissue" title="Permalink to this definition">¶</a></dt>
<dd><p>return the expression profile of the provided tissue</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tissue_name</strong> (<em>str</em>) – the name of the tissue</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>KeyError</strong> – incase the provided tissue is not provided in the database</p></li>
<li><p><strong>RuntimeError</strong> – in case any error was encountered while generating the expression profile.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a table summarizing the expression of all genes in the provided tissue.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_gene_names">
<code class="sig-name descname">get_gene_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_gene_names" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of the UNIQUE gene names currently in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of the UNIQUE gene names currently in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_genes">
<code class="sig-name descname">get_genes</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of the UNIQUE gene ids currently in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of the UNIQUE gene ids currently in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_table">
<code class="sig-name descname">get_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_table" title="Permalink to this definition">¶</a></dt>
<dd><p>return a table containing the expression value of all the genes accross all tissues in the current instance</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The expression of all genes accross all tissues in the database.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_tissues">
<code class="sig-name descname">get_tissues</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_tissues" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of the tissues in the current database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list containing the names of the UNIQUE tissues in the database.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Database.OrganismDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">OrganismDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2Fasta</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.OrganismDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Extract information about the source organsim of a collection of protein sequences 
from a fasta file and provides an API to query the results.     The function expect the input fasta file to have header written in the UNIPROT format.</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.OrganismDB.get_number_protein_per_organism">
<code class="sig-name descname">get_number_protein_per_organism</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.OrganismDB.get_number_protein_per_organism" title="Permalink to this definition">¶</a></dt>
<dd><p>provides a table containing the number of proteins per organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a table containing the number of proteins per organism</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.OrganismDB.get_org">
<code class="sig-name descname">get_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Database.OrganismDB.get_org" title="Permalink to this definition">¶</a></dt>
<dd><p>return the parent organism of the provided proteins</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prot_id</strong> (<em>str</em>) – the id of the protein of interest</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – incase the provided identifier is not in the database</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the parent organism, i.e. the source organism.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.OrganismDB.get_unique_orgs">
<code class="sig-name descname">get_unique_orgs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.OrganismDB.get_unique_orgs" title="Permalink to this definition">¶</a></dt>
<dd><p>get the number of unique organisms in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of all unique organisms in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Database.SeqDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">SeqDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2fasta</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.SeqDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>load a fasta file and constructs a lock up dictionary where sequence ids are  keys and sequences are values.</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.SeqDB.get_seq">
<code class="sig-name descname">get_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Database.SeqDB.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the corresponding sequence if the provided protein-id is defined in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_id</strong> (<em>str</em>) – The protein id to retrive its sequence.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – If the provided protein does not exist in the database</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the protein sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.SeqDB.has_sequence">
<code class="sig-name descname">has_sequence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Database.SeqDB.has_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>check if the provided sequence id is an element of the database or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence_name</strong> (<em>str</em>) – The id of the sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the database has this id, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Experiment">
<span id="iptk-classes-experiment-module"></span><h7>IPTK.Classes.Experiment module<a class="headerlink" href="#module-IPTK.Classes.Experiment" title="Permalink to this headline">¶</a></h7>
<p>This module provides an abstraction for an IP experiment.</p>
<dl class="py class">
<dt id="IPTK.Classes.Experiment.Experiment">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Experiment.</code><code class="sig-name descname">Experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proband</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Proband.Proband" title="IPTK.Classes.Proband.Proband">IPTK.Classes.Proband.Proband</a></span></em>, <em class="sig-param"><span class="n">hla_set</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.HLASet.HLASet" title="IPTK.Classes.HLASet.HLASet">IPTK.Classes.HLASet.HLASet</a></span></em>, <em class="sig-param"><span class="n">tissue</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Tissue.Tissue" title="IPTK.Classes.Tissue.Tissue">IPTK.Classes.Tissue.Tissue</a></span></em>, <em class="sig-param"><span class="n">database</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Database.SeqDB" title="IPTK.Classes.Database.SeqDB">IPTK.Classes.Database.SeqDB</a></span></em>, <em class="sig-param"><span class="n">ident_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A representation of an immunopeptidomic experiment.</p>
<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.add_org_info">
<code class="sig-name descname">add_org_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot2org</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.add_org_info" title="Permalink to this definition">¶</a></dt>
<dd><p>annotated the inferred proteins with their source organism</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prot2org</strong> (<em>ProteinSource</em>) – a dict that contain the protein id as keys and its source organism as values           and add this info to each protein inferred in the current experiment.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>RuntimeWarning</strong> – If the provided dictionary does cover all proteins in the experimental object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.annotate_proteins">
<code class="sig-name descname">annotate_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">organisms_db</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Database.OrganismDB" title="IPTK.Classes.Database.OrganismDB">IPTK.Classes.Database.OrganismDB</a></span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.annotate_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the parent organisms of each protein in the experiment from an organism database instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>organisms_db</strong> (<a class="reference internal" href="index.html#IPTK.Classes.Database.OrganismDB" title="IPTK.Classes.Database.OrganismDB"><em>OrganismDB</em></a>) – an OrgansimDB instance that will be used to annotate the proteins          identified in the experiment.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.drop_peptide_belong_to_org">
<code class="sig-name descname">drop_peptide_belong_to_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.drop_peptide_belong_to_org" title="Permalink to this definition">¶</a></dt>
<dd><p>Drop the all the peptides that belong to a user provided organism.              Note that, this function will IRREVERSIBLY remove the peptide from the experimental object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>org</strong> (<em>str</em>) – the organims name</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_binarized_results">
<code class="sig-name descname">get_binarized_results</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_binarized_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of NumPy arrays where each array represents a child peptide, parent protein mapped pair.
Please note that, The function treat each peptide-protein pair individually, that is if two peptides originating from the same protein, 
it treat them independently and the same protein will be represented twice with the two different peptides. Incase an integrative mapping is needed,
the function &#64;get_integrated_binarized_results&#64; shall be used.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of NumPy arrays containing the mapping between each peptide protein pair.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>MappedProtein</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_c_terminal_flanked_seqs">
<code class="sig-name descname">get_c_terminal_flanked_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_c_terminal_flanked_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>return the c-terminal flanking sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_length</strong> (<em>int</em>) – the length of the peptide downstream of the C-terminal of the peptide</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list sequences contain the N-terminal flanking sequence for each peptide in the instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_experiment_reference_tissue_expression">
<code class="sig-name descname">get_experiment_reference_tissue_expression</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_experiment_reference_tissue_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>return the reference gene expression for the current tissue</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that contain the expression value for ALL the genes in the instance Tissue</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_expression_of_parent_proteins">
<code class="sig-name descname">get_expression_of_parent_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">non_mapped_dval</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">- 1</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_expression_of_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>return a table containing the expression of the proteins inferred in the current experiment from the current tissue.
This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>non_mapped_dval</strong> (<em>float</em><em>, </em><em>optional</em>) – A default value to be added incase the parent protein is not define in the expression database, defaults to -1</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a table that contain the expression of the protein inferred in the database</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_flanked_peptides">
<code class="sig-name descname">get_flanked_peptides</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_flanked_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a list of sequences containing the peptides identified in the experiment padded with
the flanking regions from all the parents of each peptide.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_length</strong> (<em>int</em>) – the length of the flanking region</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of the peptides + the flanking region.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_go_location_id_parent_proteins">
<code class="sig-name descname">get_go_location_id_parent_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">not_mapped_val</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'UNK'</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_go_location_id_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the gene ontology,GO, location terms for all the identified proteins.</p>
<p>&#64;brief: 
&#64;param: not_mapped_val: the default value to return incase the GO term of the protein can not be extracted. 
&#64;note: This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>not_mapped_val</strong> (<em>str</em><em>, </em><em>optional</em>) – The default value to return incase the GO term of the protein can not be extracted, defaults to ‘UNK’</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A table that contain the GO-location term for each protein in the current instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_hla_allele">
<code class="sig-name descname">get_hla_allele</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_hla_allele" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the set of HLA alleles from which the instance peptides have been eluted</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_hla_class">
<code class="sig-name descname">get_hla_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_hla_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the HLA class</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_main_sub_cellular_location_of_parent_proteins">
<code class="sig-name descname">get_main_sub_cellular_location_of_parent_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">not_mapped_val</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'UNK'</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_main_sub_cellular_location_of_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the main cellular location for the identified proteins.
This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>not_mapped_val</strong> (<em>str</em><em>, </em><em>optional</em>) – The default value to return incase the location of a protein can not be extracted, defaults to ‘UNK’</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A table that contain the main cellular compartment for each protein in the current instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_mapped_protein">
<code class="sig-name descname">get_mapped_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_mapped_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>return an NumPy array of shape 1 x protein length where each number in the array represents 
the total number of peptides identified in the experiment that have originated from the said position 
in the protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – the protein id</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided protein id was inferred from the current experiment</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a NumPy array that contain the coverage of the protein.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_mapped_proteins">
<code class="sig-name descname">get_mapped_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>numpy.ndarray<span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_mapped_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>return a dictionary of all the proteins identified in the current experiment with all inferred
peptides mapped to them.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a dictionary that contain the mapped proteins for all the proteins in the current instance.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>MappedProteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_mono_parent_peptides">
<code class="sig-name descname">get_mono_parent_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_mono_parent_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of peptides that have only one parent protein</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of peptide instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_n_terminal_flanked_seqs">
<code class="sig-name descname">get_n_terminal_flanked_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_n_terminal_flanked_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>return the n-terminal flanking sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_length</strong> (<em>int</em>) – the length of the flanking region upstream of the N-terminal of the peptide</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list sequences contain the N-terminal flanking sequence for each peptide in the instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_negative_example">
<code class="sig-name descname">get_negative_example</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fold</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">2</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_negative_example" title="Permalink to this definition">¶</a></dt>
<dd><p>generate negative examples, i.e., non-bounding peptides from the proteins identified in the current experiment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fold</strong> (<em>int</em><em>, </em><em>optional</em>) – the number of negative example to generate relative to the number of unique identified peptides, defaults to 2</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of non-presented peptides from all inferred proteins.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_num_peptide_expression_table">
<code class="sig-name descname">get_num_peptide_expression_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_num_peptide_expression_table" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Get a table that contain the id of all parent proteins, number of peptide per-proteins and the expression value </dt><dd><p>of these parent transcripts. Please note, this method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of peptides per protein table</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_num_peptide_per_go_term">
<code class="sig-name descname">get_num_peptide_per_go_term</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_num_peptide_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the number of peptides per each GO-Term 
:return: A table that has two columns, namely, GO-Terms and Counts. 
:rtype: pd.DataFrame</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_num_peptide_per_location">
<code class="sig-name descname">get_num_peptide_per_location</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_num_peptide_per_location" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the number of peptides obtained from proteins localized to different sub-cellular compartments</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that has two columns, namely, Compartment and Counts.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_number_of_children">
<code class="sig-name descname">get_number_of_children</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_number_of_children" title="Permalink to this definition">¶</a></dt>
<dd><p>return the number of children, i.e. number of peptides belonging to a parent protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – the id of the parent protein</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the number of peptides</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_compartment">
<code class="sig-name descname">get_number_of_proteins_per_compartment</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_compartment" title="Permalink to this definition">¶</a></dt>
<dd><p>get the number of proteins from each compartment</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that has two columns, namely, Compartment and Counts.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_go_term">
<code class="sig-name descname">get_number_of_proteins_per_go_term</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>get the number of proteins from each GO-Term</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that has two columns, namely, GO-Terms and Counts.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_orgs">
<code class="sig-name descname">get_orgs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_orgs" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list containing the UNIQUE organisms identified in the current experiment</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of all UNIQUE organisms inferred from the inferred proteins.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptide">
<code class="sig-name descname">get_peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_seq</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>return a peptide instance corresponding to the user provided peptide sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pep_seq</strong> (<em>str</em>) – the peptide sequence</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the peptide sequence has not been inferred from the current database.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the peptide instance with the coresponding sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">Peptide</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptide_number_parent">
<code class="sig-name descname">get_peptide_number_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptide_number_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>return a pandas dataframe with the peptide sequence in the first columns and the 
number of parent proteins in the second column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ascending</strong> (<em>bool</em><em>, </em><em>optional</em>) – ascending sort the peptide by their number of parent proteins, defaults to False</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the number of parents for each peptide</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides">
<code class="sig-name descname">get_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set of all the peptide stored in the experimental object</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides_length">
<code class="sig-name descname">get_peptides_length</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides_length" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list containing the length of each unique peptide in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>peptides’ lengths</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides_per_organism">
<code class="sig-name descname">get_peptides_per_organism</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides_per_organism" title="Permalink to this definition">¶</a></dt>
<dd><p>return a pandas dataframe that contain the count of peptides belonging to each organism in
the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a table with two columns, namely, Organisms and Counts</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides_per_protein">
<code class="sig-name descname">get_peptides_per_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides_per_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>return a pandas dataframe that contain the number of peptides belonging to each protein 
inferred in the experiment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ascending</strong> (<em>bool</em><em>, </em><em>optional</em>) – ascending sort the proteins by their number of parent number of child peptides, defaults to False</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a table with the following columns, Proteins and Number_of_Peptides</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_poly_parental_peptides">
<code class="sig-name descname">get_poly_parental_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_poly_parental_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of peptides that have more than one parent proteins
:return: [list of peptide instance 
:rtype: Peptides</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_proband_name">
<code class="sig-name descname">get_proband_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_proband_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the proband name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_proteins">
<code class="sig-name descname">get_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Protein.Protein" title="IPTK.Classes.Protein.Protein">IPTK.Classes.Protein.Protein</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_proteins" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set of all the proteins in the experimental object</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Proteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_tissue">
<code class="sig-name descname">get_tissue</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="index.html#IPTK.Classes.Tissue.Tissue" title="IPTK.Classes.Tissue.Tissue">IPTK.Classes.Tissue.Tissue</a><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_tissue" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the tissue of the current experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="index.html#IPTK.Classes.Tissue.Tissue" title="IPTK.Classes.Tissue.Tissue">Tissue</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_tissue_name">
<code class="sig-name descname">get_tissue_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_tissue_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the tissue name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_allele_group">
<code class="sig-name descname">has_allele_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain peptides eluted from an HLA-alleles belonging to the provided allele group or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_group</strong> (<em>str</em>) – the gene group to query the hla_set against</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the gene group has a member that is a member of the instance HLASet and False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_gene">
<code class="sig-name descname">has_gene</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">locus</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain peptides eluted from an HLA-alleles belonging to the provided locus or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>locus</strong> (<em>str</em>) – the locus of the allele to query the hla_set against</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the locus has a member that is a member of the instance HLASet and False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_hla_allele">
<code class="sig-name descname">has_hla_allele</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_hla_allele" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain an eluted peptides from the provided alleles</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> (<em>str</em>) – is the name of the allele as a string</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the allele is a member of the instance HLASet and False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_protein_group">
<code class="sig-name descname">has_protein_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain peptides eluted from an HLA-alleles belonging to the provided protein group or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_group</strong> (<em>str</em>) – The protein group to query the hla_set against</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the locus has a member that is a member of the instance HLASet and False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.is_a_parent_protein">
<code class="sig-name descname">is_a_parent_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.is_a_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein</strong> – check if the protein is a member of the instance proteins or not.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the protein has been identified in the current instance, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.is_member">
<code class="sig-name descname">is_member</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptide</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.is_member" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – check if the peptide is a member of the instance peptides or not.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the peptide has been identified in the current instance, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.ExperimentalSet">
<span id="iptk-classes-experimentalset-module"></span><h7>IPTK.Classes.ExperimentalSet module<a class="headerlink" href="#module-IPTK.Classes.ExperimentalSet" title="Permalink to this headline">¶</a></h7>
<p>An Experimental set which is a collection of experiments.
The class provides an API for integrating and comparing different experimental instances.</p>
<dl class="py class">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.ExperimentalSet.</code><code class="sig-name descname">ExperimentSet</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">exp_id_pair</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>an API for integrating and comparing different experimental instances</p>
<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.add_experiment">
<code class="sig-name descname">add_experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">exp_id_pair</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.add_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>add an arbitrary number of experiments to the set</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compare_org_count_among_exps">
<code class="sig-name descname">compare_org_count_among_exps</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">abs_count</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compare_org_count_among_exps" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>org</strong> (<em>str</em>) – The name of the organism to query the database for it.</p></li>
<li><p><strong>abs_count</strong> (<em>bool</em><em>, </em><em>optional</em>) – The absolute count, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the count of the peptides that belong to a specific organism in the database.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compare_peptide_counts">
<code class="sig-name descname">compare_peptide_counts</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compare_peptide_counts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that contain the total number of peptides and per-organism peptide counts         among all experiments in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_average_distance_between_exps">
<code class="sig-name descname">compute_average_distance_between_exps</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_average_distance_between_exps" title="Permalink to this definition">¶</a></dt>
<dd><p>compute the average distance between experiments by taking the average over the z-axis
of the 3D tensor computed by the function compute_change_in_protein_representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A 2D tensor with shape of (num-experiments, num-experiments)</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_change_in_protein_representation">
<code class="sig-name descname">compute_change_in_protein_representation</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_change_in_protein_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the change in protein representation among the proteins that are presented/ detect in all of the         instance’s experiments.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a 3D tensor, T, with shape of (num-experiments, num-experiments, num-proteins),         where T[i,j,k] is a the difference between experiment i &amp; j with respect to the k th protein         :rtype: np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_correlation_in_experssion">
<code class="sig-name descname">compute_correlation_in_experssion</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_correlation_in_experssion" title="Permalink to this definition">¶</a></dt>
<dd><p>compute the correlation in parent protein gene-expression across all the experiments
in the set. See the function <strong>compute_binary_correlation</strong> in the analysis module 
for information about the computational logic.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a 2D matrix containing the coorelation in gene expression between each pair of experiments inside the current instance collection of experiments.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_length_table">
<code class="sig-name descname">compute_peptide_length_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_length_table" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that contain the length of each peptide in the experiment</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_overlap_matrix">
<code class="sig-name descname">compute_peptide_overlap_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_overlap_matrix" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a 2D matrix containing the number of peptide overlapping between each pair of experiments inside the current instance collection of experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_representation_count">
<code class="sig-name descname">compute_peptide_representation_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_representation_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of times a peptide was observed accross  experiments in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_coverage_over_the_set">
<code class="sig-name descname">compute_protein_coverage_over_the_set</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_coverage_over_the_set" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the mapped representation for each protein in accross the entire  set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_overlap_matrix">
<code class="sig-name descname">compute_protein_overlap_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_overlap_matrix" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a 2D matrix containing the number of proteins overlapping between each pair of experiments inside the current instance collection of experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_representation_count">
<code class="sig-name descname">compute_protein_representation_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_representation_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of times a protein was observed accross the experiment in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.drop_peptides_belong_to_org">
<code class="sig-name descname">drop_peptides_belong_to_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.drop_peptides_belong_to_org" title="Permalink to this definition">¶</a></dt>
<dd><p>drop all the peptides that belong to the provided organisms from all experiments in the set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>org_name</strong> (<em>str</em>) – the name of the organism to drop</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_allele_count">
<code class="sig-name descname">get_allele_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_allele_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of experiments obtained from each allele in the instance.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiment">
<code class="sig-name descname">get_experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exp_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>return the experiment pointed to by the provided experimental name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>exp_name</strong> (<em>str</em>) – the name of the experiment</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided experimental name is not in the dataset.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the experiment corresponds to the provided name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">Experiment</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_experimental_names">
<code class="sig-name descname">get_experimental_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_experimental_names" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list with all the identifiers of the experiments in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Names</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiments">
<code class="sig-name descname">get_experiments</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>Experiment<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiments" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a dict with all the experiments stored in the instance as value of ids as keys.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[<a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">Experiment</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_num_experiments_in_the_set">
<code class="sig-name descname">get_num_experiments_in_the_set</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_num_experiments_in_the_set" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of experiments currently in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_peptides_present_in_all">
<code class="sig-name descname">get_peptides_present_in_all</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_peptides_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the peptides that are observed in every experiments in the set.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_proband_count">
<code class="sig-name descname">get_proband_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_proband_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of experiments obtained from each proband in the ExperimentalSet.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_proteins_present_in_all">
<code class="sig-name descname">get_proteins_present_in_all</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_proteins_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the proteins that are inferred in all experiments of the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Proteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_tissue_counts">
<code class="sig-name descname">get_tissue_counts</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_tissue_counts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of experiments obtained from each tissue in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_total_peptide_per_org_count">
<code class="sig-name descname">get_total_peptide_per_org_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_total_peptide_per_org_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The total count of peptides per organism accross the all experiments in the set.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_orgs">
<code class="sig-name descname">get_unique_orgs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_orgs" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of the unique organisms in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_peptides">
<code class="sig-name descname">get_unique_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_peptides" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The set of unique peptides in the experimentalSet</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_proteins">
<code class="sig-name descname">get_unique_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_proteins" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the set of unique proteins in the experimentalset</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Proteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_proband">
<code class="sig-name descname">group_by_proband</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span><a class="reference internal" href="index.html#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">IPTK.Classes.ExperimentalSet.ExperimentSet</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_proband" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a map between each proband and an Experimentalset object represent all the experiments objects         belonging to this proband.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,<a class="reference internal" href="index.html#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">ExperimentSet</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_tissue">
<code class="sig-name descname">group_by_tissue</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span><a class="reference internal" href="index.html#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">IPTK.Classes.ExperimentalSet.ExperimentSet</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_tissue" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A map between each tissue and an ExperimentalSet object representing all experiments belonging to that tissue.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,<a class="reference internal" href="index.html#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">ExperimentSet</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.is_peptide_present_in_all">
<code class="sig-name descname">is_peptide_present_in_all</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptide</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.is_peptide_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – The peptide sequence to search its occurrences in every experiment contained in the set</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if peptide is present in all experiments inside the instance, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.is_protein_present_in_all">
<code class="sig-name descname">is_protein_present_in_all</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.is_protein_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein</strong> (<em>str</em>) – the protein id to search its occurrences in every experimental in the set</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if peptide is present in all experiments inside the instance, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.HLAChain">
<span id="iptk-classes-hlachain-module"></span><h7>IPTK.Classes.HLAChain module<a class="headerlink" href="#module-IPTK.Classes.HLAChain" title="Permalink to this headline">¶</a></h7>
<p>The implementation of an HLA molecules</p>
<dl class="py class">
<dt id="IPTK.Classes.HLAChain.HLAChain">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.HLAChain.</code><code class="sig-name descname">HLAChain</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_allele_group">
<code class="sig-name descname">get_allele_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The allele group</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_chain_class">
<code class="sig-name descname">get_chain_class</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_chain_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_name</strong> (<em>str</em>) – the name of the gene</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 if the gene belongs to class one and 2 if it belong to class two</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_class">
<code class="sig-name descname">get_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The HLA class</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_gene">
<code class="sig-name descname">get_gene</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_gene" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the gene name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The chain name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_protein_group">
<code class="sig-name descname">get_protein_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the protein name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.HLAMolecules">
<span id="iptk-classes-hlamolecules-module"></span><h7>IPTK.Classes.HLAMolecules module<a class="headerlink" href="#module-IPTK.Classes.HLAMolecules" title="Permalink to this headline">¶</a></h7>
<p>a representation of an HLA molecules</p>
<dl class="py class">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.HLAMolecules.</code><code class="sig-name descname">HLAMolecule</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">hla_chains</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_allele_group">
<code class="sig-name descname">get_allele_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the allele group for the instance chain/pair of chains</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>AlleleGroup</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_class">
<code class="sig-name descname">get_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The class of the HLA molecules</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_gene">
<code class="sig-name descname">get_gene</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_gene" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>gene/pair of genes coding for the current HLA molecules</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Genes</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">':'</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – the name of the allele by concatenating the names of the individual chains using         a separator, defaults to ‘:’</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>[description]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_protein_group">
<code class="sig-name descname">get_protein_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The protein group for the instance chain/pair of chains</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>ProteinGroup</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.HLASet">
<span id="iptk-classes-hlaset-module"></span><h7>IPTK.Classes.HLASet module<a class="headerlink" href="#module-IPTK.Classes.HLASet" title="Permalink to this headline">¶</a></h7>
<p>An abstraction for a collection of HLA alleles</p>
<dl class="py class">
<dt id="IPTK.Classes.HLASet.HLASet">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.HLASet.</code><code class="sig-name descname">HLASet</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">hlas</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">gene_sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">':'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.HLASet.HLASet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.get_alleles">
<code class="sig-name descname">get_alleles</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.get_alleles" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The class of the HLA-alleles in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.get_class">
<code class="sig-name descname">get_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.get_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The class of the HLA-alleles in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.get_hla_count">
<code class="sig-name descname">get_hla_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.get_hla_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the count of HLA molecules in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_allele">
<code class="sig-name descname">has_allele</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allele</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_allele" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>allele</strong> (<em>str</em>) – the name of the allele to check the instance for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if the provided allele is in the current instance, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_allele_group">
<code class="sig-name descname">has_allele_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allele_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>allele_group</strong> (<em>str</em>) – the allele group to search the set for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if at least one allele in the set belongs to the provided allele group, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_gene">
<code class="sig-name descname">has_gene</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_gene" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_name</strong> (<em>str</em>) – the gene name to search the set against.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if at least one of the alleles in the set belongs to the provided gene. False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_protein_group">
<code class="sig-name descname">has_protein_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_group</strong> – The protein group to search the set for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if at least one allele in the set belongs to the provided protein group</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Peptide">
<span id="iptk-classes-peptide-module"></span><h7>IPTK.Classes.Peptide module<a class="headerlink" href="#module-IPTK.Classes.Peptide" title="Permalink to this headline">¶</a></h7>
<p>A representation of the eluted peptides and its identified proteins.</p>
<dl class="py class">
<dt id="IPTK.Classes.Peptide.Peptide">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Peptide.</code><code class="sig-name descname">Peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_seq</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An representation of an eluted peptide.</p>
<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.add_org_2_parent">
<code class="sig-name descname">add_org_2_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.add_org_2_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>add the source organism of one of the instance parent protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prot_name</strong> (<em>str</em>) – The name of the protein, i.e. the identifier of the protein</p></li>
<li><p><strong>org</strong> (<em>str</em>) – the name of the organism</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – incase the provided protein is not a parent of the provided peptide</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.add_parent_protein">
<code class="sig-name descname">add_parent_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parent_protein</span></em>, <em class="sig-param"><span class="n">start_index</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">end_index</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.add_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>add a protein instance as a parent to the current peptide. 
The library use Python-based indexing where its 0-indexed and ranges are treated as [start, end). 
:param parent_protein: a Protein instance that act as a parent to the peptide.
:type parent_protein: Protein 
:param start_index: the position in the parent protein where the peptide starts
:type start_index: int 
:param end_index: the index of the amino acid that occurs after the last amino acid in the peptide, 
:type start_index: int</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_c_terminal_flank_seq">
<code class="sig-name descname">get_c_terminal_flank_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_len</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_c_terminal_flank_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>:param flank_len:the length of the flanking regions 
:type flank_len: int
:return: a list of string containing the sequences located downstream of the peptide in the parent protein. 
:rtype: [type]</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_flanked_peptide">
<code class="sig-name descname">get_flanked_peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_len</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_flanked_peptide" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_len</strong> (<em>int</em>) – the length of the flanking regions</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of string containing the length of the peptide + the flanking region from               both the N and C terminal of the instance peptide, from all proteins.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_length">
<code class="sig-name descname">get_length</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_length" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the length of the peptides</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_n_terminal_flank_seq">
<code class="sig-name descname">get_n_terminal_flank_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_len</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_n_terminal_flank_seq" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_len</strong> (<em>int</em>) – the length of the flanking regions</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of string containing the sequences located upstream of the peptide in the parent protein.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_non_presented_peptides">
<code class="sig-name descname">get_non_presented_peptides</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_non_presented_peptides" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>length</strong> (<em>int</em>) – The length, i.e. number of amino acids, for the non-presented peptide</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>non-presented peptide from all the parent protein of the current peptide instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_number_of_parents">
<code class="sig-name descname">get_number_of_parents</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_number_of_parents" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of instance parent proteins</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_number_parent_protein">
<code class="sig-name descname">get_number_parent_protein</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_number_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of parent proteins this instance has</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_parent">
<code class="sig-name descname">get_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_parent" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – The protein identifer</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the parent protein that has an id matching the user defined pro_id</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="index.html#IPTK.Classes.Protein.Protein" title="IPTK.Classes.Protein.Protein">Protein</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_parent_proteins">
<code class="sig-name descname">get_parent_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_parents_org">
<code class="sig-name descname">get_parents_org</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_parents_org" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list containing the name of each parent protein source organisms</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Organisms</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_peptide_seq">
<code class="sig-name descname">get_peptide_seq</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_peptide_seq" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the sequence of the peptide.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_pos_in_parent">
<code class="sig-name descname">get_pos_in_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; Tuple<span class="p">[</span>int<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_pos_in_parent" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – the id of the parent protein</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the identifier is not a parent of the instance</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the start and end position of the instance peptide in the parent pointed out by the provided identifier</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Range</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.is_child_of">
<code class="sig-name descname">is_child_of</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.is_child_of" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – is the protein id</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if a user provided protein-id is a parent for the instance peptide, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.map_to_parent_protein">
<code class="sig-name descname">map_to_parent_protein</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.map_to_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Mapped the instance peptide to the parent protein and returned a 
list of numpy arrays where each array has a size of 1 by protein length. 
within the protein the range representing the peptide is encoded as one while
the rest is zero.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of binary encoded arrays represent this mapping.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>MappedProtein</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Proband">
<span id="iptk-classes-proband-module"></span><h7>IPTK.Classes.Proband module<a class="headerlink" href="#module-IPTK.Classes.Proband" title="Permalink to this headline">¶</a></h7>
<p>A description for an IP proband</p>
<dl class="py class">
<dt id="IPTK.Classes.Proband.Proband">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Proband.</code><code class="sig-name descname">Proband</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">info</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Proband.Proband" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.Proband.Proband.get_meta_data">
<code class="sig-name descname">get_meta_data</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; dict<a class="headerlink" href="#IPTK.Classes.Proband.Proband.get_meta_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a dict that contain all the meta-data about the patient</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Proband.Proband.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Proband.Proband.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the patient</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Proband.Proband.update_info">
<code class="sig-name descname">update_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">info</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Proband.Proband.update_info" title="Permalink to this definition">¶</a></dt>
<dd><p>add new or update existing info about the patient  using an arbitrary number of key-value pair to be added to added to the instance meta-info dict</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Protein">
<span id="iptk-classes-protein-module"></span><h7>IPTK.Classes.Protein module<a class="headerlink" href="#module-IPTK.Classes.Protein" title="Permalink to this headline">¶</a></h7>
<p>A representation of a protein that has been inferred from an IP experiment.</p>
<dl class="py class">
<dt id="IPTK.Classes.Protein.Protein">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Protein.</code><code class="sig-name descname">Protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_id</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">seq</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Protein.Protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>representation of a protein that has been infered from an IP experiment.</p>
<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_id">
<code class="sig-name descname">get_id</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_id" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return the protein identifier.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_non_presented_peptide">
<code class="sig-name descname">get_non_presented_peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exc_reg_s_idx</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">exc_reg_e_idx</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_non_presented_peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>sample a peptide from the protein sequences where the sampled peptides is not part of the               experimentally identified regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exc_reg_s_idx</strong> (<em>int</em>) – the start point in the reference protein sequence of the experimentally identified peptide.</p></li>
<li><p><strong>exc_reg_e_idx</strong> (<em>int</em>) – the end point in the reference protein sequence of the experimentally identified peptide.</p></li>
<li><p><strong>length</strong> (<em>int</em>) – length the non-presented peptides.</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if the length of the peptide is bigger than the protein length</p></li>
<li><p><strong>ValueError</strong> – if the length of the peptide is smaller than or equal to zero</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a substring of the instance sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_org">
<code class="sig-name descname">get_org</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_org" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the organism in which this instance protein belong.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_peptides_map">
<code class="sig-name descname">get_peptides_map</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start_idxs</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">end_idxs</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_peptides_map" title="Permalink to this definition">¶</a></dt>
<dd><p>compute a coverage over the protein sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start_idxs</strong> (<em>Index</em>) – a list of integers representing the start positions</p></li>
<li><p><strong>end_idxs</strong> – a list of integers representing the end positions</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if start_indxs and end_idxs MUST be of equal length are not of equal length</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A numpy array with shape of 1 by the length of the protein where every element in the array donates the number of times, It has been observed in the experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_seq">
<code class="sig-name descname">get_seq</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the protein sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.set_org">
<code class="sig-name descname">set_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Protein.Protein.set_org" title="Permalink to this definition">¶</a></dt>
<dd><p>a post-instantitation mechanism to set the organism for which the protein belong.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>org</strong> (<em>str</em>) – the name of the organism</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Tissue">
<span id="iptk-classes-tissue-module"></span><h7>IPTK.Classes.Tissue module<a class="headerlink" href="#module-IPTK.Classes.Tissue" title="Permalink to this headline">¶</a></h7>
<p>A representation of the Tissue used in an IP Experiment.</p>
<dl class="py class">
<dt id="IPTK.Classes.Tissue.ExpressionProfile">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Tissue.</code><code class="sig-name descname">ExpressionProfile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">expression_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">aux_proteins</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>pandas.core.frame.DataFrame<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>a representation of tissue reference expression value.</p>
<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_gene_id_expression">
<code class="sig-name descname">get_gene_id_expression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_gene_id_expression" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_id</strong> (<em>str</em>) – the gene id to retrive its expression value from the database</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided id is not defined in the instance table</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the expression value of the provided gene id.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_gene_name_expression">
<code class="sig-name descname">get_gene_name_expression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_gene_name_expression" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_name</strong> (<em>str</em>) – the gene name to retrive its expression value from the database</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided id is not defined in the instance table</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the expression value of the provided gene name.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the tissue which the instance profile its gene expression</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_table">
<code class="sig-name descname">get_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_table" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a table that contain the expression of all the transcript in the current profile                including core and auxiliary proteins</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Tissue.Tissue">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Tissue.</code><code class="sig-name descname">Tissue</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">main_exp_value</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Database.GeneExpressionDB" title="IPTK.Classes.Database.GeneExpressionDB">IPTK.Classes.Database.GeneExpressionDB</a></span></em>, <em class="sig-param"><span class="n">main_location</span><span class="p">:</span> <span class="n"><a class="reference internal" href="index.html#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a></span></em>, <em class="sig-param"><span class="n">aux_exp_value</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Database.GeneExpressionDB" title="IPTK.Classes.Database.GeneExpressionDB">IPTK.Classes.Database.GeneExpressionDB</a><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">aux_location</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Tissue.Tissue" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.Tissue.Tissue.get_expression_profile">
<code class="sig-name descname">get_expression_profile</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="index.html#IPTK.Classes.Tissue.ExpressionProfile" title="IPTK.Classes.Tissue.ExpressionProfile">IPTK.Classes.Tissue.ExpressionProfile</a><a class="headerlink" href="#IPTK.Classes.Tissue.Tissue.get_expression_profile" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the expresion profile of the current tissue</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="index.html#IPTK.Classes.Tissue.ExpressionProfile" title="IPTK.Classes.Tissue.ExpressionProfile">ExpressionProfile</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.Tissue.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Tissue.Tissue.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the tissue</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.Tissue.get_subCellular_locations">
<code class="sig-name descname">get_subCellular_locations</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="index.html#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a><a class="headerlink" href="#IPTK.Classes.Tissue.Tissue.get_subCellular_locations" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the sub-cellular localization of all the proteins stored in current instance resources.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="index.html#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">CellularLocationDB</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes">
<span id="module-contents"></span><h7>Module contents<a class="headerlink" href="#module-IPTK.Classes" title="Permalink to this headline">¶</a></h7>
</div>
</div>
<span id="document-IPTK.IO"></span><div class="section" id="iptk-io-package">
<h6>IPTK.IO package<a class="headerlink" href="#iptk-io-package" title="Permalink to this headline">¶</a></h6>
<div class="section" id="submodules">
<h7>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h7>
</div>
<div class="section" id="module-IPTK.IO.InFunctions">
<span id="iptk-io-infunctions-module"></span><h7>IPTK.IO.InFunctions module<a class="headerlink" href="#module-IPTK.IO.InFunctions" title="Permalink to this headline">¶</a></h7>
<p>Parse different user inputs into a standard format/tables used by the library.</p>
<dl class="py function">
<dt id="IPTK.IO.InFunctions.download_pdb_entry">
<code class="sig-prename descclassname">IPTK.IO.InFunctions.</code><code class="sig-name descname">download_pdb_entry</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.IO.InFunctions.download_pdb_entry" title="Permalink to this definition">¶</a></dt>
<dd><p>Download the structure of a protein from protein databank form as mmCIF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prot_id</strong> (<em>str</em>) – the protein id</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>IOError</strong> – incase downloading and accessing the data failed</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the path to the downloaded file</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.InFunctions.fasta2dict">
<code class="sig-prename descclassname">IPTK.IO.InFunctions.</code><code class="sig-name descname">fasta2dict</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2fasta</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">filter_decoy</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">decoy_string</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'DECOY'</span></em><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span><a class="headerlink" href="#IPTK.IO.InFunctions.fasta2dict" title="Permalink to this definition">¶</a></dt>
<dd><p>loads a fasta file and construct a dict object where ids are keys and sequences are the value</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path2fasta</strong> (<em>str</em>) – The path to load the fasta file</p></li>
<li><p><strong>filter_decoy</strong> (<em>bool</em><em>, </em><em>optional</em>) – A boolean of whether or not to filter the decoy sequences from the database, defaults to True</p></li>
<li><p><strong>decoy_string</strong> (<em>str</em><em>, </em><em>optional</em>) – The decoy database prefix, only valid incase filter_decoy is set to true, defaults to ‘DECOY’</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>IOError</strong> – [description]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a dict where the protein ids are the keys and the protein sequences are the value</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.InFunctions.load_identification_table">
<code class="sig-prename descclassname">IPTK.IO.InFunctions.</code><code class="sig-name descname">load_identification_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">input_path</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.IO.InFunctions.load_identification_table" title="Permalink to this definition">¶</a></dt>
<dd><p>load &amp; process an identification table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_path</strong> (<em>str</em>) – the path two the identification table. with the following columns: peptides which hold the identified
peptide sequence, protein which hold the identified protein sequence, start_index, and end_index where
the last two columns define the position of the peptide in the parent protein.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – The separator to parse the provided table.</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>IOError</strong> – [description]</p></li>
<li><p><strong>ValueError</strong> – [description]</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>[description]</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.InFunctions.parse_mzTab_to_identification_table">
<code class="sig-prename descclassname">IPTK.IO.InFunctions.</code><code class="sig-name descname">parse_mzTab_to_identification_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2mzTab</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">path2fastaDB</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">fasta_reader_param</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'decoy_string': 'DECOY', 'filter_decoy': True}</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.IO.InFunctions.parse_mzTab_to_identification_table" title="Permalink to this definition">¶</a></dt>
<dd><p>parse a user provided mzTab to an identification table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path2mzTab</strong> (<em>str</em>) – the path to the input mzTab file</p></li>
<li><p><strong>path2fastaDB</strong> (<em>str</em>) – the path to a fasta sequence database to obtain the protein sequences</p></li>
<li><p><strong>fasta_reader_param</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict of parameters for controlling the behavior of the fasta reader , defaults to {‘filter_decoy’:True, ‘decoy_string’:’DECOY’ }</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>IOError</strong> – if the mztab file could not be open and loaded or if the fasta database could not be read</p></li>
<li><p><strong>KeyError</strong> – if a protein id defined in the mzTab file could not be extracted from a matched sequence database</p></li>
<li><p><strong>ValueError</strong> – if the peptide can not be mapped to the identified protein</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the identification table</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.InFunctions.parse_text_table">
<code class="sig-prename descclassname">IPTK.IO.InFunctions.</code><code class="sig-name descname">parse_text_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2file</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">path2fastaDB</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">','</span></em>, <em class="sig-param"><span class="n">fasta_reader_param</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'decoy_string': 'DECOY', 'filter_decoy': True}</span></em>, <em class="sig-param"><span class="n">seq_column</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Sequence'</span></em>, <em class="sig-param"><span class="n">accession_column</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Protein Accessions'</span></em>, <em class="sig-param"><span class="n">protein_group_sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">';'</span></em>, <em class="sig-param"><span class="n">remove_protein_version</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">remove_if_not_matched</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.IO.InFunctions.parse_text_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a user defined table to extract the columns containing the identification table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path2file</strong> (<em>str</em>) – The path to load the CSV file holding the results</p></li>
<li><p><strong>path2fastaDB</strong> (<em>str</em>) – The path to a fasta sequence database to obtain the protein sequences</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – The table separators, defaults to ‘,’</p></li>
<li><p><strong>fasta_reader_param</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict of parameters for controlling the behavior of the fasta reader, defaults to {‘filter_decoy’:True, ‘decoy_string’:’DECOY’ }</p></li>
<li><p><strong>seq_column</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the columns containing the peptide sequence, defaults to ‘Sequence’</p></li>
<li><p><strong>accession_column</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the column containing the protein accession , defaults to ‘Protein Accessions’</p></li>
<li><p><strong>protein_group_sep</strong> (<em>str</em><em>, </em><em>optional</em>) – The separator for the protein group,, defaults to ‘;’</p></li>
<li><p><strong>remove_protein_version</strong> (<em>bool</em><em>, </em><em>optional</em>) – A bool if true strip the version number from the protein , defaults to True</p></li>
<li><p><strong>remove_if_not_matched</strong> (<em>bool</em><em>, </em><em>optional</em>) – remove the peptide if it could not be matched to the parent protein, defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>IOError</strong> – Incase either the sequences database or the identification table can not be open and loaded</p></li>
<li><p><strong>KeyError</strong> – In case the provided column names not in the provided identification table.</p></li>
<li><p><strong>KeyError</strong> – Incase the protein sequence can not be extract from the sequence database</p></li>
<li><p><strong>ValueError</strong> – incase the peptide could not be located in the protein sequence</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>an identification table</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.InFunctions.parse_xml_based_format_to_identification_table">
<code class="sig-prename descclassname">IPTK.IO.InFunctions.</code><code class="sig-name descname">parse_xml_based_format_to_identification_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2XML_file</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">path2fastaDB</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">decoy_prefix</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'DECOY'</span></em>, <em class="sig-param"><span class="n">is_idXML</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">fasta_reader_param</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'decoy_string': 'DECOY', 'filter_decoy': True}</span></em>, <em class="sig-param"><span class="n">remove_if_not_matched</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.IO.InFunctions.parse_xml_based_format_to_identification_table" title="Permalink to this definition">¶</a></dt>
<dd><p>parse either a pepXML or an idXML file to generate an identification table ,</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path2XML_file</strong> (<em>str</em>) – The path to the input pepXML files</p></li>
<li><p><strong>path2fastaDB</strong> (<em>str</em>) – The path to a fasta sequence database to obtain the protein sequences</p></li>
<li><p><strong>decoy_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – the prefix of the decoy sequences, default is DECOY, defaults to ‘DECOY’</p></li>
<li><p><strong>is_idXML</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether or not the provided file is an idXML, default is false which assume the provided file is a pepXML file, defaults to False</p></li>
<li><p><strong>fasta_reader_param</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict of parameters for controlling the behavior of the fasta reader , defaults to {‘filter_decoy’:True, ‘decoy_string’:’DECOY’ }</p></li>
<li><p><strong>remove_if_not_matched</strong> (<em>bool</em><em>, </em><em>optional</em>) – remove the peptide if it could not be matched to the parent protein,, defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>IOError</strong> – if the fasta database could not be open</p></li>
<li><p><strong>ValueError</strong> – if the XML file can not be open</p></li>
<li><p><strong>KeyError</strong> – if a protein id defined in the mzTab file could not be extracted from a matched sequence database</p></li>
<li><p><strong>ValueError</strong> – if the peptide can not be mapped to the identified protein</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the identification table</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.IO.MEMEInterface">
<span id="iptk-io-memeinterface-module"></span><h7>IPTK.IO.MEMEInterface module<a class="headerlink" href="#module-IPTK.IO.MEMEInterface" title="Permalink to this headline">¶</a></h7>
<p>The module contains functions to to call meme software via a system call.</p>
<dl class="py function">
<dt id="IPTK.IO.MEMEInterface.call_meme">
<code class="sig-prename descclassname">IPTK.IO.MEMEInterface.</code><code class="sig-name descname">call_meme</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">input_fasta_file</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">output_dir</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">objfunc</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'classic'</span></em>, <em class="sig-param"><span class="n">test</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'mhg'</span></em>, <em class="sig-param"><span class="n">use_llr</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">shuf</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">2</span></em>, <em class="sig-param"><span class="n">hsfrac</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">cefrac</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.25</span></em>, <em class="sig-param"><span class="n">searchsize</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">maxsize</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">norand</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">csites</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">seed</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">mod</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'oops'</span></em>, <em class="sig-param"><span class="n">nmotifs</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">evt</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">- 1.0</span></em>, <em class="sig-param"><span class="n">time</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">nsite</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">minsites</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">maxsite</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">nsites</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">w</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">minw</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">maxw</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">nomatrim</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">wg</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">ws</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">noendgaps</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">maxiter</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">prior</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'dirichlet'</span></em>, <em class="sig-param"><span class="n">b</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">p</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.IO.MEMEInterface.call_meme" title="Permalink to this definition">¶</a></dt>
<dd><p>warper for making a system call to meme software for sequence motif finding         for the reset of the function parameters use the function             <strong>get_meme_help</strong> defined in the module IO, submodule MEMEInterface. </p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_fasta_file</strong> (<em>str</em>) – The path to input FASTA files.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – the output dir to write the results, <strong>IT WILL OVERWRITE EXISTING DIRECTORY</strong></p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – whether or not to print the output of calling meme to the screen, default is True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.MEMEInterface.get_meme_help">
<code class="sig-prename descclassname">IPTK.IO.MEMEInterface.</code><code class="sig-name descname">get_meme_help</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.IO.MEMEInterface.get_meme_help" title="Permalink to this definition">¶</a></dt>
<dd><p>print the command line help interface for the meme tool</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>FileNotFoundError</strong> – if meme is not callable</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.MEMEInterface.is_meme_callable">
<code class="sig-prename descclassname">IPTK.IO.MEMEInterface.</code><code class="sig-name descname">is_meme_callable</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.IO.MEMEInterface.is_meme_callable" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if meme is callable, False otherwise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.IO.OutFunctions">
<span id="iptk-io-outfunctions-module"></span><h7>IPTK.IO.OutFunctions module<a class="headerlink" href="#module-IPTK.IO.OutFunctions" title="Permalink to this headline">¶</a></h7>
<p>Write the results generated by the library into a wide variety of formats.</p>
<dl class="py function">
<dt id="IPTK.IO.OutFunctions.write_annotated_sequences">
<code class="sig-prename descclassname">IPTK.IO.OutFunctions.</code><code class="sig-name descname">write_annotated_sequences</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptides</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">labels</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">path2write</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">','</span></em>, <em class="sig-param"><span class="n">shuffle</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.IO.OutFunctions.write_annotated_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>take a list of peptides along with it sequences and write the results to a CSV file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – a list of peptide sequences</p></li>
<li><p><strong>labels</strong> (<em>List</em><em>[</em><em>int</em><em>]</em>) – a list of numerical labels associated with the peptides</p></li>
<li><p><strong>path2write</strong> (<em>str</em>) – the path to write the generated file</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – The separator in the resulting table,  defaults to ‘,’</p></li>
<li><p><strong>shuffle</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether or not to shuffle the table , defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – incase the length of the tables and labels is not matching</p></li>
<li><p><strong>IOError</strong> – In case writing the output table failed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.OutFunctions.write_auto_named_peptide_to_fasta">
<code class="sig-prename descclassname">IPTK.IO.OutFunctions.</code><code class="sig-name descname">write_auto_named_peptide_to_fasta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptides</span><span class="p">:</span> <span class="n">List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">output_file</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.IO.OutFunctions.write_auto_named_peptide_to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of peptides, generate automatic names for the peptides and write the results to the disk as FASTA files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>Peptides</em>) – a list of peptide sequences</p></li>
<li><p><strong>output_file</strong> (<em>str</em>) – the name of the output file to write the results to, it will OVERWRITE existing files</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.OutFunctions.write_mapped_tensor_to_h5py">
<code class="sig-prename descclassname">IPTK.IO.OutFunctions.</code><code class="sig-name descname">write_mapped_tensor_to_h5py</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tensor</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">path2write</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">dataSet_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'MAPPED_TENSOR'</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.IO.OutFunctions.write_mapped_tensor_to_h5py" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a mapped tensor to an hdf5 file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tensor</strong> (<em>np.ndarray</em>) – The provided tensor to  write it to the hdf5 file.</p></li>
<li><p><strong>path2write</strong> (<em>str</em>) – The path of the output file</p></li>
<li><p><strong>dataSet_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the dataset inside the mapped tensor, defaults to ‘MAPPED_TENSOR’</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>IOError</strong> – In case opening the file for writing failed</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.OutFunctions.write_named_peptides_to_fasta">
<code class="sig-prename descclassname">IPTK.IO.OutFunctions.</code><code class="sig-name descname">write_named_peptides_to_fasta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">names</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">peptides</span><span class="p">:</span> <span class="n">List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">output_file</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.IO.OutFunctions.write_named_peptides_to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of names and peptide sequences and writes them as an output file to the disk as fasta files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>names</strong> (<em>Names</em>) – A list of sequences names</p></li>
<li><p><strong>peptides</strong> (<em>Peptides</em>) – A list of peptide sequences</p></li>
<li><p><strong>output_file</strong> (<em>str</em>) – The name of the output file to write the results to, it will OVERWRITE existing files</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – Incase the length of the tables and labels is not matching</p></li>
<li><p><strong>IOError</strong> – In case writing the output file failed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.IO.OutFunctions.write_pep_file">
<code class="sig-prename descclassname">IPTK.IO.OutFunctions.</code><code class="sig-name descname">write_pep_file</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptides</span><span class="p">:</span> <span class="n">List<span class="p">[</span><a class="reference internal" href="index.html#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">output_file</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.IO.OutFunctions.write_pep_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a file and write the peptides to .pep file which is a text file that contain one peptide per line</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptides</strong> (<em>Peptides</em>) – a list of peptide sequences</p></li>
<li><p><strong>output_file</strong> (<em>str</em>) – the name of the output file to write the results to, it will OVERWRITE existing files</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>IOError</strong> – In case writing the output file failed</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.IO">
<span id="module-contents"></span><h7>Module contents<a class="headerlink" href="#module-IPTK.IO" title="Permalink to this headline">¶</a></h7>
</div>
</div>
<span id="document-IPTK.Utils"></span><div class="section" id="iptk-utils-package">
<h6>IPTK.Utils package<a class="headerlink" href="#iptk-utils-package" title="Permalink to this headline">¶</a></h6>
<div class="section" id="submodules">
<h7>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h7>
</div>
<div class="section" id="module-IPTK.Utils.DevFunctions">
<span id="iptk-utils-devfunctions-module"></span><h7>IPTK.Utils.DevFunctions module<a class="headerlink" href="#module-IPTK.Utils.DevFunctions" title="Permalink to this headline">¶</a></h7>
<p>The module contain functions that can be used for developing &amp; testing other functions of the library</p>
<dl class="py function">
<dt id="IPTK.Utils.DevFunctions.generate_random_peptide_seq">
<code class="sig-prename descclassname">IPTK.Utils.DevFunctions.</code><code class="sig-name descname">generate_random_peptide_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptide_length</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">num_peptides</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Utils.DevFunctions.generate_random_peptide_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>generate a list of random peptides for testing and developing purposes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>peptide_length</strong> (<em>int</em>) – The peptide length</p></li>
<li><p><strong>num_peptides</strong> (<em>int</em>) – The number of peptides in the generate list</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of random peptides</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.DevFunctions.simulate_an_experimental_ident_table_from_fasta">
<code class="sig-prename descclassname">IPTK.Utils.DevFunctions.</code><code class="sig-name descname">simulate_an_experimental_ident_table_from_fasta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2load</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">num_pep</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">num_prot</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Utils.DevFunctions.simulate_an_experimental_ident_table_from_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>simulate an IP identification table from a fasta file. Please Note,  if the reminder of num_pep over num_prot does not equal to zero,
the floor of this ratio will be used to sample peptides from each proteins</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path2load</strong> (<em>str</em>) – The path to load the Fasta files</p></li>
<li><p><strong>num_pep</strong> (<em>int</em>) – The number of peptides in the tables</p></li>
<li><p><strong>num_prot</strong> (<em>int</em>) – The number of UNIQUE proteins in the table</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if number of proteins or number of peptide is zero</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>an identification table</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.DevFunctions.simulate_an_expression_table">
<code class="sig-prename descclassname">IPTK.Utils.DevFunctions.</code><code class="sig-name descname">simulate_an_expression_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_transcripts</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Utils.DevFunctions.simulate_an_expression_table" title="Permalink to this definition">¶</a></dt>
<dd><p>create a dummy expression table to be used for testing and developing Tissue based classes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num_transcripts</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of transcripts that shall be contained in the transcript , defaults to 100</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – incase number of transcripts is 0</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>[description]</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.DevFunctions.simulate_mapped_array_list">
<code class="sig-prename descclassname">IPTK.Utils.DevFunctions.</code><code class="sig-name descname">simulate_mapped_array_list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">min_len</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">20</span></em>, <em class="sig-param"><span class="n">max_len</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">num_elem</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Utils.DevFunctions.simulate_mapped_array_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate a list of mapped arrays proteins to be used for developing purposes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>min_len</strong> (<em>int</em><em>, </em><em>optional</em>) – the minmum length of the protein , defaults to 20</p></li>
<li><p><strong>max_len</strong> (<em>int</em><em>, </em><em>optional</em>) – the maximum length for the protein , defaults to 100</p></li>
<li><p><strong>num_elem</strong> (<em>int</em><em>, </em><em>optional</em>) – the number of arrays in the protein , defaults to 100</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of simulated NumPy array that represent protein peptide coverage</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>List[np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.DevFunctions.simulate_random_experiment">
<code class="sig-prename descclassname">IPTK.Utils.DevFunctions.</code><code class="sig-name descname">simulate_random_experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alleles</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">path2fasta</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'TEST_TISSUE'</span></em>, <em class="sig-param"><span class="n">num_pep</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">num_prot</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em>, <em class="sig-param"><span class="n">proband_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a><a class="headerlink" href="#IPTK.Utils.DevFunctions.simulate_random_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate a random experiment objects</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alleles</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – a list of alleles names.</p></li>
<li><p><strong>path2fasta</strong> (<em>str</em>) – The path to load the database objects</p></li>
<li><p><strong>tissue_name</strong> (<em>str</em><em>, </em><em>optional</em>) – the name of the tissue, defaults to ‘TEST_TISSUE’</p></li>
<li><p><strong>num_pep</strong> (<em>int</em><em>, </em><em>optional</em>) – the number of peptides in the table, defaults to 10</p></li>
<li><p><strong>num_prot</strong> (<em>int</em><em>, </em><em>optional</em>) – number of proteins, defaults to 5</p></li>
<li><p><strong>proband_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the Proband, defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a simulated experimental object</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="index.html#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">Experiment</a></p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.Utils.Mapping">
<span id="iptk-utils-mapping-module"></span><h7>IPTK.Utils.Mapping module<a class="headerlink" href="#module-IPTK.Utils.Mapping" title="Permalink to this headline">¶</a></h7>
<p>A submodule that contain function to map different database keys</p>
<dl class="py function">
<dt id="IPTK.Utils.Mapping.map_from_uniprot_gene">
<code class="sig-prename descclassname">IPTK.Utils.Mapping.</code><code class="sig-name descname">map_from_uniprot_gene</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">uniprots</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Utils.Mapping.map_from_uniprot_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>map from uniprot id to ensemble gene ids</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>uniprots</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – a list of uniprot IDs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A table that contain the mapping between each uniprot and its corresponding Gene ID/IDs</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.Mapping.map_from_uniprot_pdb">
<code class="sig-prename descclassname">IPTK.Utils.Mapping.</code><code class="sig-name descname">map_from_uniprot_pdb</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">uniprots</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Utils.Mapping.map_from_uniprot_pdb" title="Permalink to this definition">¶</a></dt>
<dd><p>map from uniprot id to protein data bank identifiers</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>uniprots</strong> (<em>List</em><em>[</em><em>str</em><em>]</em>) – a list of uniprot IDs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A table that contain the mapping between each uniprot and its corresponding PDB ID/IDs</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.Utils.Types">
<span id="iptk-utils-types-module"></span><h7>IPTK.Utils.Types module<a class="headerlink" href="#module-IPTK.Utils.Types" title="Permalink to this headline">¶</a></h7>
<p>Contain a definition of commonly used types through the library</p>
</div>
<div class="section" id="module-IPTK.Utils.UtilityFunction">
<span id="iptk-utils-utilityfunction-module"></span><h7>IPTK.Utils.UtilityFunction module<a class="headerlink" href="#module-IPTK.Utils.UtilityFunction" title="Permalink to this headline">¶</a></h7>
<p>Utility functions that are used through the library</p>
<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.append_to_calling_string">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">append_to_calling_string</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">param</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">def_value</span></em>, <em class="sig-param"><span class="n">cur_val</span></em>, <em class="sig-param"><span class="n">calling_string</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">is_flag</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Utils.UtilityFunction.append_to_calling_string" title="Permalink to this definition">¶</a></dt>
<dd><p>help function that take a calling string, a parameter, a default value and current value     if the parameter does not equal its default value the function append the parameter with its current 
value to the calling string adding a space before the calling_string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>param</strong> (<em>str</em>) – The name of the parameter that will be append to the calling string</p></li>
<li><p><strong>def_value</strong> (<em>[</em><em>type</em><em>]</em>) – The default value for the parameter</p></li>
<li><p><strong>cur_val</strong> (<em>[</em><em>type</em><em>]</em>) – The current value for the parameter</p></li>
<li><p><strong>calling_string</strong> (<em>str</em>) – The calling string in which the parameter and the current value might be appended to it</p></li>
<li><p><strong>is_flag</strong> (<em>bool</em><em>, </em><em>optional</em>) – If the parameter is a control flag, i.e. a boolean switch, it append the parameter to the calling string without associating a value to it , defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the updated version of the calling string</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.build_sequence_table">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">build_sequence_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Utils.UtilityFunction.build_sequence_table" title="Permalink to this definition">¶</a></dt>
<dd><p>construct a sequences database from sequences dict object</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em>) – a dict that contain the protein ids as keys and sequences as values.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>pandas dataframe that contain the protein ID and the associated protein sequence</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.check_peptide_made_of_std_20_aa">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">check_peptide_made_of_std_20_aa</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptide</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Utils.UtilityFunction.check_peptide_made_of_std_20_aa" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the peptide is made of the standard 20 amino acids, if this is the case, 
it return the peptide sequence, otherwise it return an empty string</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – a peptide sequence to check its composition</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if the peptide is made of the standard 20 amino acids, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.generate_color_scale">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">generate_color_scale</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">color_ranges</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.colors.LinearSegmentedColormap<a class="headerlink" href="#IPTK.Utils.UtilityFunction.generate_color_scale" title="Permalink to this definition">¶</a></dt>
<dd><p>generate a color gradient with number of steps equal to color_ranges -1</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>color_ranges</strong> (<em>int</em>) – the number of colors in the range</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a color gradient palette</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>matplotlib.colors.LinearSegmentedColormap</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.generate_random_name">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">generate_random_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Utils.UtilityFunction.generate_random_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name_length</strong> (<em>int</em>) – Generate a random ASCII based string</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>[description]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.generate_random_protein_mapping">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">generate_random_protein_mapping</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_len</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">max_coverage</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Utils.UtilityFunction.generate_random_protein_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a NumPy array with shape of 1 by protein_len where the elements in the array 
is a random integer between zero &amp;  max_coverage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_len</strong> (<em>int</em>) – The length of the protein</p></li>
<li><p><strong>max_coverage</strong> (<em>int</em>) – The maximum peptide coverage at each position</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a NumPy array contain a simulate protein coverage</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.get_idx_peptide_in_sequence_table">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">get_idx_peptide_in_sequence_table</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">peptide</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Utils.UtilityFunction.get_idx_peptide_in_sequence_table" title="Permalink to this definition">¶</a></dt>
<dd><p>check the sequences table if the provided peptide is locate in one of its sequences and returns 
a list of protein identifiers containing the identifier of the hit proteins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence_table</strong> (<em>pd.DataFrame</em>) – pandas dataframe that contain the protein ID and the associated protein sequence</p></li>
<li><p><strong>peptide</strong> (<em>str</em>) – the peptide sequence to query the protein with</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of protein identifiers containing the identifier of the hit proteins</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.load_3d_figure">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">load_3d_figure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file_path</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Utils.UtilityFunction.load_3d_figure" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – Load a pickled 3D figure from thr provided path</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>IOError</strong> – The path of the pickled figure.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a matplotlib figure</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>plt.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.pad_mapped_proteins">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">pad_mapped_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">list_array</span><span class="p">:</span> <span class="n">List<span class="p">[</span>numpy.ndarray<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">pre_pad</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">padding_char</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Utils.UtilityFunction.pad_mapped_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>Pad the provided list of array into a 2D tensor of shape number of arrays by maxlength.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>list_array</strong> (<em>List</em><em>[</em><em>np.ndarray</em><em>]</em>) – A list of NumPy arrays where each array is a mapped_protein array,     the expected shape of these arrays is 1 by protein length.</p></li>
<li><p><strong>pre_pad</strong> (<em>bool</em><em>, </em><em>optional</em>) – pre or post padding of shorter array in the library.Default is pre-padding, defaults to True</p></li>
<li><p><strong>padding_char</strong> (<em>int</em><em>, </em><em>optional</em>) – The padding char, defaults to -1</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 2D tensor of shape number of arrays by maxlength.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.save_3d_figure">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">save_3d_figure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">outpath</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">fig2save</span><span class="p">:</span> <span class="n">matplotlib.figure.Figure</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Utils.UtilityFunction.save_3d_figure" title="Permalink to this definition">¶</a></dt>
<dd><p>write a pickled version of the a 3D figure so it can be loaded later for more interactive analysis</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outpath</strong> (<em>str</em>) – The output path of the writer function</p></li>
<li><p><strong>fig2save</strong> (<em>plt.Figure</em>) – The figure to save to the output file</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>IOError</strong> – In case writing the file failed</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.simulate_protein_binary_represention">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">simulate_protein_binary_represention</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_conditions</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">protein_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Utils.UtilityFunction.simulate_protein_binary_represention" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_conditions</strong> (<em>int</em>) – The number of conditions to simulate</p></li>
<li><p><strong>protein_length</strong> (<em>int</em>) – The Length of the protein</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 2D matrix of shape protein_length by number of conditions, where each element can be either zero.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Utils.UtilityFunction.simulate_protein_representation">
<code class="sig-prename descclassname">IPTK.Utils.UtilityFunction.</code><code class="sig-name descname">simulate_protein_representation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_conditions</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">protein_len</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">protein_coverage</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Utils.UtilityFunction.simulate_protein_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate protein peptide coverage under-different conditions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_conditions</strong> (<em>[</em><em>type</em><em>]</em>) – The number of condition to simulate</p></li>
<li><p><strong>protein_len</strong> (<em>[</em><em>type</em><em>]</em>) – The length of the protein</p></li>
<li><p><strong>protein_coverage</strong> (<em>[</em><em>type</em><em>]</em>) – The maximum protein coverage</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dict of length num_conditions contains that the condition index and a simulated protein array</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Dict[str, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.Utils">
<span id="module-contents"></span><h7>Module contents<a class="headerlink" href="#module-IPTK.Utils" title="Permalink to this headline">¶</a></h7>
</div>
</div>
<span id="document-IPTK.Visualization"></span><div class="section" id="iptk-visualization-package">
<h6>IPTK.Visualization package<a class="headerlink" href="#iptk-visualization-package" title="Permalink to this headline">¶</a></h6>
<div class="section" id="submodules">
<h7>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h7>
</div>
<div class="section" id="module-IPTK.Visualization.vizTools">
<span id="iptk-visualization-viztools-module"></span><h7>IPTK.Visualization.vizTools module<a class="headerlink" href="#module-IPTK.Visualization.vizTools" title="Permalink to this headline">¶</a></h7>
<p>The module contain visualization functions the can be used to plot the results obtained from the 
methods of the classes defined in the Class module or from the analysis functions defined in the Analysis Module.</p>
<dl class="py function">
<dt id="IPTK.Visualization.vizTools.imposed_coverage_on_3D_structure">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">imposed_coverage_on_3D_structure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2mmCIF</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">mapped_protein</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">background_color</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'black'</span></em>, <em class="sig-param"><span class="n">low</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'red'</span></em>, <em class="sig-param"><span class="n">high</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'blue'</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Visualization.vizTools.imposed_coverage_on_3D_structure" title="Permalink to this definition">¶</a></dt>
<dd><p>A function to impose the peptide coverage on top of a protein 3D structure     where the color of each position is marked by a color gradient that reflect the number of peptides 
aligned to this position. Note: Use the function with Jupyter-note book as it return an NGLWidget object that your can explore     and navigate from the browser.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path2mmCIF</strong> (<em>str</em>) – The path to load the mmCIF file containing the protein structure</p></li>
<li><p><strong>mapped_protein</strong> (<em>np.ndarray</em>) – a Numpy array of containg the number of peptides originated from each position in the array</p></li>
<li><p><strong>background_color</strong> (<em>str</em><em>, </em><em>optional</em>) – the color of the background, default is black , defaults to ‘black’</p></li>
<li><p><strong>low</strong> (<em>str</em><em>, </em><em>optional</em>) – the color of low covered position, default is red. , defaults to ‘red’</p></li>
<li><p><strong>high</strong> (<em>str</em><em>, </em><em>optional</em>) – the color of high covered position, default is violet., defaults to ‘blue’</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – incase the provided path to the structure file does not exists</p></li>
<li><p><strong>IOError</strong> – if the structure file can not be loaded or if more than one file are located in the provided path</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_change_in_presentation_between_experiment">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_change_in_presentation_between_experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">change_in_presentation_array</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">index_first</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">index_second</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">'Change in protein presentation'</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="o">=</span><span class="default_value">'Proteins'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="o">=</span><span class="default_value">'magnitude of change in protein count'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_change_in_presentation_between_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the change in protein presentation between two experiment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>change_in_presentation_array</strong> (<em>np.ndarray</em>) – a 3D tensor of shape number of experiments by number of experiment by number of identified proteins.</p></li>
<li><p><strong>index_first</strong> (<em>int</em>) – [description]</p></li>
<li><p><strong>index_second</strong> (<em>int</em>) – the index of the first experiment in the tensor</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.scatterplot function, defaults to {}</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Change in protein presentation’</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The axis on the x-axis , defaults to “Proteins”</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The axis on the y-axis, defaults to “magnitude of change in protein count”</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if the provided tensor is not of rank 3</p></li>
<li><p><strong>IndexError</strong> – if the provided indices are out of bound</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_experiment_set_counts">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_experiment_set_counts</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">counts_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">log_scale</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.pyplot.figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_experiment_set_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>visualize the number of peptides and number of peptides-per-organism per experiment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts_table</strong> (<em>pd.DataFrame</em>) – a pandas dataframe that contain the count, organism name and the count</p></li>
<li><p><strong>log_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Normalize the peptide counts one log 10, defaults to False</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.catplot function, defaults to {}</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_gene_expression_vs_num_peptides">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_gene_expression_vs_num_peptides</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exp_count_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">def_value</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Expression value'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Peptides per protein Vs. Expression Level'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_gene_expression_vs_num_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the correlation between the gene expression and the num of peptids per protein using seaborn library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_count_table</strong> (<em>pd.DataFrame</em>) – A table that contain the number of peptides and the expresion value for each protein in the database</p></li>
<li><p><strong>tissue_name</strong> (<em>str</em>) – The name of the tissue</p></li>
<li><p><strong>def_value</strong> (<em>float</em><em>, </em><em>optional</em>) – The default value for proteins that could not be mapped to the expression database, defaults to -1</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.scatter function, defaults to {}</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis, defaults to ‘Expression value’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Peptides per protein Vs. Expression Level’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_motif">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_motif</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pwm_df</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'fade_probabilities': True}</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_motif" title="Permalink to this definition">¶</a></dt>
<dd><p>A generic motif plotter that forward its argument to logomaker for making plots</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pwm_df</strong> (<em>pd.DataFrame</em>) – A pandas dataframe containing the position weighted matrix</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>PlottingKeywards</em><em>, </em><em>optional</em>) – A dictionary of parameter to control the behavior of the logo plotter, defaults to {‘fade_probabilities’:True}</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a matplotlib figure instance containing the ploted motif</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plt.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_num_peptide_per_go_term">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_num_peptide_per_go_term</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep2goTerm</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'GO-Term'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per GO Term'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_num_peptide_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of peptides obtained per Go-Term  using matplotlib library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep2goTerm</strong> (<em>pd.DataFrame</em>) – A table that contain the count of peptides from each GO-Term</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.barplot function, defaults to {}</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop peptide with unknown GO-term, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis, defaults to ‘GO-Term’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Number of peptides per GO Term’</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>[description]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plt.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_num_peptides_per_location">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_num_peptides_per_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep2loc</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Compartment'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per sub-cellular compartment'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_num_peptides_per_location" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of peptides obtained from each compartment using seaborn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep2loc</strong> (<em>pd.DataFrame</em>) – A table that contain the count of peptides from each  location</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.barplot function, defaults to {}</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop protein with unknown location, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the y-axis, defaults to ‘Compartment’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Number of peptides per sub-cellular compartment’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_num_peptides_per_organism">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_num_peptides_per_organism</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_per_org</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">log_scale</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Organism'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per organism'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_num_peptides_per_organism" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of peptides per each organism inferred from the experiment using seaborn and matlotlib.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep_per_org</strong> (<em>pd.DataFrame</em>) – A table that contain the number of peptides belonging to each organism</p></li>
<li><p><strong>log_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether or not to scale the number of peptides using a log scale, default is False, defaults to False</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.barplot function, defaults to {}</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the y-axis, defaults to ‘Organism’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Number of peptides per organism’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_num_peptides_per_parent">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_num_peptides_per_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nums_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">num_prot</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">x_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">y_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Protein ID'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per protein'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Visualization.vizTools.plot_num_peptides_per_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize a histogram of the eluted peptide length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nums_table</strong> (<em>pd.DataFrame</em>) – a pandas dataframe containing number of peptides identified from each protein.</p></li>
<li><p><strong>num_prot</strong> (<em>int</em><em>, </em><em>optional</em>) – the number of protein to show relative to the first element, for example, the first 10, 20 etc.     If the default value of -1 is used then all protein will be plotted, however, this might lead to a crowded figure,     defaults to -1.</p></li>
<li><p><strong>plotting_kwargs</strong> – a dict object containing parameters for the function     seaborn::distplot, defaults to {}</p></li>
<li><p><strong>x_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>y_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the y-axis, defaults to ‘Protein ID’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Number of peptides per protein’</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if the num_prot is bigger than the number of elements in the provided table</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_num_protein_per_go_term">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_num_protein_per_go_term</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein2goTerm</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of Proteins'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Compartment'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of proteins per sub-cellular compartment'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_num_protein_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of proteins per each GO Term</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein2goTerm</strong> (<em>pd.DataFrame</em>) – A table that contain the count of proteins from each GO-Term</p></li>
<li><p><strong>tissue_name</strong> (<em>str</em>) – a dict object containing parameters for the sns.barplot function.</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.barplot function, defaults to {}</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop protein with unknown location, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of Proteins’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis, defaults to ‘Compartment’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure, defaults to ‘Number of proteins per sub-cellular compartment’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_num_protein_per_location">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_num_protein_per_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_loc</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of Proteins'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Compartment'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of proteins per sub-cellular compartment'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_num_protein_per_location" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of proteins per each sub-cellular compartment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_loc</strong> (<em>pd.DataFrame</em>) – A table that contain the count of protein from each location.</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.barplot function, defaults to {}</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop protein with unknown location, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of Proteins’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis, defaults to ‘Compartment’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure, defaults to ‘Number of proteins per sub-cellular compartment’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_overlap_heatmap">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_overlap_heatmap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">results_df</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em><span class="sig-paren">)</span> &#x2192; seaborn.matrix.ClusterGrid<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_overlap_heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot a user provided dataframe as a cluster heatmap using seaborn library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>results_df</strong> (<em>pd.DataFrame</em>) – A pandas dataframe table that hold the overlapping number</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>PlottingKeywards</em><em>, </em><em>optional</em>) – forward parameter to the clustermap function, defaults to {}</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the generated clustermap</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>sns.matrix.ClusterGrid</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_overlay_representation">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_overlay_representation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proteins</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">legend_pos</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">2</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_overlay_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>plot an overlayed representation for the SAME protein or proteins OF EQUAL LENGTH in different conditions        in which the mapped represention of each protein are stacked on top of each other to generate        a representation for the protein representability under different conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>proteins</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>np.ndarray</em><em>]</em><em>]</em>) – a nested dict object containing for each protein a child dict that contain the mapping array and the color in the figure.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>optional</em>) – the transparency between proteins , defaults to 0.5</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to None</p></li>
<li><p><strong>legend_pos</strong> (<em>int</em><em>, </em><em>optional</em>) – the position of the legend , defaults to 2</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if the provided protein have different lengths</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a matplotlib figure containing the mapped representation</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>plt.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_paired_represention">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_paired_represention</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_one_repr</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">protein_two_repr</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">color_first</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'red'</span></em>, <em class="sig-param"><span class="n">color_second</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'blue'</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.9</span></em>, <em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">' Parired protein representation'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_paired_represention" title="Permalink to this definition">¶</a></dt>
<dd><p>visualize the difference in representation for the same protein between two experiments using matplotlib library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_one_repr</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>np.ndarray</em><em>]</em>) – a dict object containing the legand of the first condition along with its mapped array</p></li>
<li><p><strong>protein_two_repr</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>np.ndarray</em><em>]</em>) – a dict object containing the legand of the second condition along with its mapped array</p></li>
<li><p><strong>alpha</strong> (<em>the transparency of the figure.</em>) – the transparency of the figure.</p></li>
</ul>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>color_first: the color of representation for the first condition</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>color_second: the color of the second condition</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>title: the title of the figure.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A matplotlib Figure containing the representation</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>plt.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_parent_protein_expression_in_tissue">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_parent_protein_expression_in_tissue</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">expression_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">ref_expression</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">sampling_num</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'orient': 'v'}</span></em>, <em class="sig-param"><span class="n">def_value</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Normalized Expression'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_parent_protein_expression_in_tissue" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the parent protein expression in tissue relative a sampled collection of non-presented data using seaborn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>expression_table</strong> (<em>pd.DataFrame</em>) – The protein expression table which contains the expresion value for each parent protein</p></li>
<li><p><strong>ref_expression</strong> (<em>pd.DataFrame</em>) – The reference expression of the tissue under investigation.</p></li>
<li><p><strong>tissue_name</strong> (<em>str</em>) – The name of the tissue .</p></li>
<li><p><strong>sampling_num</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of times to sample from the non-prsenter. , defaults to 10</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.violinplot function., defaults to {‘orient’:’v’}</p></li>
<li><p><strong>def_value</strong> (<em>float</em><em>, </em><em>optional</em>) – The default value for proteins that could not be mapped to the expression database , defaults to -1</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis. , defaults to ‘Normalized Expression’</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if the reference gene expression table is smaller than the number of parents</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_peptide_length_dist">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_peptide_length_dist</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_length</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em>, <em class="sig-param"><span class="n">x_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Peptide Length'</span></em>, <em class="sig-param"><span class="n">y_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Frequency'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Peptide Length distribution'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Visualization.vizTools.plot_peptide_length_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize a histogram of the eluted peptide length using seaborn library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep_length</strong> (<em>List</em><em>[</em><em>int</em><em>]</em>) – [description]</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the function seaborn::distplot, defaults to {}</p></li>
<li><p><strong>x_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the x-axis , defaults to ‘Peptide Length’</p></li>
<li><p><strong>y_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the y-axis , defaults to ‘Frequency’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure, defaults to ‘Peptide Length distribution’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_peptide_length_per_experiment">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_peptide_length_per_experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">counts_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">plotting_kwargs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{}</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.pyplot.figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_peptide_length_per_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>visualize the peptide length distribution among the experiments defined in the set</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts_table</strong> (<em>pd.DataFrame</em>) – a pandas dataframe that contain the length of each peptide defined in the experiment along with     the experiment name</p></li>
<li><p><strong>plotting_kwargs</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – a dict object containing parameters for the sns.catplot function, defaults to {}</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_protein_coverage">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_protein_coverage</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapped_protein</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">col</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'r'</span></em>, <em class="sig-param"><span class="n">prot_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_protein_coverage" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the peptide coverage for a given protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mapped_protein</strong> (<em>np.ndarray</em>) – a NumPy array with shape of 1 by protein length or shape protein-length</p></li>
<li><p><strong>col</strong> (<em>str</em><em>, </em><em>optional</em>) – the color of the coverage respresentation, defaults to ‘r’</p></li>
<li><p><strong>prot_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The default protein name, defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>plt.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plot_protein_presentation_3D">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plot_protein_presentation_3D</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proteins</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">plotting_args</span><span class="o">=</span><span class="default_value">{'color': 'red'}</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plot_protein_presentation_3D" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a 3D surface representation for the SAME protein or proteins OF EQAUL LENGTH     in different conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>proteins</strong> (<em>[</em><em>type</em><em>]</em>) – a nested dict object containing for each protein a child dict that contain     the mapping array and the color in the figure.</p></li>
<li><p><strong>plotting_args</strong> (<em>dict</em><em>, </em><em>optional</em>) – a dict that contain further parameter to the plot_surface functions, defaults to {‘color’:’red’}</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure, defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if the provided proteins are of different length</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plt.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_gene_expression_vs_num_peptides">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_gene_expression_vs_num_peptides</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exp_count_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">def_value</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Expression value'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Peptides per protein Vs. Protein Expression Level'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_gene_expression_vs_num_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the correlation between the gene expression and the number of peptids per protein using plotly library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exp_count_table</strong> (<em>pd.DataFrame</em>) – A table that contain the number of peptides and the expresion value for each protein in the database</p></li>
<li><p><strong>tissue_name</strong> (<em>str</em>) – The name of the tissue</p></li>
<li><p><strong>def_value</strong> (<em>float</em><em>, </em><em>optional</em>) – The default value for proteins that could not be mapped to the expression database , defaults to -1</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the y-axis., defaults to ‘Expression value’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Peptides per protein Vs. Protein Expression Level’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_multi_traced_coverage_representation">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_multi_traced_coverage_representation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proteins</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Protein Coverage Across  '</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_multi_traced_coverage_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>plot a multi-traced representation for the same protein accross  using plotly library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>proteins</strong> (<em>[</em><em>type</em><em>]</em>) – A dict object containing for each protein the corresponding mapped array.</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure, defaults to “Protein Coverage Across  “</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a multitraced traced figure showing the coverage of proteins accross different conditions</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_num_peptide_per_go_term">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_num_peptide_per_go_term</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep2goTerm</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'GO-Term'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per GO Term'</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_num_peptide_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of peptides obtained per Go-Term  using plotly library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep2goTerm</strong> (<em>pd.DataFrame</em>) – A table that contain the count of peptides from each GO-Term</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop peptide with unknown GO-term, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis, defaults to ‘GO-Term’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure, defaults to ‘Number of peptides per GO Term’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_num_peptides_per_location">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_num_peptides_per_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep2loc</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Compartment'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per sub-cellular compartment'</span></em><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_num_peptides_per_location" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of peptides obtained from each compartment using plotly library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep2loc</strong> (<em>pd.DataFrame</em>) – A table that contain the count of peptides from each  location</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop protein with unknown location, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the x-axis, defaults to ‘Number of peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the y-axis, defaults to ‘Compartment’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – The title of the figure, defaults to ‘Number of peptides per sub-cellular compartment’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_num_peptides_per_organism">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_num_peptides_per_organism</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_per_org</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">log_scale</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of Peptides'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Organism'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per organism'</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_num_peptides_per_organism" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of peptides per each organism inferred from the experiment using plotly library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep_per_org</strong> (<em>pd.DataFrame</em>) – A table that contain the count of peptides from each organism</p></li>
<li><p><strong>log_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether or not to scale the number of peptide using a log scale, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the x-axis, defaults to ‘Number of Peptides’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label on the y-axis, defaults to ‘Organism’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure , defaults to ‘Number of peptides per organism’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_num_peptides_per_parent">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_num_peptides_per_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nums_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">num_prot</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">x_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides'</span></em>, <em class="sig-param"><span class="n">y_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Protein ID'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of peptides per protein'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_num_peptides_per_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualize a histogram of the the number of peptides per each inferred protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nums_table</strong> (<em>pd.DataFrame</em>) – a pandas dataframe containing number of peptides identified from each protein.</p></li>
<li><p><strong>num_prot</strong> – the number of protein to show relative to the first element, for example, the first 10, 20 etc.     If the default value of -1 is used then all protein will be plotted, however, this might lead to a crowded figure., defaults to -1    :type num_prot: int, optional</p></li>
<li><p><strong>x_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the x-axis , defaults to ‘Number of peptides’</p></li>
<li><p><strong>y_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the y-axis , defaults to ‘Protein ID’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – title, defaults to ‘Number of peptides per protein’</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if the num_prot is bigger than the number of elements in the provided table</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_num_protein_per_go_term">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_num_protein_per_go_term</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein2goTerm</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of Proteins'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'GO-Term'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of proteins per GO-Term'</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_num_protein_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of proteins per each GO Term using plotly library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein2goTerm</strong> (<em>pd.DataFrame</em>) – A table that contain the count of proteins from each GO-Term</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop protein with unknown location, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of Proteins’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis, defaults to ‘GO-Term’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the title of the figure, defaults to ‘Number of proteins per GO-Term’</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>[description]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_num_protein_per_location">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_num_protein_per_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_loc</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">drop_unknown</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of Proteins'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Compartment'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Number of proteins per sub-cellular compartment'</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_num_protein_per_location" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the number of proteins per each sub-cellular compartment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_loc</strong> (<em>pd.DataFrame</em>) – A table that contain the count of protein from each location</p></li>
<li><p><strong>drop_unknown</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to drop protein with unknown location, defaults to False</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the x-axis, defaults to ‘Number of Proteins’</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – the label on the y-axis, defaults to ‘Compartment’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – [description], defaults to ‘Number of proteins per sub-cellular compartment’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_overlap_heatmap">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_overlap_heatmap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">results_df</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_overlap_heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot a user provided dataframe as a heatmap using plotly library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>results_df</strong> (<em>pd.DataFrame</em>) – A pandas dataframe table that hold the overlapping number.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a plotly Figure containing the heatmap</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_paired_representation">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_paired_representation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_one_repr</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">protein_two_repr</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">' Parired protein representation'</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_paired_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>compare the peptide coverage for the same protein under different conditions using the same protein using plotly library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_one_repr</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>np.ndarray</em><em>]</em>) – a dict object containing the legand of the first condition along with its mapped array</p></li>
<li><p><strong>protein_two_repr</strong> (<em>Dict</em><em>[</em><em>str</em><em>, </em><em>np.ndarray</em><em>]</em>) – a dict object containing the legand of the second condition along with its mapped array</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A plotly Figure containing the representation</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_parent_protein_expression_in_tissue">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_parent_protein_expression_in_tissue</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">expression_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">ref_expression</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">sampling_num</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">def_value</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Normalized Expression'</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_parent_protein_expression_in_tissue" title="Permalink to this definition">¶</a></dt>
<dd><p>plot the parent protein expression in tissue relative a sampled collection of non-presented genes using plotly library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>expression_table</strong> (<em>pd.DataFrame</em>) – The protein expression table which contains the expresion value for each parent proteins.</p></li>
<li><p><strong>ref_expression</strong> (<em>pd.DataFrame</em>) – The reference expression of the tissue under investigation.</p></li>
<li><p><strong>tissue_name</strong> (<em>str</em>) – The name of the tissue.</p></li>
<li><p><strong>sampling_num</strong> (<em>int</em><em>, </em><em>optional</em>) – the number of times to sample from the non-prsenter, defaults to 10</p></li>
<li><p><strong>def_value</strong> (<em>float</em><em>, </em><em>optional</em>) – The default value for proteins that could not be mapped to the expression database, defaults to -1</p></li>
<li><p><strong>ylabel</strong> (<em>The label on the y-axis</em><em>, </em><em>optional</em>) – [description], defaults to ‘Normalized Expression’</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the reference gene expression table is smaller than the number of parents</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_peptide_length_dist">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_peptide_length_dist</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_length</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">x_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Peptide Length'</span></em>, <em class="sig-param"><span class="n">y_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Counts'</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Peptide Length distribution'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_peptide_length_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>visualize a histogram of the eluted peptide length using plotly library</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pep_length</strong> (<em>List</em><em>[</em><em>int</em><em>]</em>) – a list of integer containing the peptides’ lengths</p></li>
<li><p><strong>x_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the x-axis , defaults to ‘Peptide Length’</p></li>
<li><p><strong>y_label</strong> (<em>str</em><em>, </em><em>optional</em>) – the label of the y-axis, defaults to ‘Counts’</p></li>
<li><p><strong>title</strong> (<em>str</em><em>, </em><em>optional</em>) – the figure title, defaults to ‘Peptide Length distribution’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="IPTK.Visualization.vizTools.plotly_protein_coverage">
<code class="sig-prename descclassname">IPTK.Visualization.vizTools.</code><code class="sig-name descname">plotly_protein_coverage</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapped_protein</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">prot_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; plotly.graph_objs._figure.Figure<a class="headerlink" href="#IPTK.Visualization.vizTools.plotly_protein_coverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the peptide coverage for a given protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mapped_protein</strong> (<em>np.ndarray</em>) – A NumPy array with shape of 1 by protein length or shape protein-length</p></li>
<li><p><strong>prot_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The default protein name, defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Figure</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-IPTK.Visualization">
<span id="module-contents"></span><h7>Module contents<a class="headerlink" href="#module-IPTK.Visualization" title="Permalink to this headline">¶</a></h7>
</div>
</div>
</div>
</div>
<div class="section" id="module-IPTK">
<span id="module-contents"></span><h5>Module contents<a class="headerlink" href="#module-IPTK" title="Permalink to this headline">¶</a></h5>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">IPTK</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-license">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#contact">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-get_started">Get Started!</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-modules">IPTK</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Hesham ElAbd.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>