
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>IPTK.Classes package &#8212; IPTK 0.3.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="IPTK.IO package" href="IPTK.IO.html" />
    <link rel="prev" title="IPTK.Analysis package" href="IPTK.Analysis.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="iptk-classes-package">
<h1>IPTK.Classes package<a class="headerlink" href="#iptk-classes-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-IPTK.Classes.Annotator">
<span id="iptk-classes-annotator-module"></span><h2>IPTK.Classes.Annotator module<a class="headerlink" href="#module-IPTK.Classes.Annotator" title="Permalink to this headline">¶</a></h2>
<p>The class provided methods for visualizing different aspects of the protein biology. This is achieved through three main methods:</p>
<blockquote>
<div><p>1- add_segmented_track: which visualize information about non-overlapping protein substructures, for example, protein domains.</p>
<p>2- add_stacked_track: which visualize information about overlapping protein substructures, for example, splice variants.</p>
<p>3- add_marked_positions_track: which visualize information about positions in the protein, for example, sequence variance, or PTM.</p>
</div></blockquote>
<p>The class also provided functions for visualizing the relationship between
a protein and its eluted peptide/peptides in an analogous manner to the way
NGS reads are aligned to genomic regions. This can be useful to identify
regions in the protein with high/low number of eluted peptides, also, to
visualize this aspect of the protein along with other facests of the protein
like domain organization, PTM, sequence/splice variants.</p>
<p class="rubric">Notes</p>
<p>each figure should have a base track this can be done explicitly by calling
the function add_base_track or by implicitly by calling the function
add_coverage_plot with the parameter coverage_as_base=True.</p>
<dl class="py class">
<dt id="IPTK.Classes.Annotator.Annotator">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Annotator.</code><code class="sig-name descname">Annotator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_length</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">figure_size</span><span class="p">:</span> <span class="n">Tuple<span class="p">[</span>int<span class="p">, </span>int<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">figure_dpi</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">face_color</span><span class="o">=</span><span class="default_value">'white'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Annotator.Annotator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>higher level API to plot information about the protein, for example, PTM, Splice variant etc, using matplotlib library</p>
<dl class="py method">
<dt id="IPTK.Classes.Annotator.Annotator.add_base_track">
<code class="sig-name descname">add_base_track</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">space_fraction</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.3</span></em>, <em class="sig-param"><span class="n">protein_name_position</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">track_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'base_track'</span></em>, <em class="sig-param"><span class="n">track_label_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 8}</span></em>, <em class="sig-param"><span class="n">protein_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'A protein'</span></em>, <em class="sig-param"><span class="n">protein_name_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 10}</span></em>, <em class="sig-param"><span class="n">rect_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'capstyle': 'butt', 'color': 'olive'}</span></em>, <em class="sig-param"><span class="n">number_ticks</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">xticks_font_size</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">4</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Annotator.Annotator.add_base_track" title="Permalink to this definition">¶</a></dt>
<dd><p>add a base track to the figure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>space_fraction</strong> (<em>float</em><em>, </em><em>optional</em>) – a float between 0 and 1 that represent the fraction of space left
below and above the track. The default is 0.3 which means that the
track will be drown on a 40% while 60% are left as an empty space
below and above the track.</p></li>
<li><p><strong>protein_name_position</strong> (<em>float</em><em>, </em><em>optional</em>) – a float between 0 and 1 which control the relative position of the
protein name on the y-axis. The default is 0.5.</p></li>
<li><p><strong>track_label</strong> (<em>string</em><em>, </em><em>optional</em>) – The name on the track, which will be shown on the y-axis.
The default is “base_track”.</p></li>
<li><p><strong>track_label_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – the parameters that control the printing of the track_label,
for example, the font size and the color. These parameters should
be provided as dict that will be fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.set_ylabel</span></code>.The default is {“fontsize”:8,”color”:”black”}.</p></li>
<li><p><strong>protein_name</strong> (<em>string</em><em>, </em><em>optional</em>) – the name of the protein to be printed to the track.
The default is “A protein”.</p></li>
<li><p><strong>protein_name_dict</strong> (<em>dDict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>ict</em><em>, </em><em>optional</em>) – the parameters that control the printing of the protein name,
for example, the font size and the color. These parameters should
be provided as dict that will be fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.text()</span></code>. The default is {“fontsize”:10,”color”:”black”}.</p></li>
<li><p><strong>rect_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – a dictionary that control the character of the track itself, for example,
the color and the transparency. this dict will be fed to the function
<code class="docutils literal notranslate"><span class="pre">plt.Rectangle()</span></code>.
The default is {“color”:”olive”,”capstyle”:”butt”}.</p></li>
<li><p><strong>number_ticks</strong> (<em>int</em>) – The number of ticks on the x-axis. The default is 10.</p></li>
<li><p><strong>xticks_font_size</strong> (<em>int</em>) – The font size of the x-axis ticks. The default is 4.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">=</span><span class="n">VisTool</span><span class="p">(</span><span class="mi">250</span><span class="p">,(</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span><span class="mi">300</span><span class="p">)</span>
<span class="go">    # create a graph of size 3 inches by 5 inches with a 300 dots per</span>
<span class="go">    # inch (DPI) as a resolution metric for a protein of length 250 amino acids</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_base_track</span><span class="p">()</span>
<span class="go">    # adds a basic track using the default parameters.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_base_track</span><span class="p">(</span><span class="n">space_fraction</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="go">                            track_label=&quot;example_1&quot;,</span>
<span class="go">                            track_label_dict={&quot;fontsize&quot;:5,&quot;color&quot;:&quot;blue&quot;}</span>
<span class="go">                            number_ticks=5,</span>
<span class="go">                            xticks_font_size=6)</span>
<span class="go">    # generate a base track with 10% empty space above and below</span>
<span class="go">    #  the track. Track will have the name example_1 and it will be</span>
<span class="go">    # shown in font 5 instead of 8 and in blue color instead of black.</span>
<span class="go">    # five ticks will be shown on the x-axis using a font of size 6.</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>calling the function more than once will result in an overriding of
the previously added base track, for example, in the examples
section calling add_base_track for the second time will overrides
the graph build by the previous call.</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Annotator.Annotator.add_coverage_track">
<code class="sig-name descname">add_coverage_track</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">coverage_matrix</span><span class="p">:</span> <span class="n">numpy.ndarray</span></em>, <em class="sig-param"><span class="n">coverage_as_base</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">coverage_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'blue', 'width': 1.2}</span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'positions'</span></em>, <em class="sig-param"><span class="n">xlabel_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 6}</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'coverage'</span></em>, <em class="sig-param"><span class="n">ylabel_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 6}</span></em>, <em class="sig-param"><span class="n">number_ticks</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10</span></em>, <em class="sig-param"><span class="n">xticks_font_size</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">4</span></em>, <em class="sig-param"><span class="n">yticks_font_size</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">4</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Annotator.Annotator.add_coverage_track" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a coverage plot to the panel. The coverage plot shows the
relationship between a peptide and its experimentally detected eluted
peptide/peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>coverage_matrix</strong> (<em>np.ndarray</em>) – a protein length by one array which summarize information the protein and
the eluted peptides. The coverage matrix can be computed using the function
<code class="docutils literal notranslate"><span class="pre">AntigenicProtein.compute_protein_coverage_by_peptide</span></code>.</p></li>
<li><p><strong>coverage_as_base</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to plot the coverage as a base track for the figure.
The default is False which means that the track appended to a figure
that have a default base track which can be constructed using the
function <code class="docutils literal notranslate"><span class="pre">add_base_track</span></code>. However, if coverage_as_base  is set to
True, the function will draw the base track using the coverage matrix and
calling the function add_base_track should be avoided.</p></li>
<li><p><strong>coverage_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – The parameters that control the printing of the coverage matrix,
for example, the color. these parameters are fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.bar</span></code>. The default is {“color”:”blue”,”width”:1.2}.</p></li>
<li><p><strong>xlabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label of the x-axis of the coverage track. The default is “positions”.</p></li>
<li><p><strong>xlabel_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – The parameters that control the x-label printing, for example,
the color and/ the font size. these parameters are fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.set_xlabel</span></code>. The default is {“fontsize”:6,”color”:”black”}.</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em>, </em><em>optional</em>) – The label of the y-axis of the coverage track. The default is “coverage”.</p></li>
<li><p><strong>ylabel_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – he parameters that control the x-label printing, for example,
the color and/ the font size. these parameters are fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.set_ylabel</span></code>. The default is {“fontsize”:6,”color”:”black”}.</p></li>
<li><p><strong>number_ticks</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of ticks on the x-axis. The default is 10.</p></li>
<li><p><strong>xticks_font_size</strong> (<em>float</em><em>, </em><em>optional</em>) – The font size of the x-axis ticks. The default is 4.</p></li>
<li><p><strong>yticks_font_size</strong> (<em>float</em><em>, </em><em>optional</em>) – The font size of the y-axis ticks. The default is 4.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Annotator.Annotator.add_marked_positions_track">
<code class="sig-name descname">add_marked_positions_track</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">positions</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">height_frac</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">marker_bar_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'linestyles': 'solid'}</span></em>, <em class="sig-param"><span class="n">marker_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'red', 's': 3}</span></em>, <em class="sig-param"><span class="n">track_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'A marked positions Track'</span></em>, <em class="sig-param"><span class="n">track_label_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 8}</span></em>, <em class="sig-param"><span class="n">base_line_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'linewidth': 1}</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Annotator.Annotator.add_marked_positions_track" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>The function adds a marked position to the track which is shown to highlight
certain amino acid position within the protein, for example, a sequence
variant position, or PTM position.</p>
<dl class="simple">
<dt>positions<span class="classifier">List[int]</span></dt><dd><p>a list that contain the position/positions that should be heighlighted
in the protein sequence.</p>
</dd>
<dt>height_frac<span class="classifier">float</span></dt><dd><p>the relative hight of the marked positions. The default is 0.5 which
means that the hight of the marker will be 50% of the y-axis height.</p>
</dd>
<dt>marker_bar_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><p>The parameters of the marker position bar, for example, line width
or color. These parameters are going to be fed to the function
<code class="docutils literal notranslate"><span class="pre">plt.hlines</span></code>. The default is {“color”:”black”,”linestyles”:”solid”}.</p>
</dd>
<dt>marker_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><p>These parameters of marker point which sits on top of the marker bar,
for example, the color, the shape or the size.
The default is {“color”:”red”,”s”:3}.</p>
</dd>
<dt>track_label<span class="classifier">string, optional</span></dt><dd><p>The name of the track, which will be shown on the y-axis.
The default is “A marked positions Track”.</p>
</dd>
<dt>track_label_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><p>the parameters that control the printing of the track_label,
for example, the font size and the color. These parameters should
be provided as dict that will be fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.set_ylabel</span></code>.The default is {“fontsize”:8,”color”:”black”}.</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>base_line_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><blockquote>
<div><p>the parameters that control the shape of the base line, for example, color and/or line width. These parameters are going to be fed
to the function <code class="docutils literal notranslate"><span class="pre">axes.hlines</span></code>.
The default is {“color”:”black”,”linewidth”:1}.</p>
</div></blockquote>
<p>None.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">test_list</span><span class="o">=</span><span class="p">[</span><span class="mi">24</span><span class="p">,</span><span class="mi">26</span><span class="p">,</span><span class="mi">75</span><span class="p">,</span><span class="mi">124</span><span class="p">,</span><span class="mi">220</span><span class="p">]</span>
<span class="go"># first define a dict object that define some protein features.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">=</span><span class="n">Annotator</span><span class="p">(</span><span class="n">protein_length</span><span class="o">=</span><span class="mi">250</span><span class="p">,</span> <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="n">figure_dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="go"># creating a VisTool instance</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_base_track</span><span class="p">()</span>
<span class="go"># add a base_track</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_marked_positions_track</span><span class="p">(</span><span class="n">test_list</span><span class="p">)</span> <span class="c1"># build a marked position track using the default parameters</span>
<span class="go"># marked positions track</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_marked_positions_track</span><span class="p">(</span><span class="n">positions</span><span class="o">=</span><span class="n">test_list</span><span class="p">,</span><span class="n">height_frac</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span>
<span class="go">                                  track_label=&quot;Post_translational_modifications&quot;,</span>
<span class="go">                                  marker_bar_dict={&quot;color&quot;:&quot;blue&quot;})</span>
<span class="go"># add a second marked position track with the following parameters:</span>
<span class="go">#track name:  Post_translational_modifications</span>
<span class="go">#hight of the maker bar = 75%</span>
<span class="go">#color of the markerbar= blue</span>
</pre></div>
</div>
<p>Any panel can have zero, one or more than one marked-position track.
Thus, in the above examples calling the method <code class="docutils literal notranslate"><span class="pre">add_marked_positions_track</span></code>
for the second time does NOT override the previous marked-position track
it create a new one and added to the figure.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Annotator.Annotator.add_segmented_track">
<code class="sig-name descname">add_segmented_track</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">track_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">track_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'A segmented Track'</span></em>, <em class="sig-param"><span class="n">track_label_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 8}</span></em>, <em class="sig-param"><span class="n">track_element_names_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 8}</span></em>, <em class="sig-param"><span class="n">center_line_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">, </span>float<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'alpha': 0.5, 'linewidth': 0.5}</span></em>, <em class="sig-param"><span class="n">track_elements_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'capstyle': 'butt', 'color': 'brown'}</span></em>, <em class="sig-param"><span class="n">show_names</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Annotator.Annotator.add_segmented_track" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a segmentation track which show non-overlapping features of the protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>track_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>]</em>) – <p>a dict that contain the non-overlapping features of the protein.
The dict is assumed to have the following structure: a dict with the
feature_index as a key and associated features as values. The associated
features is a dict with the following three keys:</p>
<blockquote>
<div><p>1- Name: which contain the feature name</p>
<p>2- startIdx: which contain the start position of the protein</p>
<p>3- endIdx: which contain the end position of the protein</p>
</div></blockquote>
</p></li>
<li><p><strong>track_label</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the track, which will be shown on the y-axis.
The default is “A segmented Track”.</p></li>
<li><p><strong>track_label_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – the parameters that control the printing of the track_label,
for example, the font size and the color. These parameters should
be provided as dict that will be fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.set_ylabel</span></code>. The default is {“fontsize”:8,”color”:”black”}.</p></li>
<li><p><strong>track_element_names_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – the parameters that control the printing of the feature names on the track,
for example, the font size and the color. These parameters should
be provided as a dict that will be fed to the function <code class="docutils literal notranslate"><span class="pre">axes.text</span></code>.
The default is {“fontsize”:8,”color”:”black”}.</p></li>
<li><p><strong>center_line_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>, </em><em>float</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – The parameters that control the printing of the center line of a segmented track object.
The default is {“fontsize”:8,”color”:”black”}.</p></li>
<li><p><strong>track_elements_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>, </em><em>optional</em>) – the parameters that control the printing of the feature
rectangluar representation for example the color, the dict will be fed
to the function <code class="docutils literal notranslate"><span class="pre">plt.Rectangle</span></code>.
The default is {“color”:”brown”,”capstyle”:”butt”}.</p></li>
<li><p><strong>show_names</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether or not to show the name of the features.The default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">test_dict</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;domain1&quot;</span><span class="p">:{</span><span class="s2">&quot;Name&quot;</span><span class="p">:</span><span class="s2">&quot;domain_one&quot;</span><span class="p">,</span><span class="s2">&quot;startIdx&quot;</span><span class="p">:</span><span class="mi">55</span><span class="p">,</span><span class="s2">&quot;endIdx&quot;</span><span class="p">:</span><span class="mi">150</span><span class="p">},</span>
<span class="go">               &quot;domain2&quot;:{&quot;Name&quot;:&quot;domain_Two&quot;,&quot;startIdx&quot;:190,&quot;endIdx&quot;:225}}</span>
<span class="go"># first define a dict object that define some protein features.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">=</span><span class="n">Annotator</span><span class="p">(</span><span class="n">protein_length</span><span class="o">=</span><span class="mi">250</span><span class="p">,</span> <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="n">figure_dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="go"># creating a Annotator instance</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_base_track</span><span class="p">()</span>
<span class="go"># add a base_track</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_segmented_track</span><span class="p">(</span><span class="n">test_dict</span><span class="p">)</span> <span class="c1"># build a segmented track using the default parameters</span>
<span class="go"># add the segmented track</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_segmented_track</span><span class="p">(</span><span class="n">track_dict</span><span class="o">=</span><span class="n">test_dict</span><span class="p">,</span>
<span class="go">                                  track_label=&quot;Domains&quot;,</span>
<span class="go">                                  track_elements_dict={&quot;color&quot;:&quot;brown&quot;})</span>
<span class="go"># add a second segmented track with track name set to Domains and elements</span>
<span class="go"># of the track shown as brown rectangles.</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>Any panel can have one or more segmented-tracks. Thus, in the above examples
calling the method <code class="docutils literal notranslate"><span class="pre">add_segmented_track</span></code> for the second time does NOT
override the previous segmented track it create a new one and added to the
figure.</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Annotator.Annotator.add_stacked_track">
<code class="sig-name descname">add_stacked_track</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">track_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">track_label</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'A stacked Track'</span></em>, <em class="sig-param"><span class="n">track_label_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 8}</span></em>, <em class="sig-param"><span class="n">track_element_names_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'fontsize': 8}</span></em>, <em class="sig-param"><span class="n">track_elements_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'capstyle': 'butt', 'color': 'magenta'}</span></em>, <em class="sig-param"><span class="n">base_line_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'color': 'black', 'linewidth': 1}</span></em>, <em class="sig-param"><span class="n">show_names</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Annotator.Annotator.add_stacked_track" title="Permalink to this definition">¶</a></dt>
<dd><p>The function add a stacked_track to a visualization panel.
The stacked track is used to show overlapping protein features, for example,
different splice variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>track_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>]</em>) – <p>a dict that contain the overlapping features of the protein.
The dict is assumed to have the following structure, a dict with the
feature_index as a key and associated features as values. The associated
features is a dict with the following three keys:</p>
<blockquote>
<div><p>1- Name: which contain the feature name</p>
<p>2- startIdx: which contain the start position of the feature.</p>
<p>3- endIdx: which contain the end position of the feature.</p>
</div></blockquote>
</p>
</dd>
</dl>
<dl class="simple">
<dt>track_label<span class="classifier">str, optional</span></dt><dd><p>The name of the track, which will be shown on the y-axis.
The default is “A stacked Track”.</p>
</dd>
<dt>track_label_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><p>the parameters that control the printing of the track_label,
for example, the font size and the color. These parameters should
be provided as dict that will be fed to the function
<code class="docutils literal notranslate"><span class="pre">axes.set_ylabel</span></code>.The default is {“fontsize”:8,”color”:”black”}.</p>
</dd>
<dt>track_element_names_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><p>the parameters that control the printing of the feature names on the track,
for example, the font size and the color. These parameters should
be provided as a dict that will be fed to the function <code class="docutils literal notranslate"><span class="pre">axes.text</span></code>.
The default is {“fontsize”:8,”color”:”black”}.</p>
</dd>
<dt>track_elements_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><p>the parameters that control the printing of the feature
rectangluar representation for example the color, the dict will be fed
to the function <code class="docutils literal notranslate"><span class="pre">plt.Rectangle</span></code>.
The default is {“color”:”magenta”,”capstyle”:”butt”}.</p>
</dd>
<dt>base_line_dict<span class="classifier">Dict[str,Union[int,str]], optional</span></dt><dd><p>the parameters that control the shape of the shape of the base line,
for example, color and/or line width. These parameters are going to be fed
to the function <code class="docutils literal notranslate"><span class="pre">axes.hlines</span></code>.
The default is {“color”:”black”,”linewidth”:1}.</p>
</dd>
<dt>show_names<span class="classifier">bool, optional</span></dt><dd><p>whether or not to show the name of the features. The default is True.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">test_dict</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;feature_1&quot;</span><span class="p">:{</span><span class="s2">&quot;Name&quot;</span><span class="p">:</span><span class="s2">&quot;X&quot;</span><span class="p">,</span><span class="s2">&quot;startIdx&quot;</span><span class="p">:</span><span class="mi">55</span><span class="p">,</span><span class="s2">&quot;endIdx&quot;</span><span class="p">:</span><span class="mi">150</span><span class="p">},</span>
<span class="go">               &quot;feature_2&quot;:{&quot;Name&quot;:&quot;Y&quot;,&quot;startIdx&quot;:85,&quot;endIdx&quot;:225},</span>
<span class="go">               &quot;feature_3&quot;:{&quot;Name&quot;:&quot;Z&quot;,&quot;startIdx&quot;:160,&quot;endIdx&quot;:240}}</span>
<span class="go"> # first define a dict object that define some protein features.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">=</span><span class="n">Annotator</span><span class="p">(</span><span class="n">protein_length</span><span class="o">=</span><span class="mi">250</span><span class="p">,</span> <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="n">figure_dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="go"># creating a Annotator instance</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_base_track</span><span class="p">()</span>
<span class="go"># add a base_track</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_segmented_track</span><span class="p">(</span><span class="n">test_dict</span><span class="p">)</span> <span class="c1"># build a stacked track using the default parameters.</span>
<span class="go"># add the stacked track</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">example_1</span><span class="o">.</span><span class="n">add_segmented_track</span><span class="p">(</span><span class="n">track_dict</span><span class="o">=</span><span class="n">test_dict</span><span class="p">,</span>
<span class="go">                                  track_label=&quot;OverLappingFeat&quot;,</span>
<span class="go">                                  track_elements_dict={&quot;color&quot;:&quot;red&quot;})</span>
<span class="go"># add a second segmented track with track name set to OverLappingFeat and elements</span>
<span class="go"># of the track shown as red rectangles.</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>Any panel can have zero, one or more than one stacked-track.
Thus, in the above examples calling the method <code class="docutils literal notranslate"><span class="pre">add_stacked_track</span></code>
for the second time does NOT override the previous stacked track
it create a new one and added to the figure.</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Annotator.Annotator.get_figure">
<code class="sig-name descname">get_figure</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; matplotlib.figure.Figure<a class="headerlink" href="#IPTK.Classes.Annotator.Annotator.get_figure" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The figure with all the tracks that have been added to it.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Annotator.Annotator.save_fig">
<code class="sig-name descname">save_fig</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">output_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'.'</span></em>, <em class="sig-param"><span class="n">format_</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'png'</span></em>, <em class="sig-param"><span class="n">figure_dpi</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'same'</span></em>, <em class="sig-param"><span class="n">figure_saving_dict</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">{'facecolor': 'white'}</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Annotator.Annotator.save_fig" title="Permalink to this definition">¶</a></dt>
<dd><p>write the construct fig to the hard-disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – The name of the figure to save the file.</p></li>
<li><p><strong>output_path</strong> (<em>str</em><em> , </em><em>optional</em>) – The path to write the output, by default the function write to the
current working directory.</p></li>
<li><p><strong>format</strong> (<em>str</em><em>, </em><em>optional</em>) – The output format, this parameter will be fed to the method
<code class="docutils literal notranslate"><span class="pre">plt.savefig</span></code>. The default is “png”.</p></li>
<li><p><strong>figure_dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – The dpi of the saved figure. The deafult is same which mean the figure
will be saved using the same dpi used for creating the figure.</p></li>
<li><p><strong>figure_saving_dict</strong> (<em>Dict</em><em>[</em><em>str</em><em>,</em><em>Union</em><em>[</em><em>int</em><em>,</em><em>str</em><em>]</em><em>]</em><em>,</em><em>optional</em>) – The parameters that should be fed to the function <code class="docutils literal notranslate"><span class="pre">plt.savefig</span></code>.
The default is figure_saving_dict={“facecolor”:”white”}</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Database">
<span id="iptk-classes-database-module"></span><h2>IPTK.Classes.Database module<a class="headerlink" href="#module-IPTK.Classes.Database" title="Permalink to this headline">¶</a></h2>
<p>This submodule define a collection of container classes that are used through the library</p>
<dl class="py class">
<dt id="IPTK.Classes.Database.CellularLocationDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">CellularLocationDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2data</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'https://www.proteinatlas.org/download/subcellular_location.tsv.zip'</span></em>, <em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The class provides an API to access the cellular location information from a database   the follow the structure of the human Proteome Atlas sub-cellular location database. See <a class="reference external" href="https://www.proteinatlas.org/about/download">https://www.proteinatlas.org/about/download</a>    for more details.</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.add_to_database">
<code class="sig-name descname">add_to_database</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genes_to_add</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a></span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.add_to_database" title="Permalink to this definition">¶</a></dt>
<dd><p>add the the location of more proteins to the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>genes_to_add</strong> (<a class="reference internal" href="#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB"><em>CellularLocationDB</em></a>) – a CellularLocationDB instance containing the genes that shall be added to the database.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if the genes to add to the database are already defined in the database</p></li>
<li><p><strong>RuntimeError</strong> – Incase any other error has been encountered while merging the tables.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_approved_location">
<code class="sig-name descname">get_approved_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_approved_location" title="Permalink to this definition">¶</a></dt>
<dd><p>return the location of the provided gene id or gene name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the approved location where the protein the corresponds to the provided name or id is located.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_gene_names">
<code class="sig-name descname">get_gene_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_gene_names" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of all gene names in the dataset</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the names of all genes in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_genes">
<code class="sig-name descname">get_genes</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of all gene ids in the dataset</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>all genes ids currently defined in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_go_names">
<code class="sig-name descname">get_go_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_go_names" title="Permalink to this definition">¶</a></dt>
<dd><p>return the location of the provided gene id or gene name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the gene ontology, GO,  location where the protein the corresponds to the provided name or id is located.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_main_location">
<code class="sig-name descname">get_main_location</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">corresponds</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_main_location" title="Permalink to this definition">¶</a></dt>
<dd><p>return the main location(s) of the provided gene id or gene name. If both gene Id and gene name are provided, both gene_id has a higher precedence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the main location where the protein the corresponds to the provided name or id is located.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.CellularLocationDB.get_table">
<code class="sig-name descname">get_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.CellularLocationDB.get_table" title="Permalink to this definition">¶</a></dt>
<dd><p>return the instance table</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the location table of the instance.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Database.GeneExpressionDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">GeneExpressionDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2data</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'https://www.proteinatlas.org/download/rna_tissue_consensus.tsv.zip'</span></em>, <em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'\t'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>provides an API to access gene expression data stored in table that follows the same structure as the Human proteome Atlas Normalized RNA Expression see  <a class="reference external" href="https://www.proteinatlas.org/about/download">https://www.proteinatlas.org/about/download</a> for more details</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_expression">
<code class="sig-name descname">get_expression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a table summarizing the expression of the provided gene name or gene id accross different tissues.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id</strong> (<em>str</em><em>, </em><em>optional</em>) – the id of the gene of interest , defaults to None</p></li>
<li><p><strong>gene_name</strong> (<em>[</em><em>type</em><em>]</em><em>, </em><em>optional</em>) – the name of the gene of interest , defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if both gene_id and gene_name are None</p></li>
<li><p><strong>KeyError</strong> – if gene_id is None and gene_name is not in the database</p></li>
<li><p><strong>KeyError</strong> – if gene_name is None and gene_id is not in the database</p></li>
<li><p><strong>RuntimeError</strong> – incase some error was encountered while running retriving the elements from the database</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table summarizing the expression of the provided gene accross all tissues in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_expression_in_tissue">
<code class="sig-name descname">get_expression_in_tissue</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tissue_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_expression_in_tissue" title="Permalink to this definition">¶</a></dt>
<dd><p>return the expression profile of the provided tissue</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tissue_name</strong> (<em>str</em>) – the name of the tissue</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>KeyError</strong> – incase the provided tissue is not provided in the database</p></li>
<li><p><strong>RuntimeError</strong> – in case any error was encountered while generating the expression profile.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a table summarizing the expression of all genes in the provided tissue.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_gene_names">
<code class="sig-name descname">get_gene_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_gene_names" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of the UNIQUE gene names currently in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of the UNIQUE gene names currently in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_genes">
<code class="sig-name descname">get_genes</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of the UNIQUE gene ids currently in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of the UNIQUE gene ids currently in the database</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_table">
<code class="sig-name descname">get_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_table" title="Permalink to this definition">¶</a></dt>
<dd><p>return a table containing the expression value of all the genes accross all tissues in the current instance</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The expression of all genes accross all tissues in the database.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.GeneExpressionDB.get_tissues">
<code class="sig-name descname">get_tissues</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.GeneExpressionDB.get_tissues" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of the tissues in the current database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list containing the names of the UNIQUE tissues in the database.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Database.OrganismDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">OrganismDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2Fasta</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.OrganismDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Extract information about the source organsim of a collection of protein sequences from a fasta file and provides an API to query the results.     The function expect the input fasta file to have header written in the UNIPROT format.</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.OrganismDB.get_number_protein_per_organism">
<code class="sig-name descname">get_number_protein_per_organism</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Database.OrganismDB.get_number_protein_per_organism" title="Permalink to this definition">¶</a></dt>
<dd><p>provides a table containing the number of proteins per organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a table containing the number of proteins per organism</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.OrganismDB.get_org">
<code class="sig-name descname">get_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Database.OrganismDB.get_org" title="Permalink to this definition">¶</a></dt>
<dd><p>return the parent organism of the provided proteins</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prot_id</strong> (<em>str</em>) – the id of the protein of interest</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – incase the provided identifier is not in the database</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the parent organism, i.e. the source organism.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.OrganismDB.get_unique_orgs">
<code class="sig-name descname">get_unique_orgs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Database.OrganismDB.get_unique_orgs" title="Permalink to this definition">¶</a></dt>
<dd><p>get the number of unique organisms in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of all unique organisms in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Database.SeqDB">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Database.</code><code class="sig-name descname">SeqDB</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path2fasta</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Database.SeqDB" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>load a fasta file and constructs a lock up dictionary where sequence ids are  keys and sequences are values.</p>
<dl class="py method">
<dt id="IPTK.Classes.Database.SeqDB.get_seq">
<code class="sig-name descname">get_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Database.SeqDB.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the corresponding sequence if the provided protein-id is defined in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_id</strong> (<em>str</em>) – The protein id to retrive its sequence.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – If the provided protein does not exist in the database</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the protein sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Database.SeqDB.has_sequence">
<code class="sig-name descname">has_sequence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Database.SeqDB.has_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>check if the provided sequence id is an element of the database or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence_name</strong> (<em>str</em>) – The id of the sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the database has this id, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Experiment">
<span id="iptk-classes-experiment-module"></span><h2>IPTK.Classes.Experiment module<a class="headerlink" href="#module-IPTK.Classes.Experiment" title="Permalink to this headline">¶</a></h2>
<p>This module provides an abstraction for an IP experiment.</p>
<dl class="py class">
<dt id="IPTK.Classes.Experiment.Experiment">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Experiment.</code><code class="sig-name descname">Experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proband</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.Proband.Proband" title="IPTK.Classes.Proband.Proband">IPTK.Classes.Proband.Proband</a></span></em>, <em class="sig-param"><span class="n">hla_set</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.HLASet.HLASet" title="IPTK.Classes.HLASet.HLASet">IPTK.Classes.HLASet.HLASet</a></span></em>, <em class="sig-param"><span class="n">tissue</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.Tissue.Tissue" title="IPTK.Classes.Tissue.Tissue">IPTK.Classes.Tissue.Tissue</a></span></em>, <em class="sig-param"><span class="n">database</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.Database.SeqDB" title="IPTK.Classes.Database.SeqDB">IPTK.Classes.Database.SeqDB</a></span></em>, <em class="sig-param"><span class="n">ident_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A representation of an immunopeptidomic experiment.</p>
<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.add_org_info">
<code class="sig-name descname">add_org_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot2org</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>str<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.add_org_info" title="Permalink to this definition">¶</a></dt>
<dd><p>annotated the inferred proteins with their source organism</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prot2org</strong> (<em>ProteinSource</em>) – a dict that contain the protein id as keys and its source organism as values           and add this info to each protein inferred in the current experiment.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>RuntimeWarning</strong> – If the provided dictionary does cover all proteins in the experimental object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.annotate_proteins">
<code class="sig-name descname">annotate_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">organisms_db</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.Database.OrganismDB" title="IPTK.Classes.Database.OrganismDB">IPTK.Classes.Database.OrganismDB</a></span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.annotate_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the parent organisms of each protein in the experiment from an organism database instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>organisms_db</strong> (<a class="reference internal" href="#IPTK.Classes.Database.OrganismDB" title="IPTK.Classes.Database.OrganismDB"><em>OrganismDB</em></a>) – an OrgansimDB instance that will be used to annotate the proteins          identified in the experiment.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.drop_peptide_belong_to_org">
<code class="sig-name descname">drop_peptide_belong_to_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.drop_peptide_belong_to_org" title="Permalink to this definition">¶</a></dt>
<dd><p>Drop the all the peptides that belong to a user provided organism.              Note that, this function will IRREVERSIBLY remove the peptide from the experimental object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>org</strong> (<em>str</em>) – the organims name</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_binarized_results">
<code class="sig-name descname">get_binarized_results</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_binarized_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of NumPy arrays where each array represents a child peptide, parent protein mapped pair.
Please note that, The function treat each peptide-protein pair individually, that is if two peptides originating from the same protein,
it treat them independently and the same protein will be represented twice with the two different peptides. Incase an integrative mapping is needed,
the function &#64;get_integrated_binarized_results&#64; shall be used.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of NumPy arrays containing the mapping between each peptide protein pair.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>MappedProtein</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_c_terminal_flanked_seqs">
<code class="sig-name descname">get_c_terminal_flanked_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_c_terminal_flanked_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>return the c-terminal flanking sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_length</strong> (<em>int</em>) – the length of the peptide downstream of the C-terminal of the peptide</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list sequences contain the N-terminal flanking sequence for each peptide in the instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_experiment_reference_tissue_expression">
<code class="sig-name descname">get_experiment_reference_tissue_expression</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_experiment_reference_tissue_expression" title="Permalink to this definition">¶</a></dt>
<dd><p>return the reference gene expression for the current tissue</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that contain the expression value for ALL the genes in the instance Tissue</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_expression_of_parent_proteins">
<code class="sig-name descname">get_expression_of_parent_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">non_mapped_dval</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">- 1</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_expression_of_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>return a table containing the expression of the proteins inferred in the current experiment from the current tissue.
This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>non_mapped_dval</strong> (<em>float</em><em>, </em><em>optional</em>) – A default value to be added incase the parent protein is not define in the expression database, defaults to -1</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a table that contain the expression of the protein inferred in the database</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_flanked_peptides">
<code class="sig-name descname">get_flanked_peptides</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_flanked_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a list of sequences containing the peptides identified in the experiment padded with
the flanking regions from all the parents of each peptide.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_length</strong> (<em>int</em>) – the length of the flanking region</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of the peptides + the flanking region.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_go_location_id_parent_proteins">
<code class="sig-name descname">get_go_location_id_parent_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">not_mapped_val</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'UNK'</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_go_location_id_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the gene ontology,GO, location terms for all the identified proteins.</p>
<p>&#64;brief:
&#64;param: not_mapped_val: the default value to return incase the GO term of the protein can not be extracted.
&#64;note: This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>not_mapped_val</strong> (<em>str</em><em>, </em><em>optional</em>) – The default value to return incase the GO term of the protein can not be extracted, defaults to ‘UNK’</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A table that contain the GO-location term for each protein in the current instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_hla_allele">
<code class="sig-name descname">get_hla_allele</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_hla_allele" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the set of HLA alleles from which the instance peptides have been eluted</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_hla_class">
<code class="sig-name descname">get_hla_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_hla_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the HLA class</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_main_sub_cellular_location_of_parent_proteins">
<code class="sig-name descname">get_main_sub_cellular_location_of_parent_proteins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">not_mapped_val</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'UNK'</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_main_sub_cellular_location_of_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the main cellular location for the identified proteins.
This method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>not_mapped_val</strong> (<em>str</em><em>, </em><em>optional</em>) – The default value to return incase the location of a protein can not be extracted, defaults to ‘UNK’</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A table that contain the main cellular compartment for each protein in the current instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_mapped_protein">
<code class="sig-name descname">get_mapped_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_mapped_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>return an NumPy array of shape 1 x protein length where each number in the array represents
the total number of peptides identified in the experiment that have originated from the said position
in the protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – the protein id</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided protein id was inferred from the current experiment</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a NumPy array that contain the coverage of the protein.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_mapped_proteins">
<code class="sig-name descname">get_mapped_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>numpy.ndarray<span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_mapped_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>return a dictionary of all the proteins identified in the current experiment with all inferred
peptides mapped to them.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a dictionary that contain the mapped proteins for all the proteins in the current instance.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>MappedProteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_mono_parent_peptides">
<code class="sig-name descname">get_mono_parent_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_mono_parent_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of peptides that have only one parent protein</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of peptide instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_n_terminal_flanked_seqs">
<code class="sig-name descname">get_n_terminal_flanked_seqs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_n_terminal_flanked_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>return the n-terminal flanking sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_length</strong> (<em>int</em>) – the length of the flanking region upstream of the N-terminal of the peptide</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list sequences contain the N-terminal flanking sequence for each peptide in the instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_negative_example">
<code class="sig-name descname">get_negative_example</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fold</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">2</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_negative_example" title="Permalink to this definition">¶</a></dt>
<dd><p>generate negative examples, i.e., non-bounding peptides from the proteins identified in the current experiment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fold</strong> (<em>int</em><em>, </em><em>optional</em>) – the number of negative example to generate relative to the number of unique identified peptides, defaults to 2</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of non-presented peptides from all inferred proteins.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_num_peptide_expression_table">
<code class="sig-name descname">get_num_peptide_expression_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_num_peptide_expression_table" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Get a table that contain the id of all parent proteins, number of peptide per-proteins and the expression value</dt><dd><p>of these parent transcripts. Please note, this method need internet connection as it need to access uniprot mapping API to map uniprot IDs to gene IDs.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of peptides per protein table</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_num_peptide_per_go_term">
<code class="sig-name descname">get_num_peptide_per_go_term</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_num_peptide_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the number of peptides per each GO-Term
:return: A table that has two columns, namely, GO-Terms and Counts.
:rtype: pd.DataFrame</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_num_peptide_per_location">
<code class="sig-name descname">get_num_peptide_per_location</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_num_peptide_per_location" title="Permalink to this definition">¶</a></dt>
<dd><p>retrun the number of peptides obtained from proteins localized to different sub-cellular compartments</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that has two columns, namely, Compartment and Counts.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_number_of_children">
<code class="sig-name descname">get_number_of_children</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_number_of_children" title="Permalink to this definition">¶</a></dt>
<dd><p>return the number of children, i.e. number of peptides belonging to a parent protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – the id of the parent protein</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the number of peptides</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_compartment">
<code class="sig-name descname">get_number_of_proteins_per_compartment</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_compartment" title="Permalink to this definition">¶</a></dt>
<dd><p>get the number of proteins from each compartment</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that has two columns, namely, Compartment and Counts.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_go_term">
<code class="sig-name descname">get_number_of_proteins_per_go_term</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_number_of_proteins_per_go_term" title="Permalink to this definition">¶</a></dt>
<dd><p>get the number of proteins from each GO-Term</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that has two columns, namely, GO-Terms and Counts.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_orgs">
<code class="sig-name descname">get_orgs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_orgs" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list containing the UNIQUE organisms identified in the current experiment</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of all UNIQUE organisms inferred from the inferred proteins.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptide">
<code class="sig-name descname">get_peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_seq</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>return a peptide instance corresponding to the user provided peptide sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pep_seq</strong> (<em>str</em>) – the peptide sequence</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the peptide sequence has not been inferred from the current database.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the peptide instance with the coresponding sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">Peptide</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptide_number_parent">
<code class="sig-name descname">get_peptide_number_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptide_number_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>return a pandas dataframe with the peptide sequence in the first columns and the
number of parent proteins in the second column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ascending</strong> (<em>bool</em><em>, </em><em>optional</em>) – ascending sort the peptide by their number of parent proteins, defaults to False</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the number of parents for each peptide</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides">
<code class="sig-name descname">get_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set of all the peptide stored in the experimental object</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides_length">
<code class="sig-name descname">get_peptides_length</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides_length" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list containing the length of each unique peptide in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>peptides’ lengths</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides_per_organism">
<code class="sig-name descname">get_peptides_per_organism</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides_per_organism" title="Permalink to this definition">¶</a></dt>
<dd><p>return a pandas dataframe that contain the count of peptides belonging to each organism in
the database</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a table with two columns, namely, Organisms and Counts</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_peptides_per_protein">
<code class="sig-name descname">get_peptides_per_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_peptides_per_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>return a pandas dataframe that contain the number of peptides belonging to each protein
inferred in the experiment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ascending</strong> (<em>bool</em><em>, </em><em>optional</em>) – ascending sort the proteins by their number of parent number of child peptides, defaults to False</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a table with the following columns, Proteins and Number_of_Peptides</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_poly_parental_peptides">
<code class="sig-name descname">get_poly_parental_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_poly_parental_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>return a list of peptides that have more than one parent proteins
:return: [list of peptide instance
:rtype: Peptides</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_proband_name">
<code class="sig-name descname">get_proband_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_proband_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the proband name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_proteins">
<code class="sig-name descname">get_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Protein.Protein" title="IPTK.Classes.Protein.Protein">IPTK.Classes.Protein.Protein</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_proteins" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set of all the proteins in the experimental object</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Proteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_tissue">
<code class="sig-name descname">get_tissue</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#IPTK.Classes.Tissue.Tissue" title="IPTK.Classes.Tissue.Tissue">IPTK.Classes.Tissue.Tissue</a><a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_tissue" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the tissue of the current experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#IPTK.Classes.Tissue.Tissue" title="IPTK.Classes.Tissue.Tissue">Tissue</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.get_tissue_name">
<code class="sig-name descname">get_tissue_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.get_tissue_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the tissue name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_allele_group">
<code class="sig-name descname">has_allele_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain peptides eluted from an HLA-alleles belonging to the provided allele group or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_group</strong> (<em>str</em>) – the gene group to query the hla_set against</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the gene group has a member that is a member of the instance HLASet and False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_gene">
<code class="sig-name descname">has_gene</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">locus</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain peptides eluted from an HLA-alleles belonging to the provided locus or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>locus</strong> (<em>str</em>) – the locus of the allele to query the hla_set against</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the locus has a member that is a member of the instance HLASet and False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_hla_allele">
<code class="sig-name descname">has_hla_allele</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">individual</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_hla_allele" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain an eluted peptides from the provided alleles</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>individual</strong> (<em>str</em>) – is the name of the allele as a string</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the allele is a member of the instance HLASet and False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.has_protein_group">
<code class="sig-name descname">has_protein_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.has_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><p>return whether or not the experiment contain peptides eluted from an HLA-alleles belonging to the provided protein group or not</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_group</strong> (<em>str</em>) – The protein group to query the hla_set against</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the locus has a member that is a member of the instance HLASet and False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.is_a_parent_protein">
<code class="sig-name descname">is_a_parent_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.is_a_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein</strong> – check if the protein is a member of the instance proteins or not.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the protein has been identified in the current instance, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Experiment.Experiment.is_member">
<code class="sig-name descname">is_member</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptide</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Experiment.Experiment.is_member" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – check if the peptide is a member of the instance peptides or not.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if the peptide has been identified in the current instance, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.ExperimentalSet">
<span id="iptk-classes-experimentalset-module"></span><h2>IPTK.Classes.ExperimentalSet module<a class="headerlink" href="#module-IPTK.Classes.ExperimentalSet" title="Permalink to this headline">¶</a></h2>
<p>An Experimental set which is a collection of experiments.
The class provides an API for integrating and comparing different experimental instances.</p>
<dl class="py class">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.ExperimentalSet.</code><code class="sig-name descname">ExperimentSet</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">exp_id_pair</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>an API for integrating and comparing different experimental instances</p>
<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.add_experiment">
<code class="sig-name descname">add_experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">exp_id_pair</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.add_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>add an arbitrary number of experiments to the set</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compare_org_count_among_exps">
<code class="sig-name descname">compare_org_count_among_exps</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">abs_count</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compare_org_count_among_exps" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>org</strong> (<em>str</em>) – The name of the organism to query the database for it.</p></li>
<li><p><strong>abs_count</strong> (<em>bool</em><em>, </em><em>optional</em>) – The absolute count, defaults to False</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the count of the peptides that belong to a specific organism in the database.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compare_peptide_counts">
<code class="sig-name descname">compare_peptide_counts</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compare_peptide_counts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that contain the total number of peptides and per-organism peptide counts         among all experiments in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_average_distance_between_exps">
<code class="sig-name descname">compute_average_distance_between_exps</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_average_distance_between_exps" title="Permalink to this definition">¶</a></dt>
<dd><p>compute the average distance between experiments by taking the average over the z-axis
of the 3D tensor computed by the function compute_change_in_protein_representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A 2D tensor with shape of (num-experiments, num-experiments)</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_change_in_protein_representation">
<code class="sig-name descname">compute_change_in_protein_representation</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_change_in_protein_representation" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the change in protein representation among the proteins that are presented/ detect in all of the         instance’s experiments.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a 3D tensor, T, with shape of (num-experiments, num-experiments, num-proteins),         where T[i,j,k] is a the difference between experiment i &amp; j with respect to the k th protein         :rtype: np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_correlation_in_experssion">
<code class="sig-name descname">compute_correlation_in_experssion</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_correlation_in_experssion" title="Permalink to this definition">¶</a></dt>
<dd><p>compute the correlation in parent protein gene-expression across all the experiments
in the set. See the function <strong>compute_binary_correlation</strong> in the analysis module
for information about the computational logic.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a 2D matrix containing the coorelation in gene expression between each pair of experiments inside the current instance collection of experiments.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_length_table">
<code class="sig-name descname">compute_peptide_length_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_length_table" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A table that contain the length of each peptide in the experiment</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_overlap_matrix">
<code class="sig-name descname">compute_peptide_overlap_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_overlap_matrix" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a 2D matrix containing the number of peptide overlapping between each pair of experiments inside the current instance collection of experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_representation_count">
<code class="sig-name descname">compute_peptide_representation_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_peptide_representation_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of times a peptide was observed accross  experiments in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_coverage_over_the_set">
<code class="sig-name descname">compute_protein_coverage_over_the_set</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_coverage_over_the_set" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the mapped representation for each protein in accross the entire  set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_overlap_matrix">
<code class="sig-name descname">compute_protein_overlap_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_overlap_matrix" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a 2D matrix containing the number of proteins overlapping between each pair of experiments inside the current instance collection of experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_representation_count">
<code class="sig-name descname">compute_protein_representation_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.compute_protein_representation_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of times a protein was observed accross the experiment in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.drop_peptides_belong_to_org">
<code class="sig-name descname">drop_peptides_belong_to_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.drop_peptides_belong_to_org" title="Permalink to this definition">¶</a></dt>
<dd><p>drop all the peptides that belong to the provided organisms from all experiments in the set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>org_name</strong> (<em>str</em>) – the name of the organism to drop</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_allele_count">
<code class="sig-name descname">get_allele_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_allele_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of experiments obtained from each allele in the instance.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiment">
<code class="sig-name descname">get_experiment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exp_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">IPTK.Classes.Experiment.Experiment</a><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>return the experiment pointed to by the provided experimental name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>exp_name</strong> (<em>str</em>) – the name of the experiment</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided experimental name is not in the dataset.</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the experiment corresponds to the provided name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">Experiment</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_experimental_names">
<code class="sig-name descname">get_experimental_names</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_experimental_names" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list with all the identifiers of the experiments in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Names</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiments">
<code class="sig-name descname">get_experiments</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>Experiment<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_experiments" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a dict with all the experiments stored in the instance as value of ids as keys.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[<a class="reference internal" href="#IPTK.Classes.Experiment.Experiment" title="IPTK.Classes.Experiment.Experiment">Experiment</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_num_experiments_in_the_set">
<code class="sig-name descname">get_num_experiments_in_the_set</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_num_experiments_in_the_set" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of experiments currently in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_peptides_present_in_all">
<code class="sig-name descname">get_peptides_present_in_all</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_peptides_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the peptides that are observed in every experiments in the set.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_proband_count">
<code class="sig-name descname">get_proband_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_proband_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of experiments obtained from each proband in the ExperimentalSet.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_proteins_present_in_all">
<code class="sig-name descname">get_proteins_present_in_all</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_proteins_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the proteins that are inferred in all experiments of the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Proteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_tissue_counts">
<code class="sig-name descname">get_tissue_counts</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_tissue_counts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of experiments obtained from each tissue in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Counts</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_total_peptide_per_org_count">
<code class="sig-name descname">get_total_peptide_per_org_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_total_peptide_per_org_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The total count of peptides per organism accross the all experiments in the set.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_orgs">
<code class="sig-name descname">get_unique_orgs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_orgs" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list of the unique organisms in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_peptides">
<code class="sig-name descname">get_unique_peptides</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Peptide.Peptide" title="IPTK.Classes.Peptide.Peptide">IPTK.Classes.Peptide.Peptide</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_peptides" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The set of unique peptides in the experimentalSet</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Peptides</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_proteins">
<code class="sig-name descname">get_unique_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.get_unique_proteins" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the set of unique proteins in the experimentalset</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Proteins</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_proband">
<code class="sig-name descname">group_by_proband</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span><a class="reference internal" href="#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">IPTK.Classes.ExperimentalSet.ExperimentSet</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_proband" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a map between each proband and an Experimentalset object represent all the experiments objects         belonging to this proband.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,<a class="reference internal" href="#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">ExperimentSet</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_tissue">
<code class="sig-name descname">group_by_tissue</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span><a class="reference internal" href="#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">IPTK.Classes.ExperimentalSet.ExperimentSet</a><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.group_by_tissue" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A map between each tissue and an ExperimentalSet object representing all experiments belonging to that tissue.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,<a class="reference internal" href="#IPTK.Classes.ExperimentalSet.ExperimentSet" title="IPTK.Classes.ExperimentalSet.ExperimentSet">ExperimentSet</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.is_peptide_present_in_all">
<code class="sig-name descname">is_peptide_present_in_all</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">peptide</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.is_peptide_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>peptide</strong> (<em>str</em>) – The peptide sequence to search its occurrences in every experiment contained in the set</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if peptide is present in all experiments inside the instance, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.ExperimentalSet.ExperimentSet.is_protein_present_in_all">
<code class="sig-name descname">is_protein_present_in_all</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.ExperimentalSet.ExperimentSet.is_protein_present_in_all" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein</strong> (<em>str</em>) – the protein id to search its occurrences in every experimental in the set</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if peptide is present in all experiments inside the instance, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Features">
<span id="iptk-classes-features-module"></span><h2>IPTK.Classes.Features module<a class="headerlink" href="#module-IPTK.Classes.Features" title="Permalink to this headline">¶</a></h2>
<p>Parse the XML scheme of a uniprot protein and represent and provide a python API
for quering and accessing the results</p>
<dl class="py class">
<dt id="IPTK.Classes.Features.Features">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Features.</code><code class="sig-name descname">Features</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">uniprot_id</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">temp_dir</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Features.Features" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The class provide a template for the features associated with a protein. The following features are associated with the protein #signal peptide: dict </p>
<blockquote>
<div><p>the range of the signal peptide,if the protein has no signal, for example, a globular cytologic protein. None is used as a default, placeholder value. </p>
</div></blockquote>
<dl class="simple">
<dt>#chains:dict </dt><dd><p>the chains making up the mature protein, the protein should at least have one chain. </p>
</dd>
<dt>#domain: dict</dt><dd><p>the known domains in the protein, if no domain is defined, None is used. </p>
</dd>
<dt>#modification sites: nested dict </dt><dd><p>that contains information about the PTM site, glycosylation site and disulfide bonds. </p>
</dd>
<dt>#sequence variances: dict </dt><dd><p>which contains information about the sequence variant of a protein structure. </p>
</dd>
<dt>#split variance: dict </dt><dd><p>which contain known split variance </p>
</dd>
</dl>
<p>** Notes: Although disulfide bond is not a PTMs, it is being treated as a one here to simplify the workflow. </p>
<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_PTMs">
<code class="sig-name descname">get_PTMs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_PTMs" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>a dictionary the contains the PTMs found within the protein         the PTMs are classified into the main categories:</p>
<p>1- Modifications: which is the generic case and contain information             about any sequence modification beside disulfide bonds and glycosylation.</p>
<p>2- glycosylation: contain information about glycosylation sites</p>
<p>3- DisulfideBond: contain information about disulfide bond</p>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,Dict[str,Dict[str,Union[str,int]]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_PTMs_glycosylation">
<code class="sig-name descname">get_PTMs_glycosylation</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_PTMs_glycosylation" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The glycosylation sites found on the protein. If the protein has no glycosylation sites, the function returns None.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>[type]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_PTMs_modifications">
<code class="sig-name descname">get_PTMs_modifications</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_PTMs_modifications" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The generic modification found on the protein. If the protein has no PTM, the function returns None.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,Dict[str,Union[str,int]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_chains">
<code class="sig-name descname">get_chains</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_chains" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A dictionary the contains the chains of the protein, if no chain is defined it return None</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[Dict[str,Union[str,int]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_disulfide_bonds">
<code class="sig-name descname">get_disulfide_bonds</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_disulfide_bonds" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The disulfide sites found on the protein. If the protein has no disulfide sites, the function returns None</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>[type]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_domains">
<code class="sig-name descname">get_domains</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>int<span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_domains" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The domains defined in the protein sequence, if no domain is defined it returns None</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str, Dict[str, int]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_PTMs">
<code class="sig-name descname">get_number_PTMs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_PTMs" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of PTMs the sequence has, this include di-sulfide bonds See Note1 for more details.         If the protein has no PTMs the function returns zero</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_chains">
<code class="sig-name descname">get_number_chains</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_chains" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of chain/chains in the protein features. if no chain is defined it returns zero.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_disulfide_bonds">
<code class="sig-name descname">get_number_disulfide_bonds</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_disulfide_bonds" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of disulfide bonds the protein has, if the protein has no disulfide bonds the function return zero.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_domains">
<code class="sig-name descname">get_number_domains</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_domains" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of domains a protein has, if no domain is defined it returns zero.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_glycosylation_sites">
<code class="sig-name descname">get_number_glycosylation_sites</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_glycosylation_sites" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of glycosylation_sites the protein has, if the protein has no glycosylation sites, the function returns zero</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_modifications">
<code class="sig-name descname">get_number_modifications</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_modifications" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>returns the total number of generic modification found on the protein.         if no modification is found it return 0</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_sequence_variants">
<code class="sig-name descname">get_number_sequence_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_sequence_variants" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of sequence variants the protein has, if the protein has no sequence varient, the function returns 0.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_number_splice_variants">
<code class="sig-name descname">get_number_splice_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Features.Features.get_number_splice_variants" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The number of slice variants in the protein, if the protein has no splice variants, the function returns zero.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_sequence_variants">
<code class="sig-name descname">get_sequence_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_sequence_variants" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a dict object that contains all sequence variants within a protein, if the protein has no sequence variants the function returns None.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,Dict[str,Union[str,int]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_signal_peptide_index">
<code class="sig-name descname">get_signal_peptide_index</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Tuple<span class="p">[</span>int<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_signal_peptide_index" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The Index of the signal peptide in the protein, if not signal peptide is defined it returns None</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Tuple[int,int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.get_splice_variants">
<code class="sig-name descname">get_splice_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.get_splice_variants" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A dict object that contain the splice variants. if the protein has no splice variants the function returns None.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Dict[str,Dict[str,Union[str,int]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_PTMs">
<code class="sig-name descname">has_PTMs</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_PTMs" title="Permalink to this definition">¶</a></dt>
<dd><p>:return:True if the protein has a PTMs and False other wise
:rtype: bool</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_chains">
<code class="sig-name descname">has_chains</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_chains" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the protein has/have chain/chains as feature and False otherwise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>[type]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_disulfide_bond">
<code class="sig-name descname">has_disulfide_bond</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_disulfide_bond" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True is the protein has disulfide and False other wise</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_domains">
<code class="sig-name descname">has_domains</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_domains" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the protein has a defined domain/domains, otherwise it return False</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_glycosylation_site">
<code class="sig-name descname">has_glycosylation_site</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_glycosylation_site" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the protein has a glycosylation site and False otherwise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>[type]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_sequence_variants">
<code class="sig-name descname">has_sequence_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_sequence_variants" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the protein has a sequence variants, and False otherwise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_signal_peptide">
<code class="sig-name descname">has_signal_peptide</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_signal_peptide" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the protein has a signal peptide and False other wise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_site_modifications">
<code class="sig-name descname">has_site_modifications</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_site_modifications" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the protein has a modification site and False otherwise</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.has_splice_variants">
<code class="sig-name descname">has_splice_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Features.Features.has_splice_variants" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if the sequence has a splice variants and False otherwise.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Features.Features.summary">
<code class="sig-name descname">summary</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>int<span class="p">, </span>str<span class="p">]</span><span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Features.Features.summary" title="Permalink to this definition">¶</a></dt>
<dd><p>:return:The function return a dict object that summarizes the features of the protein.
:rtype: Dict[str,Union[str,int]]</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.HLAChain">
<span id="iptk-classes-hlachain-module"></span><h2>IPTK.Classes.HLAChain module<a class="headerlink" href="#module-IPTK.Classes.HLAChain" title="Permalink to this headline">¶</a></h2>
<p>The implementation of an HLA molecules</p>
<dl class="py class">
<dt id="IPTK.Classes.HLAChain.HLAChain">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.HLAChain.</code><code class="sig-name descname">HLAChain</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_allele_group">
<code class="sig-name descname">get_allele_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The allele group</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_chain_class">
<code class="sig-name descname">get_chain_class</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_chain_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_name</strong> (<em>str</em>) – the name of the gene</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>1 if the gene belongs to class one and 2 if it belong to class two</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_class">
<code class="sig-name descname">get_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The HLA class</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_gene">
<code class="sig-name descname">get_gene</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_gene" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the gene name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The chain name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAChain.HLAChain.get_protein_group">
<code class="sig-name descname">get_protein_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAChain.HLAChain.get_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the protein name</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.HLAMolecules">
<span id="iptk-classes-hlamolecules-module"></span><h2>IPTK.Classes.HLAMolecules module<a class="headerlink" href="#module-IPTK.Classes.HLAMolecules" title="Permalink to this headline">¶</a></h2>
<p>a representation of an HLA molecules</p>
<dl class="py class">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.HLAMolecules.</code><code class="sig-name descname">HLAMolecule</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">hla_chains</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_allele_group">
<code class="sig-name descname">get_allele_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the allele group for the instance chain/pair of chains</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>AlleleGroup</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_class">
<code class="sig-name descname">get_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The class of the HLA molecules</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_gene">
<code class="sig-name descname">get_gene</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_gene" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>gene/pair of genes coding for the current HLA molecules</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Genes</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">':'</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – the name of the allele by concatenating the names of the individual chains using         a separator, defaults to ‘:’</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>[description]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLAMolecules.HLAMolecule.get_protein_group">
<code class="sig-name descname">get_protein_group</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLAMolecules.HLAMolecule.get_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The protein group for the instance chain/pair of chains</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>ProteinGroup</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.HLASet">
<span id="iptk-classes-hlaset-module"></span><h2>IPTK.Classes.HLASet module<a class="headerlink" href="#module-IPTK.Classes.HLASet" title="Permalink to this headline">¶</a></h2>
<p>An abstraction for a collection of HLA alleles</p>
<dl class="py class">
<dt id="IPTK.Classes.HLASet.HLASet">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.HLASet.</code><code class="sig-name descname">HLASet</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">hlas</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">gene_sep</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">':'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.HLASet.HLASet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.get_alleles">
<code class="sig-name descname">get_alleles</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.get_alleles" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The class of the HLA-alleles in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.get_class">
<code class="sig-name descname">get_class</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.get_class" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The class of the HLA-alleles in the current instance</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.get_hla_count">
<code class="sig-name descname">get_hla_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.get_hla_count" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the count of HLA molecules in the set</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_allele">
<code class="sig-name descname">has_allele</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allele</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_allele" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>allele</strong> (<em>str</em>) – the name of the allele to check the instance for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if the provided allele is in the current instance, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_allele_group">
<code class="sig-name descname">has_allele_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allele_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_allele_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>allele_group</strong> (<em>str</em>) – the allele group to search the set for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if at least one allele in the set belongs to the provided allele group, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_gene">
<code class="sig-name descname">has_gene</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_gene" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_name</strong> (<em>str</em>) – the gene name to search the set against.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if at least one of the alleles in the set belongs to the provided gene. False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.HLASet.HLASet.has_protein_group">
<code class="sig-name descname">has_protein_group</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_group</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.HLASet.HLASet.has_protein_group" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>protein_group</strong> – The protein group to search the set for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True, if at least one allele in the set belongs to the provided protein group</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Peptide">
<span id="iptk-classes-peptide-module"></span><h2>IPTK.Classes.Peptide module<a class="headerlink" href="#module-IPTK.Classes.Peptide" title="Permalink to this headline">¶</a></h2>
<p>A representation of the eluted peptides and its identified proteins.</p>
<dl class="py class">
<dt id="IPTK.Classes.Peptide.Peptide">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Peptide.</code><code class="sig-name descname">Peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pep_seq</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An representation of an eluted peptide.</p>
<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.add_org_2_parent">
<code class="sig-name descname">add_org_2_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.add_org_2_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>add the source organism of one of the instance parent protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prot_name</strong> (<em>str</em>) – The name of the protein, i.e. the identifier of the protein</p></li>
<li><p><strong>org</strong> (<em>str</em>) – the name of the organism</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – incase the provided protein is not a parent of the provided peptide</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.add_parent_protein">
<code class="sig-name descname">add_parent_protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parent_protein</span></em>, <em class="sig-param"><span class="n">start_index</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">end_index</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.add_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>add a protein instance as a parent to the current peptide.
The library use Python-based indexing where its 0-indexed and ranges are treated as [start, end).
:param parent_protein: a Protein instance that act as a parent to the peptide.
:type parent_protein: Protein
:param start_index: the position in the parent protein where the peptide starts
:type start_index: int
:param end_index: the index of the amino acid that occurs after the last amino acid in the peptide,
:type start_index: int</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_c_terminal_flank_seq">
<code class="sig-name descname">get_c_terminal_flank_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_len</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_c_terminal_flank_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>:param flank_len:the length of the flanking regions
:type flank_len: int
:return: a list of string containing the sequences located downstream of the peptide in the parent protein.
:rtype: [type]</p>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_flanked_peptide">
<code class="sig-name descname">get_flanked_peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_len</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_flanked_peptide" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_len</strong> (<em>int</em>) – the length of the flanking regions</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A list of string containing the length of the peptide + the flanking region from               both the N and C terminal of the instance peptide, from all proteins.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_length">
<code class="sig-name descname">get_length</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_length" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the length of the peptides</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_n_terminal_flank_seq">
<code class="sig-name descname">get_n_terminal_flank_seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">flank_len</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_n_terminal_flank_seq" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_len</strong> (<em>int</em>) – the length of the flanking regions</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of string containing the sequences located upstream of the peptide in the parent protein.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>List[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_non_presented_peptides">
<code class="sig-name descname">get_non_presented_peptides</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_non_presented_peptides" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>length</strong> (<em>int</em>) – The length, i.e. number of amino acids, for the non-presented peptide</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>non-presented peptide from all the parent protein of the current peptide instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_number_of_parents">
<code class="sig-name descname">get_number_of_parents</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_number_of_parents" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of instance parent proteins</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_number_parent_protein">
<code class="sig-name descname">get_number_parent_protein</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_number_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the number of parent proteins this instance has</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_parent">
<code class="sig-name descname">get_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_parent" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – The protein identifer</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the parent protein that has an id matching the user defined pro_id</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#IPTK.Classes.Protein.Protein" title="IPTK.Classes.Protein.Protein">Protein</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_parent_proteins">
<code class="sig-name descname">get_parent_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_parent_proteins" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_parents_org">
<code class="sig-name descname">get_parents_org</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>str<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_parents_org" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a list containing the name of each parent protein source organisms</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Organisms</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_peptide_seq">
<code class="sig-name descname">get_peptide_seq</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_peptide_seq" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the sequence of the peptide.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.get_pos_in_parent">
<code class="sig-name descname">get_pos_in_parent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; Tuple<span class="p">[</span>int<span class="p">, </span>int<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.get_pos_in_parent" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – the id of the parent protein</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the identifier is not a parent of the instance</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the start and end position of the instance peptide in the parent pointed out by the provided identifier</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Range</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.is_child_of">
<code class="sig-name descname">is_child_of</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pro_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.is_child_of" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pro_id</strong> (<em>str</em>) – is the protein id</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if a user provided protein-id is a parent for the instance peptide, False otherwise</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Peptide.Peptide.map_to_parent_protein">
<code class="sig-name descname">map_to_parent_protein</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; List<span class="p">[</span>numpy.ndarray<span class="p">]</span><a class="headerlink" href="#IPTK.Classes.Peptide.Peptide.map_to_parent_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Mapped the instance peptide to the parent protein and returned a
list of numpy arrays where each array has a size of 1 by protein length.
within the protein the range representing the peptide is encoded as one while
the rest is zero.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of binary encoded arrays represent this mapping.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>MappedProtein</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Proband">
<span id="iptk-classes-proband-module"></span><h2>IPTK.Classes.Proband module<a class="headerlink" href="#module-IPTK.Classes.Proband" title="Permalink to this headline">¶</a></h2>
<p>A description for an IP proband</p>
<dl class="py class">
<dt id="IPTK.Classes.Proband.Proband">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Proband.</code><code class="sig-name descname">Proband</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">info</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Proband.Proband" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.Proband.Proband.get_meta_data">
<code class="sig-name descname">get_meta_data</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; dict<a class="headerlink" href="#IPTK.Classes.Proband.Proband.get_meta_data" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a dict that contain all the meta-data about the patient</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Proband.Proband.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Proband.Proband.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the patient</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Proband.Proband.update_info">
<code class="sig-name descname">update_info</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">info</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Proband.Proband.update_info" title="Permalink to this definition">¶</a></dt>
<dd><p>add new or update existing info about the patient  using an arbitrary number of key-value pair to be added to added to the instance meta-info dict</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Protein">
<span id="iptk-classes-protein-module"></span><h2>IPTK.Classes.Protein module<a class="headerlink" href="#module-IPTK.Classes.Protein" title="Permalink to this headline">¶</a></h2>
<p>A representation of a protein that has been inferred from an IP experiment.</p>
<dl class="py class">
<dt id="IPTK.Classes.Protein.Protein">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Protein.</code><code class="sig-name descname">Protein</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_id</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">seq</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>str<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Protein.Protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>representation of a protein that has been infered from an IP experiment.</p>
<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_id">
<code class="sig-name descname">get_id</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_id" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return the protein identifier.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_non_presented_peptide">
<code class="sig-name descname">get_non_presented_peptide</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">exc_reg_s_idx</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">exc_reg_e_idx</span><span class="p">:</span> <span class="n">int</span></em>, <em class="sig-param"><span class="n">length</span><span class="p">:</span> <span class="n">int</span></em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_non_presented_peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>sample a peptide from the protein sequences where the sampled peptides is not part of the               experimentally identified regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>exc_reg_s_idx</strong> (<em>int</em>) – the start point in the reference protein sequence of the experimentally identified peptide.</p></li>
<li><p><strong>exc_reg_e_idx</strong> (<em>int</em>) – the end point in the reference protein sequence of the experimentally identified peptide.</p></li>
<li><p><strong>length</strong> (<em>int</em>) – length the non-presented peptides.</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – if the length of the peptide is bigger than the protein length</p></li>
<li><p><strong>ValueError</strong> – if the length of the peptide is smaller than or equal to zero</p></li>
</ul>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a substring of the instance sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_org">
<code class="sig-name descname">get_org</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_org" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the organism in which this instance protein belong.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_peptides_map">
<code class="sig-name descname">get_peptides_map</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start_idxs</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">end_idxs</span><span class="p">:</span> <span class="n">List<span class="p">[</span>int<span class="p">]</span></span></em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_peptides_map" title="Permalink to this definition">¶</a></dt>
<dd><p>compute a coverage over the protein sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start_idxs</strong> (<em>Index</em>) – a list of integers representing the start positions</p></li>
<li><p><strong>end_idxs</strong> – a list of integers representing the end positions</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – if start_indxs and end_idxs MUST be of equal length are not of equal length</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A numpy array with shape of 1 by the length of the protein where every element in the array donates the number of times, It has been observed in the experiment.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.get_seq">
<code class="sig-name descname">get_seq</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Protein.Protein.get_seq" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the protein sequence.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Protein.Protein.set_org">
<code class="sig-name descname">set_org</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">org</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#IPTK.Classes.Protein.Protein.set_org" title="Permalink to this definition">¶</a></dt>
<dd><p>a post-instantitation mechanism to set the organism for which the protein belong.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>org</strong> (<em>str</em>) – the name of the organism</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes.Tissue">
<span id="iptk-classes-tissue-module"></span><h2>IPTK.Classes.Tissue module<a class="headerlink" href="#module-IPTK.Classes.Tissue" title="Permalink to this headline">¶</a></h2>
<p>A representation of the Tissue used in an IP Experiment.</p>
<dl class="py class">
<dt id="IPTK.Classes.Tissue.ExpressionProfile">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Tissue.</code><code class="sig-name descname">ExpressionProfile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">expression_table</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">aux_proteins</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span>pandas.core.frame.DataFrame<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>a representation of tissue reference expression value.</p>
<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_gene_id_expression">
<code class="sig-name descname">get_gene_id_expression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_id</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_gene_id_expression" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_id</strong> (<em>str</em>) – the gene id to retrive its expression value from the database</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided id is not defined in the instance table</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the expression value of the provided gene id.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_gene_name_expression">
<code class="sig-name descname">get_gene_name_expression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_gene_name_expression" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene_name</strong> (<em>str</em>) – the gene name to retrive its expression value from the database</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – if the provided id is not defined in the instance table</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the expression value of the provided gene name.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the tissue which the instance profile its gene expression</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.ExpressionProfile.get_table">
<code class="sig-name descname">get_table</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame<a class="headerlink" href="#IPTK.Classes.Tissue.ExpressionProfile.get_table" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>return a table that contain the expression of all the transcript in the current profile                including core and auxiliary proteins</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="IPTK.Classes.Tissue.Tissue">
<em class="property">class </em><code class="sig-prename descclassname">IPTK.Classes.Tissue.</code><code class="sig-name descname">Tissue</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">main_exp_value</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.Database.GeneExpressionDB" title="IPTK.Classes.Database.GeneExpressionDB">IPTK.Classes.Database.GeneExpressionDB</a></span></em>, <em class="sig-param"><span class="n">main_location</span><span class="p">:</span> <span class="n"><a class="reference internal" href="#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a></span></em>, <em class="sig-param"><span class="n">aux_exp_value</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Database.GeneExpressionDB" title="IPTK.Classes.Database.GeneExpressionDB">IPTK.Classes.Database.GeneExpressionDB</a><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">aux_location</span><span class="p">:</span> <span class="n">Optional<span class="p">[</span><a class="reference internal" href="#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#IPTK.Classes.Tissue.Tissue" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="IPTK.Classes.Tissue.Tissue.get_expression_profile">
<code class="sig-name descname">get_expression_profile</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#IPTK.Classes.Tissue.ExpressionProfile" title="IPTK.Classes.Tissue.ExpressionProfile">IPTK.Classes.Tissue.ExpressionProfile</a><a class="headerlink" href="#IPTK.Classes.Tissue.Tissue.get_expression_profile" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the expresion profile of the current tissue</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#IPTK.Classes.Tissue.ExpressionProfile" title="IPTK.Classes.Tissue.ExpressionProfile">ExpressionProfile</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.Tissue.get_name">
<code class="sig-name descname">get_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#IPTK.Classes.Tissue.Tissue.get_name" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the name of the tissue</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="IPTK.Classes.Tissue.Tissue.get_subCellular_locations">
<code class="sig-name descname">get_subCellular_locations</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; <a class="reference internal" href="#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">IPTK.Classes.Database.CellularLocationDB</a><a class="headerlink" href="#IPTK.Classes.Tissue.Tissue.get_subCellular_locations" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the sub-cellular localization of all the proteins stored in current instance resources.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#IPTK.Classes.Database.CellularLocationDB" title="IPTK.Classes.Database.CellularLocationDB">CellularLocationDB</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-IPTK.Classes">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-IPTK.Classes" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">IPTK</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html#contact">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="get_started.html">Get Started!</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">IPTK</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="IPTK.html">IPTK package</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="modules.html">IPTK</a><ul>
  <li><a href="IPTK.html">IPTK package</a><ul>
      <li>Previous: <a href="IPTK.Analysis.html" title="previous chapter">IPTK.Analysis package</a></li>
      <li>Next: <a href="IPTK.IO.html" title="next chapter">IPTK.IO package</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Hesham ElAbd.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/IPTK.Classes.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>